

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>API Reference &mdash; CoolBox  documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/gallery.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Customizing CoolBox" href="custom_elements.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> CoolBox
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation &amp; Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="quick_start_API.html">Quickstart (API)</a></li>
<li class="toctree-l1"><a class="reference internal" href="quick_start_CLI.html">Quickstart (CLI)</a></li>
<li class="toctree-l1"><a class="reference internal" href="gallery.html">CoolBox Gallery</a></li>
<li class="toctree-l1"><a class="reference internal" href="custom_elements.html">Customizing CoolBox</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">API Reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#module-coolbox.core.track">Track</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-coolbox.core.coverage">Coverage</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-coolbox.core.feature">Feature</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-coolbox.core.frame">Frame</a></li>
<li class="toctree-l2"><a class="reference internal" href="#jointview">JointView</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-coolbox.core.browser">Browser</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-coolbox.utilities">Utilities</a></li>
</ul>
</li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">CoolBox</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>API Reference</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/api.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="api-reference">
<h1>API Reference<a class="headerlink" href="#api-reference" title="Permalink to this headline">¶</a></h1>
<div class="section" id="module-coolbox.core.track">
<span id="track"></span><h2>Track<a class="headerlink" href="#module-coolbox.core.track" title="Permalink to this headline">¶</a></h2>
<p><strong>Functions:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.Arcs" title="coolbox.core.track.Arcs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Arcs</span></code></a>(file_, *args, **kwargs)</p></td>
<td><p>Compose BEDPE, Pairs track automatically based on type of file extension.(.bedpe, .pairs)</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.HiCMat" title="coolbox.core.track.HiCMat"><code class="xref py py-obj docutils literal notranslate"><span class="pre">HiCMat</span></code></a>(file, *args, **kwargs)</p></td>
<td><p>Compose DotHic or Cool track automatically based on tpye of file extension (.cool, .mcool, .hic)</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.Hist" title="coolbox.core.track.Hist"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Hist</span></code></a>(file, *args, **kwargs)</p></td>
<td><p>Compose BigWig, BedGraph, SNP, BAMCov track automatically based on type of file extension.(.bw, .bigwig, .snp, .vcf, .bedgraph, .bam, .sam)</p></td>
</tr>
</tbody>
</table>
<p><strong>Classes:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.ArcsBase" title="coolbox.core.track.ArcsBase"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ArcsBase</span></code></a>(*args, **kwargs)</p></td>
<td><p>Arcs(link) track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.BAM" title="coolbox.core.track.BAM"><code class="xref py py-obj docutils literal notranslate"><span class="pre">BAM</span></code></a>(*args, **kwargs)</p></td>
<td><p>BAM alignment track for plotting reads.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.BAMCov" title="coolbox.core.track.BAMCov"><code class="xref py py-obj docutils literal notranslate"><span class="pre">BAMCov</span></code></a>(*args, **kwargs)</p></td>
<td><p>Alignment reads coverage track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.BED" title="coolbox.core.track.BED"><code class="xref py py-obj docutils literal notranslate"><span class="pre">BED</span></code></a>(*args, **kwargs)</p></td>
<td><p>Bed Track for plotting 1d intervals data from .bed file.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.BEDPE" title="coolbox.core.track.BEDPE"><code class="xref py py-obj docutils literal notranslate"><span class="pre">BEDPE</span></code></a>(*args, **kwargs)</p></td>
<td><p>Arcs track from .bedpe file.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.BedBase" title="coolbox.core.track.BedBase"><code class="xref py py-obj docutils literal notranslate"><span class="pre">BedBase</span></code></a>(*args, **kwargs)</p></td>
<td><p>BED Base track.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.BedGraph" title="coolbox.core.track.BedGraph"><code class="xref py py-obj docutils literal notranslate"><span class="pre">BedGraph</span></code></a>(*args, **kwargs)</p></td>
<td><p>BedGraph track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.BigWig" title="coolbox.core.track.BigWig"><code class="xref py py-obj docutils literal notranslate"><span class="pre">BigWig</span></code></a>(*args, **kwargs)</p></td>
<td><p>BigWig track</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.ChromName" title="coolbox.core.track.ChromName"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ChromName</span></code></a>(*args, **kwargs)</p></td>
<td><p>Track for show chromosome name.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.Cool" title="coolbox.core.track.Cool"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Cool</span></code></a>(*args, **kwargs)</p></td>
<td><p>Cool track from .cool, .mcool file.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.DiScore" title="coolbox.core.track.DiScore"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DiScore</span></code></a>(*args, **kwargs)</p></td>
<td><p>Directionnality index track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.DotHiC" title="coolbox.core.track.DotHiC"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DotHiC</span></code></a>(*args, **kwargs)</p></td>
<td><p>HicMat track from .hic file.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.GTF" title="coolbox.core.track.GTF"><code class="xref py py-obj docutils literal notranslate"><span class="pre">GTF</span></code></a>(*args, **kwargs)</p></td>
<td><p>GTF gene annotation track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.HLine" title="coolbox.core.track.HLine"><code class="xref py py-obj docutils literal notranslate"><span class="pre">HLine</span></code></a>(*args, **kwargs)</p></td>
<td><p>Horizontal line track.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.HiCDiff" title="coolbox.core.track.HiCDiff"><code class="xref py py-obj docutils literal notranslate"><span class="pre">HiCDiff</span></code></a>(*args, **kwargs)</p></td>
<td><p>Track for express the comparison between two HiC Track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.HiCPeaks" title="coolbox.core.track.HiCPeaks"><code class="xref py py-obj docutils literal notranslate"><span class="pre">HiCPeaks</span></code></a>(*args, **kwargs)</p></td>
<td><p>Hi-C Peaks(Loops) from .bedpe file.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.HicMatBase" title="coolbox.core.track.HicMatBase"><code class="xref py py-obj docutils literal notranslate"><span class="pre">HicMatBase</span></code></a>(*args, **kwargs)</p></td>
<td><p>HicMatBase class for all track plotting matrix-like data.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.HistBase" title="coolbox.core.track.HistBase"><code class="xref py py-obj docutils literal notranslate"><span class="pre">HistBase</span></code></a>(*args, **kwargs)</p></td>
<td><p>HistBase track class</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.Ideogram" title="coolbox.core.track.Ideogram"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Ideogram</span></code></a>(*args, **kwargs)</p></td>
<td><p>The chromosome ideograme track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.InsuScore" title="coolbox.core.track.InsuScore"><code class="xref py py-obj docutils literal notranslate"><span class="pre">InsuScore</span></code></a>(*args, **kwargs)</p></td>
<td><p>Insulation score track.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.Pairs" title="coolbox.core.track.Pairs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Pairs</span></code></a>(*args, **kwargs)</p></td>
<td><p>Arcs track from .pairs file.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.SNP" title="coolbox.core.track.SNP"><code class="xref py py-obj docutils literal notranslate"><span class="pre">SNP</span></code></a>(*args, **kwargs)</p></td>
<td><p>Track for show SNPs Manhattan plot.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.Selfish" title="coolbox.core.track.Selfish"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Selfish</span></code></a>(*args, **kwargs)</p></td>
<td><p>Differential chromatin interaction.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.Spacer" title="coolbox.core.track.Spacer"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Spacer</span></code></a>(*args, **kwargs)</p></td>
<td><p>The spacer track Does not have any real content, just used to split two tracks.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.TAD" title="coolbox.core.track.TAD"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TAD</span></code></a>(*args, **kwargs)</p></td>
<td><p>Tad tack from bed file</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.Track" title="coolbox.core.track.Track"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Track</span></code></a>(*args, **kwargs)</p></td>
<td><p>Track base class.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.Virtual4C" title="coolbox.core.track.Virtual4C"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Virtual4C</span></code></a>(*args, **kwargs)</p></td>
<td><p>Track for view virtual 4C related to a certain genome position, and a HiC Track (include <cite>Cool</cite> and <cite>DotHiC</cite>).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.XAxis" title="coolbox.core.track.XAxis"><code class="xref py py-obj docutils literal notranslate"><span class="pre">XAxis</span></code></a>(*args, **kwargs)</p></td>
<td><p>The x axis track.</p></td>
</tr>
</tbody>
</table>
<dl class="py function">
<dt class="sig sig-object py" id="coolbox.core.track.Arcs">
<span class="sig-prename descclassname"><span class="pre">coolbox.core.track.</span></span><span class="sig-name descname"><span class="pre">Arcs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> &#x2192; <a class="reference internal" href="#coolbox.core.track.ArcsBase" title="coolbox.core.track.arcs.base.ArcsBase"><span class="pre">coolbox.core.track.arcs.base.ArcsBase</span></a><a class="headerlink" href="#coolbox.core.track.Arcs" title="Permalink to this definition">¶</a></dt>
<dd><p>Compose BEDPE, Pairs track automatically based on type of file extension.(.bedpe, .pairs)</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.track.ArcsBase">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.track.</span></span><span class="sig-name descname"><span class="pre">ArcsBase</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.ArcsBase" title="Permalink to this definition">¶</a></dt>
<dd><p>Arcs(link) track.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>style</strong><span class="classifier">str, optional</span></dt><dd><p>Style of arcs plot: ‘arcs’, ‘hicpeaks’, default ‘arcs’</p>
</dd>
<dt><strong>score_to_width</strong><span class="classifier">str, optional</span></dt><dd><p>Mapping function of score to width, default: ‘0.5 + math.sqrt(score)’</p>
</dd>
<dt><strong>line_width</strong><span class="classifier">float, optional</span></dt><dd><p>Width of arc line.</p>
</dd>
<dt><strong>open_region</strong><span class="classifier">bool, optional</span></dt><dd><p>If specified to True, will fetch the contacts on side in the region,
default True</p>
</dd>
<dt><strong>diameter_to_height</strong><span class="classifier">str, optional</span></dt><dd><p>Mapping function of arc diameter(interval end - start) to height.
You can specify to ‘max_diameter’ let all arcs has same height.
default ‘max_height * diameter / max_diameter’.</p>
</dd>
<dt><strong>orientation</strong><span class="classifier">str, optional</span></dt><dd><p>Track orientation, use ‘inverted’ for inverted track plot. default None</p>
</dd>
<dt><strong>line_style</strong><span class="classifier">str, optional</span></dt><dd><p>Border line style, default ‘solid’</p>
</dd>
<dt><strong>fill</strong><span class="classifier">bool, optional</span></dt><dd><p>Fill center or not, default False.</p>
</dd>
<dt><strong>fill_color</strong><span class="classifier">str, optional</span></dt><dd><p>Fill color, use ‘bed_rgb’ for specify color in file,
default ‘bed_rgb’.</p>
</dd>
<dt><strong>fill_alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Alpha value of fill region.
default 0.2</p>
</dd>
<dt><strong>side</strong><span class="classifier">{‘upper’, ‘lower’, ‘both’}</span></dt><dd><p>Plot peak in which side of the matrix.
NOTE: This parameters is useful only if the Cool track in matrix format.</p>
</dd>
<dt><strong>alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Alpha value of track, default 0.8.</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional;</span></dt><dd><p>Color of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_coverage</span></code>(coverage[, pos])</p></td>
<td><p>Append coverage to this track.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.ArcsBase.fetch_plot_data" title="coolbox.core.track.ArcsBase.fetch_plot_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code></a>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pile_coverages</span></code>(coverages[, pos])</p></td>
<td><p>Pile a stack of coverages with self’s coverages</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.ArcsBase.plot" title="coolbox.core.track.ArcsBase.plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code></a>(ax, gr, **kwargs)</p></td>
<td><p>Plot arc connecting two positions on a linear scale representing interactions between bins.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_arcs</span></code>(ax, gr, gr2, intervals)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_coverages</span></code>(ax, gr, gr2)</p></td>
<td><p>Plot all coverages on given axes.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_hicpeaks</span></code>(ax, gr, gr2, intervals)</p></td>
<td><p>plot hic peaks uppon a HicMatBase track.</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 63%" />
<col style="width: 37%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>has_prop</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_contacts</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_label</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.ArcsBase.fetch_plot_data" title="coolbox.core.track.ArcsBase.fetch_plot_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code></a>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.ArcsBase.plot" title="coolbox.core.track.ArcsBase.plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code></a>(ax, gr, **kwargs)</p></td>
<td><p>Plot arc connecting two positions on a linear scale representing interactions between bins.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.ArcsBase.fetch_plot_data">
<span class="sig-name descname"><span class="pre">fetch_plot_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> &#x2192; <span class="pre">pandas.core.frame.DataFrame</span><a class="headerlink" href="#coolbox.core.track.ArcsBase.fetch_plot_data" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd><p>GenomeRange is an object with properties of [chrom, start, end].</p>
</dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>intervals</strong><span class="classifier">pandas.core.frame.DataFrame</span></dt><dd><p>Can be two types:
1: with columns: [‘pos1’, ‘pos2’, ‘score’] ‘score’ is optional
2: with columns: [‘start1’, ‘end1’, ‘start2’, ‘end2’, ‘score’] ‘score’ is optional</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.ArcsBase.plot">
<span class="sig-name descname"><span class="pre">plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.ArcsBase.plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot arc connecting two positions on a linear scale representing interactions between bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.track.BAM">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.track.</span></span><span class="sig-name descname"><span class="pre">BAM</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.BAM" title="Permalink to this definition">¶</a></dt>
<dd><p>BAM alignment track for plotting reads.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>file</strong><span class="classifier">str</span></dt><dd><p>Path to .bam .sam file.</p>
</dd>
<dt><strong>length_ratio_thresh</strong><span class="classifier">float</span></dt><dd><p>Length ratio threshold of show alignments. (Default 0.01)</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional;</span></dt><dd><p>Color of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_coverage</span></code>(coverage[, pos])</p></td>
<td><p>Append coverage to this track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.BAM.fetch_data" title="coolbox.core.track.BAM.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code>(gr, **kwargs)</p></td>
<td><p>Fetch the plot data for the given GenomeRange.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pile_coverages</span></code>(coverages[, pos])</p></td>
<td><p>Pile a stack of coverages with self’s coverages</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.BAM.plot" title="coolbox.core.track.BAM.plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code></a>(ax, gr, **kwargs)</p></td>
<td><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_coverages</span></code>(ax, gr, gr2)</p></td>
<td><p>Plot all coverages on given axes.</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 66%" />
<col style="width: 34%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>fetch_intervals</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>has_prop</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_align</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_label</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.BAM.fetch_data" title="coolbox.core.track.BAM.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.BAM.plot" title="coolbox.core.track.BAM.plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code></a>(ax, gr, **kwargs)</p></td>
<td><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.BAM.fetch_data">
<span class="sig-name descname"><span class="pre">fetch_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> &#x2192; <span class="pre">pandas.core.frame.DataFrame</span><a class="headerlink" href="#coolbox.core.track.BAM.fetch_data" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl>
<dt><strong>intervals</strong><span class="classifier">pandas.core.frame.DataFrame</span></dt><dd><p>Sam interval table.
The DataFrame table should has columns like:</p>
<dl class="simple">
<dt>columns = [“qname”, “flag”, “rname”, “pos”, “mapq”, “cigar”,</dt><dd><p>“rnext”, “pnext”, “tlen”, “seq”, “qual”, “options”]</p>
</dd>
</dl>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.BAM.plot">
<span class="sig-name descname"><span class="pre">plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.BAM.plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.track.BAMCov">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.track.</span></span><span class="sig-name descname"><span class="pre">BAMCov</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.BAMCov" title="Permalink to this definition">¶</a></dt>
<dd><p>Alignment reads coverage track.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>file: str</strong></dt><dd><p>File path of bam file.</p>
</dd>
<dt><strong>num_bins: int, optional</strong></dt><dd><p>Number of bins to plot hist fig.</p>
</dd>
<dt><strong>style</strong><span class="classifier">str, optional</span></dt><dd><p>Track graph type, format {‘line’, ‘fill’, ‘heatmap’, ‘scatter’}</p>
</dd>
<dt><strong>fmt</strong><span class="classifier">str, optional</span></dt><dd><p>Line styles used when the plot style is <cite>line</cite>, will be passed as <cite>fmt</cite> parameter in <cite>matplotlib.pyplot.plot</cite></p>
</dd>
<dt><strong>line_width</strong><span class="classifier">int, optional</span></dt><dd><p>Value used when the plot style is <cite>line</cite>, will be passed as <cite>linewidth</cite> parameter in <cite>matplotlib.pyplot.plot</cite></p>
</dd>
<dt><strong>size</strong><span class="classifier">int, optional</span></dt><dd><p>Value used when the plot style is <cite>scatter</cite>, will be passed as <cite>s</cite> parameter in <cite>matplotlib.pyplot.scatter</cite></p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional</span></dt><dd><p>Main color</p>
</dd>
<dt><strong>threshold_color</strong><span class="classifier">str, optional</span></dt><dd><p>Track’s color for values greater than specified threshold.</p>
</dd>
<dt><strong>threshold</strong><span class="classifier">float, optional</span></dt><dd><p>Threshold used when the plot style is <cite>line</cite> or <cite>scatter</cite>, values greater than this thresh will be
plotted with <cite>color = threshold_color</cite></p>
</dd>
<dt><strong>cmap: str, optional</strong></dt><dd><p>Cmap used when the plot type is heatmap, will be passed as <cite>cmap</cite> paramerter in <cite>matplotlib.pyplot.matshow</cite></p>
</dd>
<dt><strong>alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Alpha value of plot, default 1.0</p>
</dd>
<dt><strong>orientation</strong><span class="classifier">str, optional</span></dt><dd><p>Track orientation, use ‘inverted’ for inverted track plot.</p>
</dd>
<dt><strong>data_range_style</strong><span class="classifier">{bool, ‘text’, ‘y-axis’}, optional</span></dt><dd><p>Show_data_range or not, default True.</p>
</dd>
<dt><strong>max_value</strong><span class="classifier">{float, ‘auto’}, optional</span></dt><dd><p>Max value of track. ‘auto’ for specify max value automatically, default ‘auto’.</p>
</dd>
<dt><strong>min_value</strong><span class="classifier">{float, ‘auto’}, optional</span></dt><dd><p>Min value of track. ‘auto’ for specify max value automatically, default ‘auto’.</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_coverage</span></code>(coverage[, pos])</p></td>
<td><p>Append coverage to this track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.BAMCov.fetch_data" title="coolbox.core.track.BAMCov.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pile_coverages</span></code>(coverages[, pos])</p></td>
<td><p>Pile a stack of coverages with self’s coverages</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code>(ax, gr, **kwargs)</p></td>
<td><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_coverages</span></code>(ax, gr, gr2)</p></td>
<td><p>Plot all coverages on given axes.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_yaxis_range</span></code>(plot_axis, y_ax)</p></td>
<td><p>Plot the scale of the y axis with respect to the plot_axis</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 66%" />
<col style="width: 34%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>adjust_plot</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>fetch_coverage</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>has_prop</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_data_range</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_fill</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_heatmap</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_hist</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_label</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_line</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_scatter</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_stairs</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_text_range</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.BAMCov.fetch_data" title="coolbox.core.track.BAMCov.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.BAMCov.fetch_data">
<span class="sig-name descname"><span class="pre">fetch_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> &#x2192; <span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">numpy.ndarray</span><span class="p"><span class="pre">,</span> </span><span class="pre">numpy.ndarray</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span> </span><span class="pre">pandas.core.frame.DataFrame</span><span class="p"><span class="pre">,</span> </span><span class="pre">numpy.ndarray</span><span class="p"><span class="pre">]</span></span><a class="headerlink" href="#coolbox.core.track.BAMCov.fetch_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetch the raw data for the given GenomeRange.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.track.BED">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.track.</span></span><span class="sig-name descname"><span class="pre">BED</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.BED" title="Permalink to this definition">¶</a></dt>
<dd><p>Bed Track for plotting 1d intervals data from .bed file.
The input bed file can be bed3/bed6/bed9/bed12</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gene_style: {‘flybase’, ‘normal’}</strong></dt><dd></dd>
<dt><strong>display</strong><span class="classifier">{‘stacked’, ‘interlaced’, ‘collapsed’}, optional</span></dt><dd><p>Display mode. (Default: ‘stacked’)</p>
</dd>
<dt><strong>fontsize</strong><span class="classifier">int, optional</span></dt><dd><p>Font size. (Default: BED.DEFAULT_FONTSIZE)</p>
</dd>
<dt><strong>labels</strong><span class="classifier">{True, False, ‘auto’}, optional</span></dt><dd><p>Draw bed name or not. ‘auto’ for automate decision according to density.
(Default: ‘auto’)</p>
</dd>
<dt><strong>interval_height</strong><span class="classifier">int, optional</span></dt><dd><p>The height of the interval. (Default: 100)</p>
</dd>
<dt><strong>num_rows</strong><span class="classifier">int, optional</span></dt><dd><p>Set the max interval rows. (Default: unlimited interval rows)</p>
</dd>
<dt><strong>row_height</strong><span class="classifier">float</span></dt><dd><p>Height of a row. default 0.5</p>
</dd>
<dt><strong>file: str</strong></dt><dd><p>The file path of <cite>.bed</cite> file.</p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional</span></dt><dd><p>Track color, ‘bed_rgb’ for auto specify color according to bed record.
(Default: ‘bed_rgb’)</p>
</dd>
<dt><strong>border_color</strong><span class="classifier">str, optional</span></dt><dd><p>Border_color of gene. (Default: ‘black’)</p>
</dd>
<dt><strong>max_value</strong><span class="classifier">float, optional</span></dt><dd><p>Max score. (Default: inf)</p>
</dd>
<dt><strong>min_value</strong><span class="classifier">float, optional</span></dt><dd><p>Min score. (Default: -inf)</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_coverage</span></code>(coverage[, pos])</p></td>
<td><p>Append coverage to this track.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">draw_gene_simple</span></code>(ax, bed, ypos, rgb, edgecolor)</p></td>
<td><p>draws an interval with direction (if given)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">draw_gene_with_introns</span></code>(ax, bed, ypos, rgb, …)</p></td>
<td><p>draws a gene like in flybase gbrowse.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">draw_gene_with_introns_flybase_style</span></code>(ax, …)</p></td>
<td><p>draws a gene using different styles</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_intervals</span></code>(bgz_file, gr)</p></td>
<td><p>Fetch intervals within input chromosome range.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code>(gr, **kwargs)</p></td>
<td><p>Fetch the plot data for the given GenomeRange.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_y_pos</span></code>(free_row)</p></td>
<td><p>The y_pos is set such that regions to be plotted do not overlap (stacked).</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">intervals2dataframe</span></code>(intervals, bed_type)</p></td>
<td><p>Convert intervals list to pandas.DataFrame</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pile_coverages</span></code>(coverages[, pos])</p></td>
<td><p>Pile a stack of coverages with self’s coverages</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.BED.plot" title="coolbox.core.track.BED.plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code></a>(ax, gr, **kwargs)</p></td>
<td><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_coverages</span></code>(ax, gr, gr2)</p></td>
<td><p>Plot all coverages on given axes.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_colormap</span></code>(df)</p></td>
<td><p>As min_score and max_score change every plot, we compute them for every plot</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 72%" />
<col style="width: 28%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>draw_arrow</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>get_rgb_and_edge_color</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>get_track_height</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>has_prop</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>infer_bed_type</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>init_colormap</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>load_range</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_genes</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_label</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.BED.plot" title="coolbox.core.track.BED.plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code></a>(ax, gr, **kwargs)</p></td>
<td><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.BED.plot">
<span class="sig-name descname"><span class="pre">plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.BED.plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.track.BEDPE">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.track.</span></span><span class="sig-name descname"><span class="pre">BEDPE</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.BEDPE" title="Permalink to this definition">¶</a></dt>
<dd><p>Arcs track from .bedpe file.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>file: str</strong></dt><dd><p>Path of .bedpe file</p>
</dd>
<dt><strong>pos</strong><span class="classifier">str, optional</span></dt><dd><p>Method for choosing arch anchor for bedpe data: ‘start’, ‘end’, ‘mid’, default ‘mid’</p>
</dd>
<dt><strong>style</strong><span class="classifier">str, optional</span></dt><dd><p>Style of arcs plot: ‘arcs’, ‘hicpeaks’, default ‘arcs’</p>
</dd>
<dt><strong>score_to_width</strong><span class="classifier">str, optional</span></dt><dd><p>Mapping function of score to width, default: ‘0.5 + math.sqrt(score)’</p>
</dd>
<dt><strong>line_width</strong><span class="classifier">float, optional</span></dt><dd><p>Width of arc line.</p>
</dd>
<dt><strong>open_region</strong><span class="classifier">bool, optional</span></dt><dd><p>If specified to True, will fetch the contacts on side in the region,
default True</p>
</dd>
<dt><strong>diameter_to_height</strong><span class="classifier">str, optional</span></dt><dd><p>Mapping function of arc diameter(interval end - start) to height.
You can specify to ‘max_diameter’ let all arcs has same height.
default ‘max_height * diameter / max_diameter’.</p>
</dd>
<dt><strong>orientation</strong><span class="classifier">str, optional</span></dt><dd><p>Track orientation, use ‘inverted’ for inverted track plot. default None</p>
</dd>
<dt><strong>line_style</strong><span class="classifier">str, optional</span></dt><dd><p>Border line style, default ‘solid’</p>
</dd>
<dt><strong>fill</strong><span class="classifier">bool, optional</span></dt><dd><p>Fill center or not, default False.</p>
</dd>
<dt><strong>fill_color</strong><span class="classifier">str, optional</span></dt><dd><p>Fill color, use ‘bed_rgb’ for specify color in file,
default ‘bed_rgb’.</p>
</dd>
<dt><strong>fill_alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Alpha value of fill region.
default 0.2</p>
</dd>
<dt><strong>side</strong><span class="classifier">{‘upper’, ‘lower’, ‘both’}</span></dt><dd><p>Plot peak in which side of the matrix.
NOTE: This parameters is useful only if the Cool track in matrix format.</p>
</dd>
<dt><strong>alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Alpha value of track, default 0.8.</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional;</span></dt><dd><p>Color of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_coverage</span></code>(coverage[, pos])</p></td>
<td><p>Append coverage to this track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.BEDPE.fetch_data" title="coolbox.core.track.BEDPE.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.BEDPE.fetch_plot_data" title="coolbox.core.track.BEDPE.fetch_plot_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code></a>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pile_coverages</span></code>(coverages[, pos])</p></td>
<td><p>Pile a stack of coverages with self’s coverages</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code>(ax, gr, **kwargs)</p></td>
<td><p>Plot arc connecting two positions on a linear scale representing interactions between bins.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_arcs</span></code>(ax, gr, gr2, intervals)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_coverages</span></code>(ax, gr, gr2)</p></td>
<td><p>Plot all coverages on given axes.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_hicpeaks</span></code>(ax, gr, gr2, intervals)</p></td>
<td><p>plot hic peaks uppon a HicMatBase track.</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 66%" />
<col style="width: 34%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>fetch_intervals</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>has_prop</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_contacts</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_label</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.BEDPE.fetch_data" title="coolbox.core.track.BEDPE.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.BEDPE.fetch_plot_data" title="coolbox.core.track.BEDPE.fetch_plot_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code></a>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.BEDPE.fetch_data">
<span class="sig-name descname"><span class="pre">fetch_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> &#x2192; <span class="pre">pandas.core.frame.DataFrame</span><a class="headerlink" href="#coolbox.core.track.BEDPE.fetch_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetch the raw data for the given GenomeRange.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.BEDPE.fetch_plot_data">
<span class="sig-name descname"><span class="pre">fetch_plot_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> &#x2192; <span class="pre">pandas.core.frame.DataFrame</span><a class="headerlink" href="#coolbox.core.track.BEDPE.fetch_plot_data" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd><p>GenomeRange is an object with properties of [chrom, start, end].</p>
</dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>intervals</strong><span class="classifier">pandas.core.frame.DataFrame</span></dt><dd><p>Can be two types:
1: with columns: [‘pos1’, ‘pos2’, ‘score’] ‘score’ is optional
2: with columns: [‘start1’, ‘end1’, ‘start2’, ‘end2’, ‘score’] ‘score’ is optional</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.track.BedBase">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.track.</span></span><span class="sig-name descname"><span class="pre">BedBase</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.BedBase" title="Permalink to this definition">¶</a></dt>
<dd><p>BED Base track.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>file: str</strong></dt><dd><p>The file path of <cite>.bed</cite> file.</p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional</span></dt><dd><p>Track color, ‘bed_rgb’ for auto specify color according to bed record.
(Default: ‘bed_rgb’)</p>
</dd>
<dt><strong>border_color</strong><span class="classifier">str, optional</span></dt><dd><p>Border_color of gene. (Default: ‘black’)</p>
</dd>
<dt><strong>max_value</strong><span class="classifier">float, optional</span></dt><dd><p>Max score. (Default: inf)</p>
</dd>
<dt><strong>min_value</strong><span class="classifier">float, optional</span></dt><dd><p>Min score. (Default: -inf)</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_coverage</span></code>(coverage[, pos])</p></td>
<td><p>Append coverage to this track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.BedBase.fetch_data" title="coolbox.core.track.BedBase.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code>(gr, **kwargs)</p></td>
<td><p>Fetch the plot data for the given GenomeRange.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pile_coverages</span></code>(coverages[, pos])</p></td>
<td><p>Pile a stack of coverages with self’s coverages</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code>(ax, gr, **kwargs)</p></td>
<td><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_coverages</span></code>(ax, gr, gr2)</p></td>
<td><p>Plot all coverages on given axes.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.BedBase.set_colormap" title="coolbox.core.track.BedBase.set_colormap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_colormap</span></code></a>(df)</p></td>
<td><p>As min_score and max_score change every plot, we compute them for every plot</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 72%" />
<col style="width: 28%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>get_rgb_and_edge_color</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>has_prop</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>infer_bed_type</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>init_colormap</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_label</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.BedBase.fetch_data" title="coolbox.core.track.BedBase.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.BedBase.set_colormap" title="coolbox.core.track.BedBase.set_colormap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_colormap</span></code></a>(df)</p></td>
<td><p>As min_score and max_score change every plot, we compute them for every plot</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.BedBase.fetch_data">
<span class="sig-name descname"><span class="pre">fetch_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> &#x2192; <span class="pre">pandas.core.frame.DataFrame</span><a class="headerlink" href="#coolbox.core.track.BedBase.fetch_data" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl>
<dt><strong>intervals</strong><span class="classifier">pandas.core.frame.DataFrame</span></dt><dd><p>BED interval table. The table should be in format like:</p>
<dl class="simple">
<dt>bed_fields = [‘chromosome’, ‘start’, ‘end’,</dt><dd><p>‘name’, ‘score’, ‘strand’,
‘thick_start’, ‘thick_end’,
‘rgb’, ‘block_count’,
‘block_sizes’, ‘block_starts’]</p>
</dd>
</dl>
<p>The table can be in bed6/bed9/bed12 format and the trailing columns can be omited.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.BedBase.set_colormap">
<span class="sig-name descname"><span class="pre">set_colormap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.BedBase.set_colormap" title="Permalink to this definition">¶</a></dt>
<dd><p>As min_score and max_score change every plot, we compute them for every plot</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.track.BedGraph">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.track.</span></span><span class="sig-name descname"><span class="pre">BedGraph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.BedGraph" title="Permalink to this definition">¶</a></dt>
<dd><p>BedGraph track.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>file</strong><span class="classifier">str</span></dt><dd><p>File path of bedgraph file.</p>
</dd>
<dt><strong>style</strong><span class="classifier">str, optional</span></dt><dd><p>Track graph type, format {‘line’, ‘fill’, ‘heatmap’, ‘scatter’}</p>
</dd>
<dt><strong>fmt</strong><span class="classifier">str, optional</span></dt><dd><p>Line styles used when the plot style is <cite>line</cite>, will be passed as <cite>fmt</cite> parameter in <cite>matplotlib.pyplot.plot</cite></p>
</dd>
<dt><strong>line_width</strong><span class="classifier">int, optional</span></dt><dd><p>Value used when the plot style is <cite>line</cite>, will be passed as <cite>linewidth</cite> parameter in <cite>matplotlib.pyplot.plot</cite></p>
</dd>
<dt><strong>size</strong><span class="classifier">int, optional</span></dt><dd><p>Value used when the plot style is <cite>scatter</cite>, will be passed as <cite>s</cite> parameter in <cite>matplotlib.pyplot.scatter</cite></p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional</span></dt><dd><p>Main color</p>
</dd>
<dt><strong>threshold_color</strong><span class="classifier">str, optional</span></dt><dd><p>Track’s color for values greater than specified threshold.</p>
</dd>
<dt><strong>threshold</strong><span class="classifier">float, optional</span></dt><dd><p>Threshold used when the plot style is <cite>line</cite> or <cite>scatter</cite>, values greater than this thresh will be
plotted with <cite>color = threshold_color</cite></p>
</dd>
<dt><strong>cmap: str, optional</strong></dt><dd><p>Cmap used when the plot type is heatmap, will be passed as <cite>cmap</cite> paramerter in <cite>matplotlib.pyplot.matshow</cite></p>
</dd>
<dt><strong>alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Alpha value of plot, default 1.0</p>
</dd>
<dt><strong>orientation</strong><span class="classifier">str, optional</span></dt><dd><p>Track orientation, use ‘inverted’ for inverted track plot.</p>
</dd>
<dt><strong>data_range_style</strong><span class="classifier">{bool, ‘text’, ‘y-axis’}, optional</span></dt><dd><p>Show_data_range or not, default True.</p>
</dd>
<dt><strong>max_value</strong><span class="classifier">{float, ‘auto’}, optional</span></dt><dd><p>Max value of track. ‘auto’ for specify max value automatically, default ‘auto’.</p>
</dd>
<dt><strong>min_value</strong><span class="classifier">{float, ‘auto’}, optional</span></dt><dd><p>Min value of track. ‘auto’ for specify max value automatically, default ‘auto’.</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_coverage</span></code>(coverage[, pos])</p></td>
<td><p>Append coverage to this track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.BedGraph.fetch_data" title="coolbox.core.track.BedGraph.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.BedGraph.fetch_plot_data" title="coolbox.core.track.BedGraph.fetch_plot_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code></a>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pile_coverages</span></code>(coverages[, pos])</p></td>
<td><p>Pile a stack of coverages with self’s coverages</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code>(ax, gr, **kwargs)</p></td>
<td><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_coverages</span></code>(ax, gr, gr2)</p></td>
<td><p>Plot all coverages on given axes.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_yaxis_range</span></code>(plot_axis, y_ax)</p></td>
<td><p>Plot the scale of the y axis with respect to the plot_axis</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 66%" />
<col style="width: 34%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>adjust_plot</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>has_prop</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>load</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_data_range</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_fill</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_heatmap</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_hist</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_label</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_line</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_scatter</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_stairs</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_text_range</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.BedGraph.fetch_data" title="coolbox.core.track.BedGraph.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.BedGraph.fetch_plot_data" title="coolbox.core.track.BedGraph.fetch_plot_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code></a>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.BedGraph.fetch_data">
<span class="sig-name descname"><span class="pre">fetch_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> &#x2192; <span class="pre">pandas.core.frame.DataFrame</span><a class="headerlink" href="#coolbox.core.track.BedGraph.fetch_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetch the raw data for the given GenomeRange.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.BedGraph.fetch_plot_data">
<span class="sig-name descname"><span class="pre">fetch_plot_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> &#x2192; <span class="pre">pandas.core.frame.DataFrame</span><a class="headerlink" href="#coolbox.core.track.BedGraph.fetch_plot_data" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd><p>GenomeRange is an object with properties of [chrom, start, end].</p>
</dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>hist_data: Union[Tuple[np.ndarray, np.ndarray], pd.DataFrame, np.ndarray]</dt><dd><p>data used for plotting.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.track.BigWig">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.track.</span></span><span class="sig-name descname"><span class="pre">BigWig</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.BigWig" title="Permalink to this definition">¶</a></dt>
<dd><p>BigWig track</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>file</strong><span class="classifier">str</span></dt><dd><p>File path of bigwig file.</p>
</dd>
<dt><strong>num_bins</strong><span class="classifier">int, optional</span></dt><dd><p>Number of bins to plot the hist in current range, default 700.</p>
</dd>
<dt><strong>style</strong><span class="classifier">str, optional</span></dt><dd><p>Track graph type, format {‘line’, ‘fill’, ‘heatmap’, ‘scatter’}</p>
</dd>
<dt><strong>fmt</strong><span class="classifier">str, optional</span></dt><dd><p>Line styles used when the plot style is <cite>line</cite>, will be passed as <cite>fmt</cite> parameter in <cite>matplotlib.pyplot.plot</cite></p>
</dd>
<dt><strong>line_width</strong><span class="classifier">int, optional</span></dt><dd><p>Value used when the plot style is <cite>line</cite>, will be passed as <cite>linewidth</cite> parameter in <cite>matplotlib.pyplot.plot</cite></p>
</dd>
<dt><strong>size</strong><span class="classifier">int, optional</span></dt><dd><p>Value used when the plot style is <cite>scatter</cite>, will be passed as <cite>s</cite> parameter in <cite>matplotlib.pyplot.scatter</cite></p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional</span></dt><dd><p>Main color</p>
</dd>
<dt><strong>threshold_color</strong><span class="classifier">str, optional</span></dt><dd><p>Track’s color for values greater than specified threshold.</p>
</dd>
<dt><strong>threshold</strong><span class="classifier">float, optional</span></dt><dd><p>Threshold used when the plot style is <cite>line</cite> or <cite>scatter</cite>, values greater than this thresh will be
plotted with <cite>color = threshold_color</cite></p>
</dd>
<dt><strong>cmap: str, optional</strong></dt><dd><p>Cmap used when the plot type is heatmap, will be passed as <cite>cmap</cite> paramerter in <cite>matplotlib.pyplot.matshow</cite></p>
</dd>
<dt><strong>alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Alpha value of plot, default 1.0</p>
</dd>
<dt><strong>orientation</strong><span class="classifier">str, optional</span></dt><dd><p>Track orientation, use ‘inverted’ for inverted track plot.</p>
</dd>
<dt><strong>data_range_style</strong><span class="classifier">{bool, ‘text’, ‘y-axis’}, optional</span></dt><dd><p>Show_data_range or not, default True.</p>
</dd>
<dt><strong>max_value</strong><span class="classifier">{float, ‘auto’}, optional</span></dt><dd><p>Max value of track. ‘auto’ for specify max value automatically, default ‘auto’.</p>
</dd>
<dt><strong>min_value</strong><span class="classifier">{float, ‘auto’}, optional</span></dt><dd><p>Min value of track. ‘auto’ for specify max value automatically, default ‘auto’.</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_coverage</span></code>(coverage[, pos])</p></td>
<td><p>Append coverage to this track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.BigWig.fetch_data" title="coolbox.core.track.BigWig.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.BigWig.fetch_plot_data" title="coolbox.core.track.BigWig.fetch_plot_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code></a>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.BigWig.fetch_scores" title="coolbox.core.track.BigWig.fetch_scores"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_scores</span></code></a>(genome_range, num_bins[, …])</p></td>
<td><p>Fetch bins scores within input chromosome range.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pile_coverages</span></code>(coverages[, pos])</p></td>
<td><p>Pile a stack of coverages with self’s coverages</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code>(ax, gr, **kwargs)</p></td>
<td><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_coverages</span></code>(ax, gr, gr2)</p></td>
<td><p>Plot all coverages on given axes.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_yaxis_range</span></code>(plot_axis, y_ax)</p></td>
<td><p>Plot the scale of the y axis with respect to the plot_axis</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 67%" />
<col style="width: 33%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>adjust_plot</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>check_chrom_name</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>get_num_bins</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>has_prop</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_data_range</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_fill</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_heatmap</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_hist</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_label</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_line</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_scatter</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_stairs</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_text_range</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.BigWig.fetch_data" title="coolbox.core.track.BigWig.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.BigWig.fetch_plot_data" title="coolbox.core.track.BigWig.fetch_plot_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code></a>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.BigWig.fetch_scores" title="coolbox.core.track.BigWig.fetch_scores"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_scores</span></code></a>(genome_range, num_bins[, …])</p></td>
<td><p>Fetch bins scores within input chromosome range.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.BigWig.fetch_data">
<span class="sig-name descname"><span class="pre">fetch_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.BigWig.fetch_data" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr</strong><span class="classifier">GenomeRange</span></dt><dd></dd>
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.BigWig.fetch_plot_data">
<span class="sig-name descname"><span class="pre">fetch_plot_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.BigWig.fetch_plot_data" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd><p>GenomeRange is an object with properties of [chrom, start, end].</p>
</dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>hist_data: Union[Tuple[np.ndarray, np.ndarray], pd.DataFrame, np.ndarray]</dt><dd><p>data used for plotting.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.BigWig.fetch_scores">
<span class="sig-name descname"><span class="pre">fetch_scores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome_range</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_bins</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_try_nums</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.BigWig.fetch_scores" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetch bins scores within input chromosome range.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.track.ChromName">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.track.</span></span><span class="sig-name descname"><span class="pre">ChromName</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.ChromName" title="Permalink to this definition">¶</a></dt>
<dd><p>Track for show chromosome name.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>fontsize</strong><span class="classifier">float</span></dt><dd><p>Font name to show.</p>
</dd>
<dt><strong>offset</strong><span class="classifier">float</span></dt><dd><p>Offset ratio to the start position.</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional;</span></dt><dd><p>Color of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_coverage</span></code>(coverage[, pos])</p></td>
<td><p>Append coverage to this track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.ChromName.fetch_data" title="coolbox.core.track.ChromName.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code>(gr, **kwargs)</p></td>
<td><p>Fetch the plot data for the given GenomeRange.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pile_coverages</span></code>(coverages[, pos])</p></td>
<td><p>Pile a stack of coverages with self’s coverages</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.ChromName.plot" title="coolbox.core.track.ChromName.plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code></a>(ax, gr, **kwargs)</p></td>
<td><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_coverages</span></code>(ax, gr, gr2)</p></td>
<td><p>Plot all coverages on given axes.</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 58%" />
<col style="width: 42%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>has_prop</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_label</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.ChromName.fetch_data" title="coolbox.core.track.ChromName.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.ChromName.plot" title="coolbox.core.track.ChromName.plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code></a>(ax, gr, **kwargs)</p></td>
<td><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.ChromName.fetch_data">
<span class="sig-name descname"><span class="pre">fetch_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.ChromName.fetch_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetch the raw data for the given GenomeRange.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.ChromName.plot">
<span class="sig-name descname"><span class="pre">plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.ChromName.plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.track.Cool">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.track.</span></span><span class="sig-name descname"><span class="pre">Cool</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.Cool" title="Permalink to this definition">¶</a></dt>
<dd><p>Cool track from .cool, .mcool file.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>file: str</strong></dt><dd><p>The file path of .cool, .mcool file</p>
</dd>
<dt><strong>balance: bool</strong></dt><dd><p>If use the balanced contact matrix.</p>
</dd>
<dt><strong>style</strong><span class="classifier">{‘triangular’, ‘window’, ‘matrix’}, optional</span></dt><dd><p>Matrix style, default ‘window’.</p>
</dd>
<dt><strong>cmap</strong><span class="classifier">str, optional</span></dt><dd><p>Color map of hic matrix, default “JuiceBoxLike”.</p>
</dd>
<dt><strong>color_bar</strong><span class="classifier">{‘vertical’, ‘horizontal’, ‘no’}, optional</span></dt><dd><p>Color bar style. default ‘vertical’.</p>
</dd>
<dt><strong>depth_ratio</strong><span class="classifier">float, optional</span></dt><dd><p>Depth ratio of triangular matrix, use ‘full’ for full depth. default ‘full’.</p>
</dd>
<dt><strong>max_value</strong><span class="classifier">{float, ‘auto’}, optional</span></dt><dd><p>Max value of hic matrix, use ‘auto’ for specify max value automatically, default ‘auto’.</p>
</dd>
<dt><strong>min_value</strong><span class="classifier">{float, ‘auto’}, optional</span></dt><dd><p>Min value of hic matrix, use ‘auto’ for specify min value automatically, default ‘auto’.</p>
</dd>
<dt><strong>orientation</strong><span class="classifier">str, optional</span></dt><dd><p>Track orientation, use ‘inverted’ for inverted track plot.</p>
</dd>
<dt><strong>resolution</strong><span class="classifier">{int, ‘auto’}, optional</span></dt><dd><p>Matrix resolution, default ‘auto’.</p>
</dd>
<dt><strong>transform</strong><span class="classifier">{str, bool}, optional</span></dt><dd><p>Transform for matrix, like ‘log2’, ‘log10’, default False.</p>
</dd>
<dt><strong>normalize</strong><span class="classifier">{‘zscore’, ‘expect’, ‘total’, False}</span></dt><dd><p>Normalization method, default False.</p>
</dd>
<dt><strong>norm: {‘log’, ‘no’}</strong></dt><dd><p>Method used when normalizing. default ‘no’</p>
</dd>
<dt><strong>gaussian_sigma</strong><span class="classifier">{float, False}, optional</span></dt><dd><p>Do gaussian filter(with sigma, for example 1.0) on matrix if specified. default False.</p>
</dd>
<dt><strong>process_func</strong><span class="classifier">{callable, str, False}, optional</span></dt><dd><p>Process matrix with a user-defined function(receive a matrix, return a processed matrix). default False.</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional;</span></dt><dd><p>Color of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_coverage</span></code>(coverage[, pos])</p></td>
<td><p>Append coverage to this track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.Cool.fetch_data" title="coolbox.core.track.Cool.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr[, gr2])</p></td>
<td><p>Fetch the raw matrix should be plotted.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.Cool.fetch_pixels" title="coolbox.core.track.Cool.fetch_pixels"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_pixels</span></code></a>(gr[, gr2])</p></td>
<td><p>Fetch the pixels table of upper triangle of the original contact matrix(not processed).</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code>(gr[, gr2])</p></td>
<td><p>Fetch 2d contact matrix, the matrix may be processed in case ‘transform’, ‘normalize’, ‘gaussian_sigma’, ‘process_func’ exits in properties.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_track_height</span></code>(frame_width, *args)</p></td>
<td><p>calculate track height dynamically.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pile_coverages</span></code>(coverages[, pos])</p></td>
<td><p>Pile a stack of coverages with self’s coverages</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code>(ax, gr[, gr2])</p></td>
<td><p>Plot matrix</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_coverages</span></code>(ax, gr, gr2)</p></td>
<td><p>Plot all coverages on given axes.</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 74%" />
<col style="width: 26%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>JuiceBoxLikeColor</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>JuiceBoxLikeColor2</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>adjust_figure</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>diagonal_mean</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>diagonal_mean_std</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>draw_colorbar</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>fetch_window_genome_range</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>fill_zero_nan</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>gaussian_matrix</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>has_prop</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>infer_binsize</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>normalize_cis</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>normalize_matrix</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>normalize_trans</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_colorbar</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_label</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_matrix</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>process_matrix</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>transform_matrix</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.Cool.fetch_data" title="coolbox.core.track.Cool.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr[, gr2])</p></td>
<td><p>Fetch the raw matrix should be plotted.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.Cool.fetch_pixels" title="coolbox.core.track.Cool.fetch_pixels"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_pixels</span></code></a>(gr[, gr2])</p></td>
<td><p>Fetch the pixels table of upper triangle of the original contact matrix(not processed).</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.Cool.fetch_data">
<span class="sig-name descname"><span class="pre">fetch_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">gr2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> &#x2192; <span class="pre">numpy.ndarray</span><a class="headerlink" href="#coolbox.core.track.Cool.fetch_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetch the raw matrix should be plotted. Normally it’s a matrix with raw contacts</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr2</strong><span class="classifier">GenomeRange, optional, keyword argument</span></dt><dd></dd>
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>matrix</strong><span class="classifier">np.array</span></dt><dd><p>Hi-C raw contact matrix.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.Cool.fetch_pixels">
<span class="sig-name descname"><span class="pre">fetch_pixels</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">gr2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.Cool.fetch_pixels" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetch the pixels table of upper triangle of the original contact matrix(not processed).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr2</strong><span class="classifier">GenomeRange, optional.</span></dt><dd></dd>
<dt><strong>balance</strong><span class="classifier">bool, optional</span></dt><dd><p>balance matrix or not,
default <cite>self.is_balance</cite>.</p>
</dd>
<dt><strong>resolution</strong><span class="classifier">{‘auto’, int}</span></dt><dd><p>resolution of the data. for example 5000.
‘auto’ for calculate resolution automatically.
default ‘auto’</p>
</dd>
<dt><strong>join</strong><span class="classifier">bool</span></dt><dd><p>whether to expand the bin ID columns
into (chrom, start, end).
default True</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>pixels</strong><span class="classifier">pandas.core.frame.DataFrame</span></dt><dd><p>Hi-C pixels table.
The pixel table contains the non-zero upper triangle entries of the contact map.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.track.DiScore">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.track.</span></span><span class="sig-name descname"><span class="pre">DiScore</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.DiScore" title="Permalink to this definition">¶</a></dt>
<dd><p>Directionnality index track.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>hicmat: {str, HicMatBase}</strong></dt><dd><p>The input hicmat file or HicMatBase object used to calculate di score.</p>
</dd>
<dt><strong>window_size: int, optional</strong></dt><dd><p>Width of the diamond region along the matrix used to calculate di. default: 40</p>
</dd>
<dt><strong>method: {‘standard’, ‘adaptive’}, optional</strong></dt><dd><p>Method used for calculating di. default: ‘adaptive’</p>
</dd>
<dt><strong>args_hic</strong><span class="classifier">dict, optional</span></dt><dd><p>Key words arguments used to create Cool/DotHic tack when the input hicmat is a file path</p>
</dd>
<dt><strong>style</strong><span class="classifier">str, optional</span></dt><dd><p>Track graph type, format {‘line’, ‘fill’, ‘heatmap’, ‘scatter’}</p>
</dd>
<dt><strong>fmt</strong><span class="classifier">str, optional</span></dt><dd><p>Line styles used when the plot style is <cite>line</cite>, will be passed as <cite>fmt</cite> parameter in <cite>matplotlib.pyplot.plot</cite></p>
</dd>
<dt><strong>line_width</strong><span class="classifier">int, optional</span></dt><dd><p>Value used when the plot style is <cite>line</cite>, will be passed as <cite>linewidth</cite> parameter in <cite>matplotlib.pyplot.plot</cite></p>
</dd>
<dt><strong>size</strong><span class="classifier">int, optional</span></dt><dd><p>Value used when the plot style is <cite>scatter</cite>, will be passed as <cite>s</cite> parameter in <cite>matplotlib.pyplot.scatter</cite></p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional</span></dt><dd><p>Main color</p>
</dd>
<dt><strong>threshold_color</strong><span class="classifier">str, optional</span></dt><dd><p>Track’s color for values greater than specified threshold.</p>
</dd>
<dt><strong>threshold</strong><span class="classifier">float, optional</span></dt><dd><p>Threshold used when the plot style is <cite>line</cite> or <cite>scatter</cite>, values greater than this thresh will be
plotted with <cite>color = threshold_color</cite></p>
</dd>
<dt><strong>cmap: str, optional</strong></dt><dd><p>Cmap used when the plot type is heatmap, will be passed as <cite>cmap</cite> paramerter in <cite>matplotlib.pyplot.matshow</cite></p>
</dd>
<dt><strong>alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Alpha value of plot, default 1.0</p>
</dd>
<dt><strong>orientation</strong><span class="classifier">str, optional</span></dt><dd><p>Track orientation, use ‘inverted’ for inverted track plot.</p>
</dd>
<dt><strong>data_range_style</strong><span class="classifier">{bool, ‘text’, ‘y-axis’}, optional</span></dt><dd><p>Show_data_range or not, default True.</p>
</dd>
<dt><strong>max_value</strong><span class="classifier">{float, ‘auto’}, optional</span></dt><dd><p>Max value of track. ‘auto’ for specify max value automatically, default ‘auto’.</p>
</dd>
<dt><strong>min_value</strong><span class="classifier">{float, ‘auto’}, optional</span></dt><dd><p>Min value of track. ‘auto’ for specify max value automatically, default ‘auto’.</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_coverage</span></code>(coverage[, pos])</p></td>
<td><p>Append coverage to this track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.DiScore.fetch_data" title="coolbox.core.track.DiScore.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pile_coverages</span></code>(coverages[, pos])</p></td>
<td><p>Pile a stack of coverages with self’s coverages</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code>(ax, gr, **kwargs)</p></td>
<td><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_coverages</span></code>(ax, gr, gr2)</p></td>
<td><p>Plot all coverages on given axes.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_yaxis_range</span></code>(plot_axis, y_ax)</p></td>
<td><p>Plot the scale of the y axis with respect to the plot_axis</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 66%" />
<col style="width: 34%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>adjust_plot</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>di_methods</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>has_prop</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_data_range</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_fill</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_heatmap</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_hist</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_label</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_line</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_scatter</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_stairs</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_text_range</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.DiScore.fetch_data" title="coolbox.core.track.DiScore.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.DiScore.fetch_data">
<span class="sig-name descname"><span class="pre">fetch_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> &#x2192; <span class="pre">numpy.ndarray</span><a class="headerlink" href="#coolbox.core.track.DiScore.fetch_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetch the raw data for the given GenomeRange.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.track.DotHiC">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.track.</span></span><span class="sig-name descname"><span class="pre">DotHiC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.DotHiC" title="Permalink to this definition">¶</a></dt>
<dd><p>HicMat track from .hic file.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>file: str</strong></dt><dd><p>The file path of .hic file.</p>
</dd>
<dt><strong>balance</strong><span class="classifier">{bool, ‘KR’, ‘VC’, ‘VC_SQRT’}, optional</span></dt><dd><p>Matrix balance method,
default True(‘KR’ balance)</p>
</dd>
<dt><strong>style</strong><span class="classifier">{‘triangular’, ‘window’, ‘matrix’}, optional</span></dt><dd><p>Matrix style, default ‘window’.</p>
</dd>
<dt><strong>cmap</strong><span class="classifier">str, optional</span></dt><dd><p>Color map of hic matrix, default “JuiceBoxLike”.</p>
</dd>
<dt><strong>color_bar</strong><span class="classifier">{‘vertical’, ‘horizontal’, ‘no’}, optional</span></dt><dd><p>Color bar style. default ‘vertical’.</p>
</dd>
<dt><strong>depth_ratio</strong><span class="classifier">float, optional</span></dt><dd><p>Depth ratio of triangular matrix, use ‘full’ for full depth. default ‘full’.</p>
</dd>
<dt><strong>max_value</strong><span class="classifier">{float, ‘auto’}, optional</span></dt><dd><p>Max value of hic matrix, use ‘auto’ for specify max value automatically, default ‘auto’.</p>
</dd>
<dt><strong>min_value</strong><span class="classifier">{float, ‘auto’}, optional</span></dt><dd><p>Min value of hic matrix, use ‘auto’ for specify min value automatically, default ‘auto’.</p>
</dd>
<dt><strong>orientation</strong><span class="classifier">str, optional</span></dt><dd><p>Track orientation, use ‘inverted’ for inverted track plot.</p>
</dd>
<dt><strong>resolution</strong><span class="classifier">{int, ‘auto’}, optional</span></dt><dd><p>Matrix resolution, default ‘auto’.</p>
</dd>
<dt><strong>transform</strong><span class="classifier">{str, bool}, optional</span></dt><dd><p>Transform for matrix, like ‘log2’, ‘log10’, default False.</p>
</dd>
<dt><strong>normalize</strong><span class="classifier">{‘zscore’, ‘expect’, ‘total’, False}</span></dt><dd><p>Normalization method, default False.</p>
</dd>
<dt><strong>norm: {‘log’, ‘no’}</strong></dt><dd><p>Method used when normalizing. default ‘no’</p>
</dd>
<dt><strong>gaussian_sigma</strong><span class="classifier">{float, False}, optional</span></dt><dd><p>Do gaussian filter(with sigma, for example 1.0) on matrix if specified. default False.</p>
</dd>
<dt><strong>process_func</strong><span class="classifier">{callable, str, False}, optional</span></dt><dd><p>Process matrix with a user-defined function(receive a matrix, return a processed matrix). default False.</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional;</span></dt><dd><p>Color of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_coverage</span></code>(coverage[, pos])</p></td>
<td><p>Append coverage to this track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.DotHiC.fetch_data" title="coolbox.core.track.DotHiC.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr[, gr2])</p></td>
<td><p>Fetch the raw matrix should be plotted.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.DotHiC.fetch_pixels" title="coolbox.core.track.DotHiC.fetch_pixels"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_pixels</span></code></a>(gr[, gr2, balance])</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code>(gr[, gr2])</p></td>
<td><p>Fetch 2d contact matrix, the matrix may be processed in case ‘transform’, ‘normalize’, ‘gaussian_sigma’, ‘process_func’ exits in properties.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_track_height</span></code>(frame_width, *args)</p></td>
<td><p>calculate track height dynamically.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pile_coverages</span></code>(coverages[, pos])</p></td>
<td><p>Pile a stack of coverages with self’s coverages</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code>(ax, gr[, gr2])</p></td>
<td><p>Plot matrix</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_coverages</span></code>(ax, gr, gr2)</p></td>
<td><p>Plot all coverages on given axes.</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 74%" />
<col style="width: 26%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>JuiceBoxLikeColor</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>JuiceBoxLikeColor2</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>adjust_figure</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>diagonal_mean</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>diagonal_mean_std</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>draw_colorbar</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>fetch_window_genome_range</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>fill_zero_nan</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>gaussian_matrix</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>has_prop</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>infer_binsize</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>normalize_cis</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>normalize_matrix</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>normalize_trans</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_colorbar</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_label</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_matrix</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>process_matrix</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>transform_matrix</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.DotHiC.fetch_data" title="coolbox.core.track.DotHiC.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr[, gr2])</p></td>
<td><p>Fetch the raw matrix should be plotted.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.DotHiC.fetch_pixels" title="coolbox.core.track.DotHiC.fetch_pixels"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_pixels</span></code></a>(gr[, gr2, balance])</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.DotHiC.fetch_data">
<span class="sig-name descname"><span class="pre">fetch_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gr2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> &#x2192; <span class="pre">numpy.ndarray</span><a class="headerlink" href="#coolbox.core.track.DotHiC.fetch_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetch the raw matrix should be plotted. Normally it’s a matrix with raw contacts</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr2</strong><span class="classifier">GenomeRange, optional, keyword argument</span></dt><dd></dd>
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>matrix</strong><span class="classifier">np.array</span></dt><dd><p>Hi-C raw contact matrix.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.DotHiC.fetch_pixels">
<span class="sig-name descname"><span class="pre">fetch_pixels</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gr2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">balance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.DotHiC.fetch_pixels" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr</strong><span class="classifier">{str, GenomeRange}</span></dt><dd><p>Intervals within input chromosome range.</p>
</dd>
<dt><strong>gr2</strong><span class="classifier">{str, GenomeRange}</span></dt><dd><p>Intervals within input chromsome range2.</p>
</dd>
<dt><strong>balance</strong><span class="classifier">{bool, ‘KR’, ‘VC’, ‘VC_SQRT’}, optional</span></dt><dd><p>matrix balance method,
default <cite>self.balance</cite>.</p>
</dd>
<dt><strong>resolution</strong><span class="classifier">{‘auto’, int}</span></dt><dd><p>resolution of the data. for example 5000.
‘auto’ for calculate resolution automatically.
default ‘auto’</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>pixels</strong><span class="classifier">pandas.core.frame.DataFrame</span></dt><dd><p>Hi-C pixels table.
The pixel table contains the non-zero upper triangle entries of the contact map.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.track.GTF">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.track.</span></span><span class="sig-name descname"><span class="pre">GTF</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.GTF" title="Permalink to this definition">¶</a></dt>
<dd><p>GTF gene annotation track.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>file</strong><span class="classifier">str</span></dt><dd><p>Path to .gtf(or .gtf.bgz) file.</p>
</dd>
<dt><strong>row_filter</strong><span class="classifier">str, optional</span></dt><dd><p>Row filter expression, only keep the rows for draw. (Default ‘feature == “gene”’)</p>
</dd>
<dt><strong>length_ratio_thresh</strong><span class="classifier">float</span></dt><dd><p>Length ratio threshold of features, (Default 0.01)</p>
</dd>
<dt><strong>color</strong><span class="classifier">{str, ‘random’}, optional</span></dt><dd><p>When the color is random, color for each gene will be randomly selected.</p>
</dd>
<dt><strong>name_attr</strong><span class="classifier">{‘auto’, ‘gene_name’, ‘gene_id’, str}, optional</span></dt><dd><p>Use which attribute to show feature’s name.
Default use ‘auto’(try ‘gene_name’ -&gt; ‘gene_id’ -&gt; ‘position_string’)</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_coverage</span></code>(coverage[, pos])</p></td>
<td><p>Append coverage to this track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.GTF.fetch_data" title="coolbox.core.track.GTF.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.GTF.fetch_intervals" title="coolbox.core.track.GTF.fetch_intervals"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_intervals</span></code></a>(gr)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code>(gr, **kwargs)</p></td>
<td><p>Fetch the plot data for the given GenomeRange.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pile_coverages</span></code>(coverages[, pos])</p></td>
<td><p>Pile a stack of coverages with self’s coverages</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.GTF.plot" title="coolbox.core.track.GTF.plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code></a>(ax, gr, **kwargs)</p></td>
<td><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_coverages</span></code>(ax, gr, gr2)</p></td>
<td><p>Plot all coverages on given axes.</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 58%" />
<col style="width: 42%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>has_prop</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_label</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.GTF.fetch_data" title="coolbox.core.track.GTF.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.GTF.fetch_intervals" title="coolbox.core.track.GTF.fetch_intervals"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_intervals</span></code></a>(gr)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.GTF.plot" title="coolbox.core.track.GTF.plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code></a>(ax, gr, **kwargs)</p></td>
<td><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.GTF.fetch_data">
<span class="sig-name descname"><span class="pre">fetch_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.GTF.fetch_data" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl>
<dt>df: pandas.DataFrame</dt><dd><p>should be with the format like:</p>
<dl class="simple">
<dt>columns = [‘seqname’, ‘source’, ‘feature’, ‘start’, ‘end’,</dt><dd><p>‘score’, ‘strand’, ‘frame’, ‘attribute’, ‘feature_name’]</p>
</dd>
</dl>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.GTF.fetch_intervals">
<span class="sig-name descname"><span class="pre">fetch_intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.GTF.fetch_intervals" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr</strong><span class="classifier">{str, GenomeRange}</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>intervals</strong><span class="classifier">pandas.core.frame.DataFrame</span></dt><dd><p>Annotation interval table.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.GTF.plot">
<span class="sig-name descname"><span class="pre">plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.GTF.plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.track.HLine">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.track.</span></span><span class="sig-name descname"><span class="pre">HLine</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.HLine" title="Permalink to this definition">¶</a></dt>
<dd><p>Horizontal line track.
Used for add a horizontal line between two tracks.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>line_width</strong><span class="classifier">float, optional</span></dt><dd><p>(Default: HLine.DEFAULT_LINE_WIDTH)</p>
</dd>
<dt><strong>line_style</strong><span class="classifier">str, optional</span></dt><dd><p>(Default: HLine.DEFAULT_LINE_STYLE)</p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional</span></dt><dd><p>(Default: HLine.DEFAULT_COLOR)</p>
</dd>
<dt><strong>alpha</strong><span class="classifier">float, optional</span></dt><dd><p>(Default: HLine.DEFAULT_ALPHA)</p>
</dd>
<dt><strong>height</strong><span class="classifier">float, optional</span></dt><dd><p>The height of Spacer track. (Default: HLine.DEFAULT_HEIGHT)</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>Track’s name.</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_coverage</span></code>(coverage[, pos])</p></td>
<td><p>Append coverage to this track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.HLine.fetch_data" title="coolbox.core.track.HLine.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code>(gr, **kwargs)</p></td>
<td><p>Fetch the plot data for the given GenomeRange.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pile_coverages</span></code>(coverages[, pos])</p></td>
<td><p>Pile a stack of coverages with self’s coverages</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.HLine.plot" title="coolbox.core.track.HLine.plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code></a>(ax, gr, **kwargs)</p></td>
<td><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_coverages</span></code>(ax, gr, gr2)</p></td>
<td><p>Plot all coverages on given axes.</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 58%" />
<col style="width: 42%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>has_prop</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_label</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.HLine.fetch_data" title="coolbox.core.track.HLine.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.HLine.plot" title="coolbox.core.track.HLine.plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code></a>(ax, gr, **kwargs)</p></td>
<td><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.HLine.fetch_data">
<span class="sig-name descname"><span class="pre">fetch_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.HLine.fetch_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetch the raw data for the given GenomeRange.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.HLine.plot">
<span class="sig-name descname"><span class="pre">plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.HLine.plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.track.HiCDiff">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.track.</span></span><span class="sig-name descname"><span class="pre">HiCDiff</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.HiCDiff" title="Permalink to this definition">¶</a></dt>
<dd><p>Track for express the comparison between two HiC Track.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>hic1</strong><span class="classifier">coolbox.api.track.Cool</span></dt><dd><p>First HiC Track or hic file path(.cool, .mcool, .hic).</p>
</dd>
<dt><strong>hic2</strong><span class="classifier">coolbox.api.track.Cool</span></dt><dd><p>Second HiC Track or hic file path(.cool, .mcool, .hic).</p>
</dd>
<dt><strong>args_hic: dict</strong></dt><dd><p>Key word arguments send to create Cool/DoctHic instance if the input hic1/hic2 is file.</p>
</dd>
<dt><strong>diff_method</strong><span class="classifier">str</span></dt><dd><p>Difference method (‘diff’, ‘log2fc’), default ‘diff’</p>
</dd>
<dt><strong>style</strong><span class="classifier">{‘triangular’, ‘window’, ‘matrix’}, optional</span></dt><dd><p>Matrix style, default ‘window’.</p>
</dd>
<dt><strong>cmap</strong><span class="classifier">str, optional</span></dt><dd><p>Color map of hic matrix, default “JuiceBoxLike”.</p>
</dd>
<dt><strong>color_bar</strong><span class="classifier">{‘vertical’, ‘horizontal’, ‘no’}, optional</span></dt><dd><p>Color bar style. default ‘vertical’.</p>
</dd>
<dt><strong>depth_ratio</strong><span class="classifier">float, optional</span></dt><dd><p>Depth ratio of triangular matrix, use ‘full’ for full depth. default ‘full’.</p>
</dd>
<dt><strong>max_value</strong><span class="classifier">{float, ‘auto’}, optional</span></dt><dd><p>Max value of hic matrix, use ‘auto’ for specify max value automatically, default ‘auto’.</p>
</dd>
<dt><strong>min_value</strong><span class="classifier">{float, ‘auto’}, optional</span></dt><dd><p>Min value of hic matrix, use ‘auto’ for specify min value automatically, default ‘auto’.</p>
</dd>
<dt><strong>orientation</strong><span class="classifier">str, optional</span></dt><dd><p>Track orientation, use ‘inverted’ for inverted track plot.</p>
</dd>
<dt><strong>resolution</strong><span class="classifier">{int, ‘auto’}, optional</span></dt><dd><p>Matrix resolution, default ‘auto’.</p>
</dd>
<dt><strong>transform</strong><span class="classifier">{str, bool}, optional</span></dt><dd><p>Transform for matrix, like ‘log2’, ‘log10’, default False.</p>
</dd>
<dt><strong>normalize</strong><span class="classifier">{‘zscore’, ‘expect’, ‘total’, False}</span></dt><dd><p>Normalization method, default False.</p>
</dd>
<dt><strong>norm: {‘log’, ‘no’}</strong></dt><dd><p>Method used when normalizing. default ‘no’</p>
</dd>
<dt><strong>gaussian_sigma</strong><span class="classifier">{float, False}, optional</span></dt><dd><p>Do gaussian filter(with sigma, for example 1.0) on matrix if specified. default False.</p>
</dd>
<dt><strong>process_func</strong><span class="classifier">{callable, str, False}, optional</span></dt><dd><p>Process matrix with a user-defined function(receive a matrix, return a processed matrix). default False.</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional;</span></dt><dd><p>Color of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_coverage</span></code>(coverage[, pos])</p></td>
<td><p>Append coverage to this track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.HiCDiff.fetch_data" title="coolbox.core.track.HiCDiff.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw matrix should be plotted.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code>(gr[, gr2])</p></td>
<td><p>Fetch 2d contact matrix, the matrix may be processed in case ‘transform’, ‘normalize’, ‘gaussian_sigma’, ‘process_func’ exits in properties.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_track_height</span></code>(frame_width, *args)</p></td>
<td><p>calculate track height dynamically.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pile_coverages</span></code>(coverages[, pos])</p></td>
<td><p>Pile a stack of coverages with self’s coverages</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code>(ax, gr[, gr2])</p></td>
<td><p>Plot matrix</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_coverages</span></code>(ax, gr, gr2)</p></td>
<td><p>Plot all coverages on given axes.</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 74%" />
<col style="width: 26%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>JuiceBoxLikeColor</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>JuiceBoxLikeColor2</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>adjust_figure</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>diagonal_mean</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>diagonal_mean_std</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>diff_matrix</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>draw_colorbar</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>fetch_pixels</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>fetch_window_genome_range</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>fill_zero_nan</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>gaussian_matrix</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>has_prop</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>infer_binsize</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>normalize_cis</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>normalize_matrix</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>normalize_trans</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_colorbar</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_label</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_matrix</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>process_matrix</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>transform_matrix</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.HiCDiff.fetch_data" title="coolbox.core.track.HiCDiff.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw matrix should be plotted.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.HiCDiff.fetch_data">
<span class="sig-name descname"><span class="pre">fetch_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> &#x2192; <span class="pre">numpy.ndarray</span><a class="headerlink" href="#coolbox.core.track.HiCDiff.fetch_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetch the raw matrix should be plotted. Normally it’s a matrix with raw contacts</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr2</strong><span class="classifier">GenomeRange, optional, keyword argument</span></dt><dd></dd>
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>matrix</strong><span class="classifier">np.array</span></dt><dd><p>Hi-C raw contact matrix.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="coolbox.core.track.HiCMat">
<span class="sig-prename descclassname"><span class="pre">coolbox.core.track.</span></span><span class="sig-name descname"><span class="pre">HiCMat</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span> </span><a class="reference internal" href="#coolbox.core.track.HicMatBase" title="coolbox.core.track.hicmat.base.HicMatBase"><span class="pre">coolbox.core.track.hicmat.base.HicMatBase</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> &#x2192; <a class="reference internal" href="#coolbox.core.track.HicMatBase" title="coolbox.core.track.hicmat.base.HicMatBase"><span class="pre">coolbox.core.track.hicmat.base.HicMatBase</span></a><a class="headerlink" href="#coolbox.core.track.HiCMat" title="Permalink to this definition">¶</a></dt>
<dd><p>Compose DotHic or Cool track automatically based on tpye of file extension (.cool, .mcool, .hic)</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.track.HiCPeaks">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.track.</span></span><span class="sig-name descname"><span class="pre">HiCPeaks</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.HiCPeaks" title="Permalink to this definition">¶</a></dt>
<dd><p>Hi-C Peaks(Loops) from .bedpe file. Used to show the peaks on the Hi-C interaction map.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>file: str</strong></dt><dd><p>Path of .bedpe file</p>
</dd>
<dt><strong>pos</strong><span class="classifier">str, optional</span></dt><dd><p>Method for choosing arch anchor for bedpe data: ‘start’, ‘end’, ‘mid’, default ‘mid’</p>
</dd>
<dt><strong>style</strong><span class="classifier">str, optional</span></dt><dd><p>Style of arcs plot: ‘arcs’, ‘hicpeaks’, default ‘arcs’</p>
</dd>
<dt><strong>score_to_width</strong><span class="classifier">str, optional</span></dt><dd><p>Mapping function of score to width, default: ‘0.5 + math.sqrt(score)’</p>
</dd>
<dt><strong>line_width</strong><span class="classifier">float, optional</span></dt><dd><p>Width of arc line.</p>
</dd>
<dt><strong>open_region</strong><span class="classifier">bool, optional</span></dt><dd><p>If specified to True, will fetch the contacts on side in the region,
default True</p>
</dd>
<dt><strong>diameter_to_height</strong><span class="classifier">str, optional</span></dt><dd><p>Mapping function of arc diameter(interval end - start) to height.
You can specify to ‘max_diameter’ let all arcs has same height.
default ‘max_height * diameter / max_diameter’.</p>
</dd>
<dt><strong>orientation</strong><span class="classifier">str, optional</span></dt><dd><p>Track orientation, use ‘inverted’ for inverted track plot. default None</p>
</dd>
<dt><strong>line_style</strong><span class="classifier">str, optional</span></dt><dd><p>Border line style, default ‘solid’</p>
</dd>
<dt><strong>fill</strong><span class="classifier">bool, optional</span></dt><dd><p>Fill center or not, default False.</p>
</dd>
<dt><strong>fill_color</strong><span class="classifier">str, optional</span></dt><dd><p>Fill color, use ‘bed_rgb’ for specify color in file,
default ‘bed_rgb’.</p>
</dd>
<dt><strong>fill_alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Alpha value of fill region.
default 0.2</p>
</dd>
<dt><strong>side</strong><span class="classifier">{‘upper’, ‘lower’, ‘both’}</span></dt><dd><p>Plot peak in which side of the matrix.
NOTE: This parameters is useful only if the Cool track in matrix format.</p>
</dd>
<dt><strong>alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Alpha value of track, default 0.8.</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional;</span></dt><dd><p>Color of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_coverage</span></code>(coverage[, pos])</p></td>
<td><p>Append coverage to this track.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pile_coverages</span></code>(coverages[, pos])</p></td>
<td><p>Pile a stack of coverages with self’s coverages</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code>(ax, gr, **kwargs)</p></td>
<td><p>Plot arc connecting two positions on a linear scale representing interactions between bins.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_arcs</span></code>(ax, gr, gr2, intervals)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_coverages</span></code>(ax, gr, gr2)</p></td>
<td><p>Plot all coverages on given axes.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_hicpeaks</span></code>(ax, gr, gr2, intervals)</p></td>
<td><p>plot hic peaks uppon a HicMatBase track.</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 66%" />
<col style="width: 34%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>fetch_intervals</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>has_prop</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_contacts</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_label</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.track.HicMatBase">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.track.</span></span><span class="sig-name descname"><span class="pre">HicMatBase</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.HicMatBase" title="Permalink to this definition">¶</a></dt>
<dd><p>HicMatBase class for all track plotting matrix-like data.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>style</strong><span class="classifier">{‘triangular’, ‘window’, ‘matrix’}, optional</span></dt><dd><p>Matrix style, default ‘window’.</p>
</dd>
<dt><strong>cmap</strong><span class="classifier">str, optional</span></dt><dd><p>Color map of hic matrix, default “JuiceBoxLike”.</p>
</dd>
<dt><strong>color_bar</strong><span class="classifier">{‘vertical’, ‘horizontal’, ‘no’}, optional</span></dt><dd><p>Color bar style. default ‘vertical’.</p>
</dd>
<dt><strong>depth_ratio</strong><span class="classifier">float, optional</span></dt><dd><p>Depth ratio of triangular matrix, use ‘full’ for full depth. default ‘full’.</p>
</dd>
<dt><strong>max_value</strong><span class="classifier">{float, ‘auto’}, optional</span></dt><dd><p>Max value of hic matrix, use ‘auto’ for specify max value automatically, default ‘auto’.</p>
</dd>
<dt><strong>min_value</strong><span class="classifier">{float, ‘auto’}, optional</span></dt><dd><p>Min value of hic matrix, use ‘auto’ for specify min value automatically, default ‘auto’.</p>
</dd>
<dt><strong>orientation</strong><span class="classifier">str, optional</span></dt><dd><p>Track orientation, use ‘inverted’ for inverted track plot.</p>
</dd>
<dt><strong>resolution</strong><span class="classifier">{int, ‘auto’}, optional</span></dt><dd><p>Matrix resolution, default ‘auto’.</p>
</dd>
<dt><strong>transform</strong><span class="classifier">{str, bool}, optional</span></dt><dd><p>Transform for matrix, like ‘log2’, ‘log10’, default False.</p>
</dd>
<dt><strong>normalize</strong><span class="classifier">{‘zscore’, ‘expect’, ‘total’, False}</span></dt><dd><p>Normalization method, default False.</p>
</dd>
<dt><strong>norm: {‘log’, ‘no’}</strong></dt><dd><p>Method used when normalizing. default ‘no’</p>
</dd>
<dt><strong>gaussian_sigma</strong><span class="classifier">{float, False}, optional</span></dt><dd><p>Do gaussian filter(with sigma, for example 1.0) on matrix if specified. default False.</p>
</dd>
<dt><strong>process_func</strong><span class="classifier">{callable, str, False}, optional</span></dt><dd><p>Process matrix with a user-defined function(receive a matrix, return a processed matrix). default False.</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional;</span></dt><dd><p>Color of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_coverage</span></code>(coverage[, pos])</p></td>
<td><p>Append coverage to this track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.HicMatBase.fetch_data" title="coolbox.core.track.HicMatBase.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr[, gr2])</p></td>
<td><p>Fetch the raw matrix should be plotted.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.HicMatBase.fetch_plot_data" title="coolbox.core.track.HicMatBase.fetch_plot_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code></a>(gr[, gr2])</p></td>
<td><p>Fetch 2d contact matrix, the matrix may be processed in case ‘transform’, ‘normalize’, ‘gaussian_sigma’, ‘process_func’ exits in properties.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_track_height</span></code>(frame_width, *args)</p></td>
<td><p>calculate track height dynamically.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pile_coverages</span></code>(coverages[, pos])</p></td>
<td><p>Pile a stack of coverages with self’s coverages</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.HicMatBase.plot" title="coolbox.core.track.HicMatBase.plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code></a>(ax, gr[, gr2])</p></td>
<td><p>Plot matrix</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_coverages</span></code>(ax, gr, gr2)</p></td>
<td><p>Plot all coverages on given axes.</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 74%" />
<col style="width: 26%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>JuiceBoxLikeColor</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>JuiceBoxLikeColor2</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>adjust_figure</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>diagonal_mean</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>diagonal_mean_std</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>draw_colorbar</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>fetch_pixels</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>fetch_window_genome_range</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>fill_zero_nan</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>gaussian_matrix</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>has_prop</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>infer_binsize</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>normalize_cis</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>normalize_matrix</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>normalize_trans</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_colorbar</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_label</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_matrix</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>process_matrix</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>transform_matrix</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.HicMatBase.fetch_data" title="coolbox.core.track.HicMatBase.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr[, gr2])</p></td>
<td><p>Fetch the raw matrix should be plotted.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.HicMatBase.fetch_plot_data" title="coolbox.core.track.HicMatBase.fetch_plot_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code></a>(gr[, gr2])</p></td>
<td><p>Fetch 2d contact matrix, the matrix may be processed in case ‘transform’, ‘normalize’, ‘gaussian_sigma’, ‘process_func’ exits in properties.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.HicMatBase.plot" title="coolbox.core.track.HicMatBase.plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code></a>(ax, gr[, gr2])</p></td>
<td><p>Plot matrix</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.HicMatBase.fetch_data">
<span class="sig-name descname"><span class="pre">fetch_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">gr2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> &#x2192; <span class="pre">numpy.ndarray</span><a class="headerlink" href="#coolbox.core.track.HicMatBase.fetch_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetch the raw matrix should be plotted. Normally it’s a matrix with raw contacts</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr2</strong><span class="classifier">GenomeRange, optional, keyword argument</span></dt><dd></dd>
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>matrix</strong><span class="classifier">np.array</span></dt><dd><p>Hi-C raw contact matrix.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.HicMatBase.fetch_plot_data">
<span class="sig-name descname"><span class="pre">fetch_plot_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">gr2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> &#x2192; <span class="pre">numpy.ndarray</span><a class="headerlink" href="#coolbox.core.track.HicMatBase.fetch_plot_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetch 2d contact matrix, the matrix may be processed in case
‘transform’, ‘normalize’, ‘gaussian_sigma’, ‘process_func’ exits in properties.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr2</strong><span class="classifier">GenomeRange, optional</span></dt><dd></dd>
<dt><strong>gr_updated: bool, optional</strong></dt><dd><p>If the input GenomeRange has been updated. default False
Default False means that the input gr will be expanded in window mode</p>
</dd>
<dt><strong>gr: GenomeRange</strong></dt><dd><p>GenomeRange is an object with properties of [chrom, start, end].</p>
</dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>matrix</strong><span class="classifier">np.array</span></dt><dd><p>Processed hic matrix used for plotting.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.HicMatBase.plot">
<span class="sig-name descname"><span class="pre">plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">gr2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.HicMatBase.plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot matrix</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>ax</strong></dt><dd></dd>
<dt><strong>gr2</strong><span class="classifier">GenomeRange, optional</span></dt><dd></dd>
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="coolbox.core.track.Hist">
<span class="sig-prename descclassname"><span class="pre">coolbox.core.track.</span></span><span class="sig-name descname"><span class="pre">Hist</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> &#x2192; <a class="reference internal" href="#coolbox.core.track.HistBase" title="coolbox.core.track.hist.base.HistBase"><span class="pre">coolbox.core.track.hist.base.HistBase</span></a><a class="headerlink" href="#coolbox.core.track.Hist" title="Permalink to this definition">¶</a></dt>
<dd><p>Compose BigWig, BedGraph, SNP, BAMCov track automatically based on type of file
extension.(.bw, .bigwig, .snp, .vcf, .bedgraph, .bam, .sam)</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.track.HistBase">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.track.</span></span><span class="sig-name descname"><span class="pre">HistBase</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.HistBase" title="Permalink to this definition">¶</a></dt>
<dd><p>HistBase track class</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>style</strong><span class="classifier">str, optional</span></dt><dd><p>Track graph type, format {‘line’, ‘fill’, ‘heatmap’, ‘scatter’}</p>
</dd>
<dt><strong>fmt</strong><span class="classifier">str, optional</span></dt><dd><p>Line styles used when the plot style is <cite>line</cite>, will be passed as <cite>fmt</cite> parameter in <cite>matplotlib.pyplot.plot</cite></p>
</dd>
<dt><strong>line_width</strong><span class="classifier">int, optional</span></dt><dd><p>Value used when the plot style is <cite>line</cite>, will be passed as <cite>linewidth</cite> parameter in <cite>matplotlib.pyplot.plot</cite></p>
</dd>
<dt><strong>size</strong><span class="classifier">int, optional</span></dt><dd><p>Value used when the plot style is <cite>scatter</cite>, will be passed as <cite>s</cite> parameter in <cite>matplotlib.pyplot.scatter</cite></p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional</span></dt><dd><p>Main color</p>
</dd>
<dt><strong>threshold_color</strong><span class="classifier">str, optional</span></dt><dd><p>Track’s color for values greater than specified threshold.</p>
</dd>
<dt><strong>threshold</strong><span class="classifier">float, optional</span></dt><dd><p>Threshold used when the plot style is <cite>line</cite> or <cite>scatter</cite>, values greater than this thresh will be
plotted with <cite>color = threshold_color</cite></p>
</dd>
<dt><strong>cmap: str, optional</strong></dt><dd><p>Cmap used when the plot type is heatmap, will be passed as <cite>cmap</cite> paramerter in <cite>matplotlib.pyplot.matshow</cite></p>
</dd>
<dt><strong>alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Alpha value of plot, default 1.0</p>
</dd>
<dt><strong>orientation</strong><span class="classifier">str, optional</span></dt><dd><p>Track orientation, use ‘inverted’ for inverted track plot.</p>
</dd>
<dt><strong>data_range_style</strong><span class="classifier">{bool, ‘text’, ‘y-axis’}, optional</span></dt><dd><p>Show_data_range or not, default True.</p>
</dd>
<dt><strong>max_value</strong><span class="classifier">{float, ‘auto’}, optional</span></dt><dd><p>Max value of track. ‘auto’ for specify max value automatically, default ‘auto’.</p>
</dd>
<dt><strong>min_value</strong><span class="classifier">{float, ‘auto’}, optional</span></dt><dd><p>Min value of track. ‘auto’ for specify max value automatically, default ‘auto’.</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_coverage</span></code>(coverage[, pos])</p></td>
<td><p>Append coverage to this track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.HistBase.fetch_data" title="coolbox.core.track.HistBase.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.HistBase.fetch_plot_data" title="coolbox.core.track.HistBase.fetch_plot_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code></a>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pile_coverages</span></code>(coverages[, pos])</p></td>
<td><p>Pile a stack of coverages with self’s coverages</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.HistBase.plot" title="coolbox.core.track.HistBase.plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code></a>(ax, gr, **kwargs)</p></td>
<td><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_coverages</span></code>(ax, gr, gr2)</p></td>
<td><p>Plot all coverages on given axes.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_yaxis_range</span></code>(plot_axis, y_ax)</p></td>
<td><p>Plot the scale of the y axis with respect to the plot_axis</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 66%" />
<col style="width: 34%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>adjust_plot</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>has_prop</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_data_range</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_fill</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_heatmap</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_hist</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_label</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_line</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_scatter</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_stairs</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_text_range</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.HistBase.fetch_data" title="coolbox.core.track.HistBase.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.HistBase.fetch_plot_data" title="coolbox.core.track.HistBase.fetch_plot_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code></a>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.HistBase.plot" title="coolbox.core.track.HistBase.plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code></a>(ax, gr, **kwargs)</p></td>
<td><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.HistBase.fetch_data">
<span class="sig-name descname"><span class="pre">fetch_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> &#x2192; <span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">numpy.ndarray</span><span class="p"><span class="pre">,</span> </span><span class="pre">numpy.ndarray</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span> </span><span class="pre">pandas.core.frame.DataFrame</span><span class="p"><span class="pre">,</span> </span><span class="pre">numpy.ndarray</span><span class="p"><span class="pre">]</span></span><a class="headerlink" href="#coolbox.core.track.HistBase.fetch_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetch the raw data for the given GenomeRange.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.HistBase.fetch_plot_data">
<span class="sig-name descname"><span class="pre">fetch_plot_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> &#x2192; <span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">numpy.ndarray</span><span class="p"><span class="pre">,</span> </span><span class="pre">numpy.ndarray</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span> </span><span class="pre">pandas.core.frame.DataFrame</span><span class="p"><span class="pre">,</span> </span><span class="pre">numpy.ndarray</span><span class="p"><span class="pre">]</span></span><a class="headerlink" href="#coolbox.core.track.HistBase.fetch_plot_data" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd><p>GenomeRange is an object with properties of [chrom, start, end].</p>
</dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>hist_data: Union[Tuple[np.ndarray, np.ndarray], pd.DataFrame, np.ndarray]</dt><dd><p>data used for plotting.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.HistBase.plot">
<span class="sig-name descname"><span class="pre">plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.HistBase.plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.track.Ideogram">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.track.</span></span><span class="sig-name descname"><span class="pre">Ideogram</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.Ideogram" title="Permalink to this definition">¶</a></dt>
<dd><p>The chromosome ideograme track.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>file_</strong><span class="classifier">str</span></dt><dd><p>Path to chromosome ideogram txt file,
ideogram file is download from the UCSC Table Browser CytoBandIdeo table (in “all table” group).
see: <a class="reference external" href="http://genome.ucsc.edu/cgi-bin/hgTables?hgta_group=allTables&amp;hgta_table=cytoBandIdeo">http://genome.ucsc.edu/cgi-bin/hgTables?hgta_group=allTables&amp;hgta_table=cytoBandIdeo</a></p>
</dd>
<dt><strong>color_scheme</strong><span class="classifier">dict, optional</span></dt><dd><p>Color scheme of ideogram, default: Ideogram.DEFAULT_COLOR_SCHEME</p>
</dd>
<dt><strong>show_band_name</strong><span class="classifier">bool, optional</span></dt><dd><p>Show band name or not. default True.</p>
</dd>
<dt><strong>font_size</strong><span class="classifier">int, optional</span></dt><dd><p>Band name font size.</p>
</dd>
<dt><strong>border_color</strong><span class="classifier">str, optional</span></dt><dd><p>Border color. default: ‘#000000’</p>
</dd>
<dt><strong>border_width</strong><span class="classifier">float, optional</span></dt><dd><p>Border width. default: 1.2</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional;</span></dt><dd><p>Color of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_coverage</span></code>(coverage[, pos])</p></td>
<td><p>Append coverage to this track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.Ideogram.fetch_data" title="coolbox.core.track.Ideogram.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code>(gr, **kwargs)</p></td>
<td><p>Fetch the plot data for the given GenomeRange.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pile_coverages</span></code>(coverages[, pos])</p></td>
<td><p>Pile a stack of coverages with self’s coverages</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.Ideogram.plot" title="coolbox.core.track.Ideogram.plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code></a>(ax, gr, **kwargs)</p></td>
<td><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_coverages</span></code>(ax, gr, gr2)</p></td>
<td><p>Plot all coverages on given axes.</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 68%" />
<col style="width: 32%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>has_prop</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>lookup_band_color</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_label</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_text</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.Ideogram.fetch_data" title="coolbox.core.track.Ideogram.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.Ideogram.plot" title="coolbox.core.track.Ideogram.plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code></a>(ax, gr, **kwargs)</p></td>
<td><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.Ideogram.fetch_data">
<span class="sig-name descname"><span class="pre">fetch_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.Ideogram.fetch_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetch the raw data for the given GenomeRange.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.Ideogram.plot">
<span class="sig-name descname"><span class="pre">plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.Ideogram.plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.track.InsuScore">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.track.</span></span><span class="sig-name descname"><span class="pre">InsuScore</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.InsuScore" title="Permalink to this definition">¶</a></dt>
<dd><p>Insulation score track.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>hicmat: {str, HicMatBase}</strong></dt><dd><p>The input hicmat file or HicMatBase object used to calculate di score.</p>
</dd>
<dt><strong>window_size: {int, str}, optional</strong></dt><dd><p>Width of the diamond region along the matrix used to calculate di. default: 20
window_size can also be a str with format like “20-40” representing a range of window_sizes.</p>
</dd>
<dt><strong>normalize: bool, optional</strong></dt><dd><p>Weather to log-nomalize the insulation score array. default: true</p>
</dd>
<dt><strong>method: {‘standard’, ‘adaptive’}, optional</strong></dt><dd><p>Method used for calculating di. default: ‘adaptive’</p>
</dd>
<dt><strong>args_hic</strong><span class="classifier">dict, optional</span></dt><dd><p>Key words arguments used to create Cool/DotHic tack when the input hicmat is a file path</p>
</dd>
<dt><strong>style</strong><span class="classifier">str, optional</span></dt><dd><p>Track graph type, format {‘line’, ‘fill’, ‘heatmap’, ‘scatter’}</p>
</dd>
<dt><strong>fmt</strong><span class="classifier">str, optional</span></dt><dd><p>Line styles used when the plot style is <cite>line</cite>, will be passed as <cite>fmt</cite> parameter in <cite>matplotlib.pyplot.plot</cite></p>
</dd>
<dt><strong>line_width</strong><span class="classifier">int, optional</span></dt><dd><p>Value used when the plot style is <cite>line</cite>, will be passed as <cite>linewidth</cite> parameter in <cite>matplotlib.pyplot.plot</cite></p>
</dd>
<dt><strong>size</strong><span class="classifier">int, optional</span></dt><dd><p>Value used when the plot style is <cite>scatter</cite>, will be passed as <cite>s</cite> parameter in <cite>matplotlib.pyplot.scatter</cite></p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional</span></dt><dd><p>Main color</p>
</dd>
<dt><strong>threshold_color</strong><span class="classifier">str, optional</span></dt><dd><p>Track’s color for values greater than specified threshold.</p>
</dd>
<dt><strong>threshold</strong><span class="classifier">float, optional</span></dt><dd><p>Threshold used when the plot style is <cite>line</cite> or <cite>scatter</cite>, values greater than this thresh will be
plotted with <cite>color = threshold_color</cite></p>
</dd>
<dt><strong>cmap: str, optional</strong></dt><dd><p>Cmap used when the plot type is heatmap, will be passed as <cite>cmap</cite> paramerter in <cite>matplotlib.pyplot.matshow</cite></p>
</dd>
<dt><strong>alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Alpha value of plot, default 1.0</p>
</dd>
<dt><strong>orientation</strong><span class="classifier">str, optional</span></dt><dd><p>Track orientation, use ‘inverted’ for inverted track plot.</p>
</dd>
<dt><strong>data_range_style</strong><span class="classifier">{bool, ‘text’, ‘y-axis’}, optional</span></dt><dd><p>Show_data_range or not, default True.</p>
</dd>
<dt><strong>max_value</strong><span class="classifier">{float, ‘auto’}, optional</span></dt><dd><p>Max value of track. ‘auto’ for specify max value automatically, default ‘auto’.</p>
</dd>
<dt><strong>min_value</strong><span class="classifier">{float, ‘auto’}, optional</span></dt><dd><p>Min value of track. ‘auto’ for specify max value automatically, default ‘auto’.</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_coverage</span></code>(coverage[, pos])</p></td>
<td><p>Append coverage to this track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.InsuScore.fetch_data" title="coolbox.core.track.InsuScore.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pile_coverages</span></code>(coverages[, pos])</p></td>
<td><p>Pile a stack of coverages with self’s coverages</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code>(ax, gr, **kwargs)</p></td>
<td><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_coverages</span></code>(ax, gr, gr2)</p></td>
<td><p>Plot all coverages on given axes.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_yaxis_range</span></code>(plot_axis, y_ax)</p></td>
<td><p>Plot the scale of the y axis with respect to the plot_axis</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 66%" />
<col style="width: 34%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>adjust_plot</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>has_prop</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_data_range</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_fill</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_heatmap</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_hist</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_label</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_line</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_scatter</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_stairs</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_text_range</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.InsuScore.fetch_data" title="coolbox.core.track.InsuScore.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.InsuScore.fetch_data">
<span class="sig-name descname"><span class="pre">fetch_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> &#x2192; <span class="pre">numpy.ndarray</span><a class="headerlink" href="#coolbox.core.track.InsuScore.fetch_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetch the raw data for the given GenomeRange.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.track.Pairs">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.track.</span></span><span class="sig-name descname"><span class="pre">Pairs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.Pairs" title="Permalink to this definition">¶</a></dt>
<dd><p>Arcs track from .pairs file.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>file: str</strong></dt><dd><p>Path of .pairs file</p>
</dd>
<dt><strong>style</strong><span class="classifier">str, optional</span></dt><dd><p>Style of arcs plot: ‘arcs’, ‘hicpeaks’, default ‘arcs’</p>
</dd>
<dt><strong>score_to_width</strong><span class="classifier">str, optional</span></dt><dd><p>Mapping function of score to width, default: ‘0.5 + math.sqrt(score)’</p>
</dd>
<dt><strong>line_width</strong><span class="classifier">float, optional</span></dt><dd><p>Width of arc line.</p>
</dd>
<dt><strong>open_region</strong><span class="classifier">bool, optional</span></dt><dd><p>If specified to True, will fetch the contacts on side in the region,
default True</p>
</dd>
<dt><strong>diameter_to_height</strong><span class="classifier">str, optional</span></dt><dd><p>Mapping function of arc diameter(interval end - start) to height.
You can specify to ‘max_diameter’ let all arcs has same height.
default ‘max_height * diameter / max_diameter’.</p>
</dd>
<dt><strong>orientation</strong><span class="classifier">str, optional</span></dt><dd><p>Track orientation, use ‘inverted’ for inverted track plot. default None</p>
</dd>
<dt><strong>line_style</strong><span class="classifier">str, optional</span></dt><dd><p>Border line style, default ‘solid’</p>
</dd>
<dt><strong>fill</strong><span class="classifier">bool, optional</span></dt><dd><p>Fill center or not, default False.</p>
</dd>
<dt><strong>fill_color</strong><span class="classifier">str, optional</span></dt><dd><p>Fill color, use ‘bed_rgb’ for specify color in file,
default ‘bed_rgb’.</p>
</dd>
<dt><strong>fill_alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Alpha value of fill region.
default 0.2</p>
</dd>
<dt><strong>side</strong><span class="classifier">{‘upper’, ‘lower’, ‘both’}</span></dt><dd><p>Plot peak in which side of the matrix.
NOTE: This parameters is useful only if the Cool track in matrix format.</p>
</dd>
<dt><strong>alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Alpha value of track, default 0.8.</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional;</span></dt><dd><p>Color of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_coverage</span></code>(coverage[, pos])</p></td>
<td><p>Append coverage to this track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.Pairs.fetch_data" title="coolbox.core.track.Pairs.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pile_coverages</span></code>(coverages[, pos])</p></td>
<td><p>Pile a stack of coverages with self’s coverages</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code>(ax, gr, **kwargs)</p></td>
<td><p>Plot arc connecting two positions on a linear scale representing interactions between bins.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_arcs</span></code>(ax, gr, gr2, intervals)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_coverages</span></code>(ax, gr, gr2)</p></td>
<td><p>Plot all coverages on given axes.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_hicpeaks</span></code>(ax, gr, gr2, intervals)</p></td>
<td><p>plot hic peaks uppon a HicMatBase track.</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 66%" />
<col style="width: 34%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>fetch_intervals</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>has_prop</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_contacts</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_label</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.Pairs.fetch_data" title="coolbox.core.track.Pairs.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.Pairs.fetch_data">
<span class="sig-name descname"><span class="pre">fetch_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.Pairs.fetch_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetch the raw data for the given GenomeRange.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.track.SNP">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.track.</span></span><span class="sig-name descname"><span class="pre">SNP</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.SNP" title="Permalink to this definition">¶</a></dt>
<dd><p>Track for show SNPs Manhattan plot.</p>
<p>Input file is a tab-split file contains SNP’s chrom, position, pvalue information.
You should specify it’s column indexes by <cite>col_chrom</cite>, <cite>col_pos</cite>, <cite>col_pval</cite> arguments.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>file</strong><span class="classifier">str</span></dt><dd><p>Path to input .snp/.vcf file.</p>
</dd>
<dt><strong>col_chrom</strong><span class="classifier">int</span></dt><dd><p>Column index of seqname(chromosome).</p>
</dd>
<dt><strong>col_pos</strong><span class="classifier">int</span></dt><dd><p>Column index of snp position.</p>
</dd>
<dt><strong>col_pval</strong><span class="classifier">int</span></dt><dd><p>Column index of snp p-value.</p>
</dd>
<dt><strong>pval_transform</strong><span class="classifier">{‘-log2’, ‘-log10’}</span></dt><dd><p>Transform the p value. Default ‘-log10’.</p>
</dd>
<dt><strong>style</strong><span class="classifier">str, optional</span></dt><dd><p>Track graph type, format {‘line’, ‘fill’, ‘heatmap’, ‘scatter’}</p>
</dd>
<dt><strong>fmt</strong><span class="classifier">str, optional</span></dt><dd><p>Line styles used when the plot style is <cite>line</cite>, will be passed as <cite>fmt</cite> parameter in <cite>matplotlib.pyplot.plot</cite></p>
</dd>
<dt><strong>line_width</strong><span class="classifier">int, optional</span></dt><dd><p>Value used when the plot style is <cite>line</cite>, will be passed as <cite>linewidth</cite> parameter in <cite>matplotlib.pyplot.plot</cite></p>
</dd>
<dt><strong>size</strong><span class="classifier">int, optional</span></dt><dd><p>Value used when the plot style is <cite>scatter</cite>, will be passed as <cite>s</cite> parameter in <cite>matplotlib.pyplot.scatter</cite></p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional</span></dt><dd><p>Main color</p>
</dd>
<dt><strong>threshold_color</strong><span class="classifier">str, optional</span></dt><dd><p>Track’s color for values greater than specified threshold.</p>
</dd>
<dt><strong>threshold</strong><span class="classifier">float, optional</span></dt><dd><p>Threshold used when the plot style is <cite>line</cite> or <cite>scatter</cite>, values greater than this thresh will be
plotted with <cite>color = threshold_color</cite></p>
</dd>
<dt><strong>cmap: str, optional</strong></dt><dd><p>Cmap used when the plot type is heatmap, will be passed as <cite>cmap</cite> paramerter in <cite>matplotlib.pyplot.matshow</cite></p>
</dd>
<dt><strong>alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Alpha value of plot, default 1.0</p>
</dd>
<dt><strong>orientation</strong><span class="classifier">str, optional</span></dt><dd><p>Track orientation, use ‘inverted’ for inverted track plot.</p>
</dd>
<dt><strong>data_range_style</strong><span class="classifier">{bool, ‘text’, ‘y-axis’}, optional</span></dt><dd><p>Show_data_range or not, default True.</p>
</dd>
<dt><strong>max_value</strong><span class="classifier">{float, ‘auto’}, optional</span></dt><dd><p>Max value of track. ‘auto’ for specify max value automatically, default ‘auto’.</p>
</dd>
<dt><strong>min_value</strong><span class="classifier">{float, ‘auto’}, optional</span></dt><dd><p>Min value of track. ‘auto’ for specify max value automatically, default ‘auto’.</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_coverage</span></code>(coverage[, pos])</p></td>
<td><p>Append coverage to this track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.SNP.fetch_data" title="coolbox.core.track.SNP.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.SNP.fetch_plot_data" title="coolbox.core.track.SNP.fetch_plot_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code></a>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pile_coverages</span></code>(coverages[, pos])</p></td>
<td><p>Pile a stack of coverages with self’s coverages</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code>(ax, gr, **kwargs)</p></td>
<td><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_coverages</span></code>(ax, gr, gr2)</p></td>
<td><p>Plot all coverages on given axes.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_yaxis_range</span></code>(plot_axis, y_ax)</p></td>
<td><p>Plot the scale of the y axis with respect to the plot_axis</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 66%" />
<col style="width: 34%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>adjust_plot</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>has_prop</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>load_range</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_data_range</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_fill</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_heatmap</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_hist</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_label</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_line</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_scatter</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_stairs</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_text_range</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>transform_fn</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.SNP.fetch_data" title="coolbox.core.track.SNP.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.SNP.fetch_plot_data" title="coolbox.core.track.SNP.fetch_plot_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code></a>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.SNP.fetch_data">
<span class="sig-name descname"><span class="pre">fetch_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.SNP.fetch_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetch the raw data for the given GenomeRange.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.SNP.fetch_plot_data">
<span class="sig-name descname"><span class="pre">fetch_plot_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.SNP.fetch_plot_data" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd><p>GenomeRange is an object with properties of [chrom, start, end].</p>
</dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>hist_data: Union[Tuple[np.ndarray, np.ndarray], pd.DataFrame, np.ndarray]</dt><dd><p>data used for plotting.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.track.Selfish">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.track.</span></span><span class="sig-name descname"><span class="pre">Selfish</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.Selfish" title="Permalink to this definition">¶</a></dt>
<dd><p>Differential chromatin interaction.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>hic1</strong><span class="classifier">coolbox.api.track.Cool</span></dt><dd><p>First HiC Track or hic file path(.cool, .mcool, .hic).</p>
</dd>
<dt><strong>hic2</strong><span class="classifier">coolbox.api.track.Cool</span></dt><dd><p>Second HiC Track or hic file path(.cool, .mcool, .hic).</p>
</dd>
<dt><strong>hic_args</strong><span class="classifier">dict, optional</span></dt><dd><p>Argument to create Hi-C instance, only in use
when first or second argument is a path.</p>
</dd>
<dt><strong>sigma0</strong><span class="classifier">float, optional</span></dt><dd><p>Initial sigma, parameter of SELFISH method. Default: 1.6</p>
</dd>
<dt><strong>s</strong><span class="classifier">int, optional</span></dt><dd><p>Iteration count parameter of SELFISH method. Default: 10</p>
</dd>
<dt><strong>style</strong><span class="classifier">{‘triangular’, ‘window’, ‘matrix’}, optional</span></dt><dd><p>Matrix style, default ‘window’.</p>
</dd>
<dt><strong>cmap</strong><span class="classifier">str, optional</span></dt><dd><p>Color map of hic matrix, default “JuiceBoxLike”.</p>
</dd>
<dt><strong>color_bar</strong><span class="classifier">{‘vertical’, ‘horizontal’, ‘no’}, optional</span></dt><dd><p>Color bar style. default ‘vertical’.</p>
</dd>
<dt><strong>depth_ratio</strong><span class="classifier">float, optional</span></dt><dd><p>Depth ratio of triangular matrix, use ‘full’ for full depth. default ‘full’.</p>
</dd>
<dt><strong>max_value</strong><span class="classifier">{float, ‘auto’}, optional</span></dt><dd><p>Max value of hic matrix, use ‘auto’ for specify max value automatically, default ‘auto’.</p>
</dd>
<dt><strong>min_value</strong><span class="classifier">{float, ‘auto’}, optional</span></dt><dd><p>Min value of hic matrix, use ‘auto’ for specify min value automatically, default ‘auto’.</p>
</dd>
<dt><strong>orientation</strong><span class="classifier">str, optional</span></dt><dd><p>Track orientation, use ‘inverted’ for inverted track plot.</p>
</dd>
<dt><strong>resolution</strong><span class="classifier">{int, ‘auto’}, optional</span></dt><dd><p>Matrix resolution, default ‘auto’.</p>
</dd>
<dt><strong>transform</strong><span class="classifier">{str, bool}, optional</span></dt><dd><p>Transform for matrix, like ‘log2’, ‘log10’, default False.</p>
</dd>
<dt><strong>normalize</strong><span class="classifier">{‘zscore’, ‘expect’, ‘total’, False}</span></dt><dd><p>Normalization method, default False.</p>
</dd>
<dt><strong>norm: {‘log’, ‘no’}</strong></dt><dd><p>Method used when normalizing. default ‘no’</p>
</dd>
<dt><strong>gaussian_sigma</strong><span class="classifier">{float, False}, optional</span></dt><dd><p>Do gaussian filter(with sigma, for example 1.0) on matrix if specified. default False.</p>
</dd>
<dt><strong>process_func</strong><span class="classifier">{callable, str, False}, optional</span></dt><dd><p>Process matrix with a user-defined function(receive a matrix, return a processed matrix). default False.</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional;</span></dt><dd><p>Color of ax</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">References</p>
<p>Abbas Roayaei Ardakany, Ferhat Ay, Stefano Lonardi,
Selfish: discovery of differential chromatin interactions via a self-similarity measure,
Bioinformatics, Volume 35, Issue 14, July 2019, Pages i145–i153,
<a class="reference external" href="https://doi.org/10.1093/bioinformatics/btz362">https://doi.org/10.1093/bioinformatics/btz362</a></p>
<dl class="field-list">
<dt class="field-odd">Attributes</dt>
<dd class="field-odd"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_coverage</span></code>(coverage[, pos])</p></td>
<td><p>Append coverage to this track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.Selfish.fetch_data" title="coolbox.core.track.Selfish.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw matrix should be plotted.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code>(gr[, gr2])</p></td>
<td><p>Fetch 2d contact matrix, the matrix may be processed in case ‘transform’, ‘normalize’, ‘gaussian_sigma’, ‘process_func’ exits in properties.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_track_height</span></code>(frame_width, *args)</p></td>
<td><p>calculate track height dynamically.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pile_coverages</span></code>(coverages[, pos])</p></td>
<td><p>Pile a stack of coverages with self’s coverages</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code>(ax, gr[, gr2])</p></td>
<td><p>Plot matrix</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_coverages</span></code>(ax, gr, gr2)</p></td>
<td><p>Plot all coverages on given axes.</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 74%" />
<col style="width: 26%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>JuiceBoxLikeColor</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>JuiceBoxLikeColor2</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>adjust_figure</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>diagonal_mean</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>diagonal_mean_std</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>diff_matrix</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>draw_colorbar</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>fetch_pixels</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>fetch_window_genome_range</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>fill_zero_nan</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>gaussian_matrix</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>has_prop</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>infer_binsize</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>normalize_cis</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>normalize_matrix</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>normalize_trans</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_colorbar</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_label</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_matrix</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>process_matrix</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>transform_matrix</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.Selfish.fetch_data" title="coolbox.core.track.Selfish.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw matrix should be plotted.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.Selfish.fetch_data">
<span class="sig-name descname"><span class="pre">fetch_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> &#x2192; <span class="pre">numpy.ndarray</span><a class="headerlink" href="#coolbox.core.track.Selfish.fetch_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetch the raw matrix should be plotted. Normally it’s a matrix with raw contacts</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr2</strong><span class="classifier">GenomeRange, optional, keyword argument</span></dt><dd></dd>
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>matrix</strong><span class="classifier">np.array</span></dt><dd><p>Hi-C raw contact matrix.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.track.Spacer">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.track.</span></span><span class="sig-name descname"><span class="pre">Spacer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.Spacer" title="Permalink to this definition">¶</a></dt>
<dd><p>The spacer track
Does not have any real content, just used to split two tracks.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>height</strong><span class="classifier">float, optional</span></dt><dd><p>The height of Spacer track. (Default: Spacer.DEFAULT_HEIGHT)</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional;</span></dt><dd><p>Color of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_coverage</span></code>(coverage[, pos])</p></td>
<td><p>Append coverage to this track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.Spacer.fetch_data" title="coolbox.core.track.Spacer.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code>(gr, **kwargs)</p></td>
<td><p>Fetch the plot data for the given GenomeRange.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pile_coverages</span></code>(coverages[, pos])</p></td>
<td><p>Pile a stack of coverages with self’s coverages</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.Spacer.plot" title="coolbox.core.track.Spacer.plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code></a>(ax, gr, **kwargs)</p></td>
<td><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_coverages</span></code>(ax, gr, gr2)</p></td>
<td><p>Plot all coverages on given axes.</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 58%" />
<col style="width: 42%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>has_prop</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_label</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.Spacer.fetch_data" title="coolbox.core.track.Spacer.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.Spacer.plot" title="coolbox.core.track.Spacer.plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code></a>(ax, gr, **kwargs)</p></td>
<td><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.Spacer.fetch_data">
<span class="sig-name descname"><span class="pre">fetch_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.Spacer.fetch_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetch the raw data for the given GenomeRange.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.Spacer.plot">
<span class="sig-name descname"><span class="pre">plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.Spacer.plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.track.TAD">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.track.</span></span><span class="sig-name descname"><span class="pre">TAD</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.TAD" title="Permalink to this definition">¶</a></dt>
<dd><p>Tad tack from bed file</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>border_style: str, optional</strong></dt><dd><p>Border style of tad. (Default: ‘solid’)</p>
</dd>
<dt><strong>border_width: int, optional</strong></dt><dd><p>Border width of tad. (Default: ‘2.0’)</p>
</dd>
<dt><strong>show_score</strong><span class="classifier">bool</span></dt><dd><p>Show bed score or not.
default False.</p>
</dd>
<dt><strong>score_font_size</strong><span class="classifier">{‘auto’, int}</span></dt><dd><p>Score text font size.
default ‘auto’</p>
</dd>
<dt><strong>score_font_color</strong><span class="classifier">str</span></dt><dd><p>Score text color.
default ‘#000000’</p>
</dd>
<dt><strong>score_height_ratio</strong><span class="classifier">float</span></dt><dd><p>(text tag height) / (TAD height). used for adjust the position of Score text.
default 0.5</p>
</dd>
<dt><strong>border_only</strong><span class="classifier">bool</span></dt><dd><p>Only show border, default False</p>
</dd>
<dt><strong>depth_ratio</strong><span class="classifier">{float, ‘auto’, ‘full’}</span></dt><dd><p>Depth ratio of triangular, use ‘full’ for full depth, use ‘auto’ for calculate depth by current genome_range. default ‘auto’.</p>
</dd>
<dt><strong>orientation</strong><span class="classifier">{‘normal’, ‘inverted’}</span></dt><dd><p>Invert y-axis or not, default ‘normal’</p>
</dd>
<dt><strong>file: str</strong></dt><dd><p>The file path of <cite>.bed</cite> file.</p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional</span></dt><dd><p>Track color, ‘bed_rgb’ for auto specify color according to bed record.
(Default: ‘bed_rgb’)</p>
</dd>
<dt><strong>border_color</strong><span class="classifier">str, optional</span></dt><dd><p>Border_color of gene. (Default: ‘black’)</p>
</dd>
<dt><strong>max_value</strong><span class="classifier">float, optional</span></dt><dd><p>Max score. (Default: inf)</p>
</dd>
<dt><strong>min_value</strong><span class="classifier">float, optional</span></dt><dd><p>Min score. (Default: -inf)</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_coverage</span></code>(coverage[, pos])</p></td>
<td><p>Append coverage to this track.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_intervals</span></code>(bgz_file, gr)</p></td>
<td><p>Fetch intervals within input chromosome range.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code>(gr, **kwargs)</p></td>
<td><p>Fetch the plot data for the given GenomeRange.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">intervals2dataframe</span></code>(intervals, bed_type)</p></td>
<td><p>Convert intervals list to pandas.DataFrame</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pile_coverages</span></code>(coverages[, pos])</p></td>
<td><p>Pile a stack of coverages with self’s coverages</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.TAD.plot" title="coolbox.core.track.TAD.plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code></a>(ax, gr, **kwargs)</p></td>
<td><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_coverages</span></code>(ax, gr, gr2)</p></td>
<td><p>Plot all coverages on given axes.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_tads</span></code>(ax, gr, tads)</p></td>
<td><p>Plots the boundaries as triangles in the given ax.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_triangular</span></code>(ax, gr, region, ymax, depth)</p></td>
<td><p>/             /              /            _____________________</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_colormap</span></code>(df)</p></td>
<td><p>As min_score and max_score change every plot, we compute them for every plot</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 72%" />
<col style="width: 28%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>get_depth_ratio</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>get_rgb_and_edge_color</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>get_track_height</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>has_prop</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>infer_bed_type</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>init_colormap</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>load_range</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_box</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_label</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_score</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.TAD.plot" title="coolbox.core.track.TAD.plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code></a>(ax, gr, **kwargs)</p></td>
<td><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.TAD.plot">
<span class="sig-name descname"><span class="pre">plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.TAD.plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.track.Track">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.track.</span></span><span class="sig-name descname"><span class="pre">Track</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.Track" title="Permalink to this definition">¶</a></dt>
<dd><p>Track base class.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional;</span></dt><dd><p>Color of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.Track.append_coverage" title="coolbox.core.track.Track.append_coverage"><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_coverage</span></code></a>(coverage[, pos])</p></td>
<td><p>Append coverage to this track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.Track.fetch_data" title="coolbox.core.track.Track.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.Track.fetch_plot_data" title="coolbox.core.track.Track.fetch_plot_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the plot data for the given GenomeRange.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.Track.pile_coverages" title="coolbox.core.track.Track.pile_coverages"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pile_coverages</span></code></a>(coverages[, pos])</p></td>
<td><p>Pile a stack of coverages with self’s coverages</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.Track.plot" title="coolbox.core.track.Track.plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code></a>(ax, gr, **kwargs)</p></td>
<td><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.Track.plot_coverages" title="coolbox.core.track.Track.plot_coverages"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_coverages</span></code></a>(ax, gr, gr2)</p></td>
<td><p>Plot all coverages on given axes.</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 58%" />
<col style="width: 42%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>has_prop</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_label</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.Track.append_coverage" title="coolbox.core.track.Track.append_coverage"><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_coverage</span></code></a>(coverage[, pos])</p></td>
<td><p>Append coverage to this track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.Track.fetch_data" title="coolbox.core.track.Track.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.Track.fetch_plot_data" title="coolbox.core.track.Track.fetch_plot_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the plot data for the given GenomeRange.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.Track.pile_coverages" title="coolbox.core.track.Track.pile_coverages"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pile_coverages</span></code></a>(coverages[, pos])</p></td>
<td><p>Pile a stack of coverages with self’s coverages</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.Track.plot" title="coolbox.core.track.Track.plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code></a>(ax, gr, **kwargs)</p></td>
<td><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.Track.plot_coverages" title="coolbox.core.track.Track.plot_coverages"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_coverages</span></code></a>(ax, gr, gr2)</p></td>
<td><p>Plot all coverages on given axes.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.Track.append_coverage">
<span class="sig-name descname"><span class="pre">append_coverage</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">coverage</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pos</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'top'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.Track.append_coverage" title="Permalink to this definition">¶</a></dt>
<dd><p>Append coverage to this track.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl>
<dt><strong>coverage</strong><span class="classifier"><cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverage object to be piled.</p>
</dd>
<dt><strong>pos</strong><span class="classifier">{‘top’, ‘bottom’}, optional</span></dt><dd><p>Add coverages to top or bottom. (Default: ‘top’)</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.Track.fetch_data">
<span class="sig-name descname"><span class="pre">fetch_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.Track.fetch_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetch the raw data for the given GenomeRange.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.Track.fetch_plot_data">
<span class="sig-name descname"><span class="pre">fetch_plot_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.Track.fetch_plot_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetch the plot data for the given GenomeRange.
The returned data would be used in <cite>plot</cite> function, the default implementation directly return <cite>fetch_data</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd><p>GenomeRange is an object with properties of [chrom, start, end].</p>
</dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.Track.pile_coverages">
<span class="sig-name descname"><span class="pre">pile_coverages</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">coverages</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pos</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'top'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.Track.pile_coverages" title="Permalink to this definition">¶</a></dt>
<dd><p>Pile a stack of coverages with self’s coverages</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite> or <cite>coolbox.api.coverage.CoverageStack</cite></span></dt><dd><p>Coverage objects to be piled.</p>
</dd>
<dt><strong>pos</strong><span class="classifier">{‘top’, ‘bottom’}, optional</span></dt><dd><p>Add coverages to top or bottom. (Default: ‘top’)</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.Track.plot">
<span class="sig-name descname"><span class="pre">plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.Track.plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.Track.plot_coverages">
<span class="sig-name descname"><span class="pre">plot_coverages</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">gr2</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.Track.plot_coverages" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot all coverages on given axes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr</strong><span class="classifier">GenomeRange</span></dt><dd><p>First genome range.</p>
</dd>
<dt><strong>gr2</strong><span class="classifier">GenomeRange</span></dt><dd><p>Second genome range.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.track.Virtual4C">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.track.</span></span><span class="sig-name descname"><span class="pre">Virtual4C</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.Virtual4C" title="Permalink to this definition">¶</a></dt>
<dd><p>Track for view virtual 4C related to a certain genome position,
and a HiC Track (include <cite>Cool</cite> and <cite>DotHiC</cite>).</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl>
<dt><strong>hicmat</strong><span class="classifier">{str, <cite>Cool</cite>, <cite>DotHiC</cite>}</span></dt><dd><p>Related hic track or Hi-C file path.</p>
</dd>
<dt><strong>genome_position</strong><span class="classifier">str</span></dt><dd><p>related genome position, like: ‘chr1:2000000-2000000’</p>
</dd>
<dt><strong>bin_width</strong><span class="classifier">int, optional</span></dt><dd><p>How many bin used for calculate the mean value.
default 3</p>
</dd>
<dt><strong>args_hic</strong><span class="classifier">dict, optional</span></dt><dd><p>Argument for create hic track, needed only if first argument is a path.</p>
</dd>
<dt><strong>hicmat: {str, HicMatBase}</strong></dt><dd><p>The input hicmat file or HicMatBase object used to calculate di score.</p>
</dd>
<dt><strong>style</strong><span class="classifier">str, optional</span></dt><dd><p>Track graph type, format {‘line’, ‘fill’, ‘heatmap’, ‘scatter’}</p>
</dd>
<dt><strong>fmt</strong><span class="classifier">str, optional</span></dt><dd><p>Line styles used when the plot style is <cite>line</cite>, will be passed as <cite>fmt</cite> parameter in <cite>matplotlib.pyplot.plot</cite></p>
</dd>
<dt><strong>line_width</strong><span class="classifier">int, optional</span></dt><dd><p>Value used when the plot style is <cite>line</cite>, will be passed as <cite>linewidth</cite> parameter in <cite>matplotlib.pyplot.plot</cite></p>
</dd>
<dt><strong>size</strong><span class="classifier">int, optional</span></dt><dd><p>Value used when the plot style is <cite>scatter</cite>, will be passed as <cite>s</cite> parameter in <cite>matplotlib.pyplot.scatter</cite></p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional</span></dt><dd><p>Main color</p>
</dd>
<dt><strong>threshold_color</strong><span class="classifier">str, optional</span></dt><dd><p>Track’s color for values greater than specified threshold.</p>
</dd>
<dt><strong>threshold</strong><span class="classifier">float, optional</span></dt><dd><p>Threshold used when the plot style is <cite>line</cite> or <cite>scatter</cite>, values greater than this thresh will be
plotted with <cite>color = threshold_color</cite></p>
</dd>
<dt><strong>cmap: str, optional</strong></dt><dd><p>Cmap used when the plot type is heatmap, will be passed as <cite>cmap</cite> paramerter in <cite>matplotlib.pyplot.matshow</cite></p>
</dd>
<dt><strong>alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Alpha value of plot, default 1.0</p>
</dd>
<dt><strong>orientation</strong><span class="classifier">str, optional</span></dt><dd><p>Track orientation, use ‘inverted’ for inverted track plot.</p>
</dd>
<dt><strong>data_range_style</strong><span class="classifier">{bool, ‘text’, ‘y-axis’}, optional</span></dt><dd><p>Show_data_range or not, default True.</p>
</dd>
<dt><strong>max_value</strong><span class="classifier">{float, ‘auto’}, optional</span></dt><dd><p>Max value of track. ‘auto’ for specify max value automatically, default ‘auto’.</p>
</dd>
<dt><strong>min_value</strong><span class="classifier">{float, ‘auto’}, optional</span></dt><dd><p>Min value of track. ‘auto’ for specify max value automatically, default ‘auto’.</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_coverage</span></code>(coverage[, pos])</p></td>
<td><p>Append coverage to this track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.Virtual4C.fetch_data" title="coolbox.core.track.Virtual4C.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pile_coverages</span></code>(coverages[, pos])</p></td>
<td><p>Pile a stack of coverages with self’s coverages</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code>(ax, gr, **kwargs)</p></td>
<td><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_coverages</span></code>(ax, gr, gr2)</p></td>
<td><p>Plot all coverages on given axes.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_yaxis_range</span></code>(plot_axis, y_ax)</p></td>
<td><p>Plot the scale of the y axis with respect to the plot_axis</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 66%" />
<col style="width: 34%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>adjust_plot</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>has_prop</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_data_range</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_fill</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_heatmap</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_hist</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_label</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_line</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_scatter</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_stairs</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_text_range</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.Virtual4C.fetch_data" title="coolbox.core.track.Virtual4C.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.Virtual4C.fetch_data">
<span class="sig-name descname"><span class="pre">fetch_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> &#x2192; <span class="pre">numpy.ndarray</span><a class="headerlink" href="#coolbox.core.track.Virtual4C.fetch_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetch the raw data for the given GenomeRange.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.track.XAxis">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.track.</span></span><span class="sig-name descname"><span class="pre">XAxis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.XAxis" title="Permalink to this definition">¶</a></dt>
<dd><p>The x axis track.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>height</strong><span class="classifier">float, optional</span></dt><dd><p>Height of Spacer track. (Default: XAxis.DEFAULT_HEIGHT)</p>
</dd>
<dt><strong>fontsize</strong><span class="classifier">int, optional</span></dt><dd><p>Font size of XAxis. (Default: XAxis.DEFAULT_FONTSIZE)</p>
</dd>
<dt><strong>where</strong><span class="classifier">{‘top’, ‘bottom’}, optional</span></dt><dd><p>The position of tick labels relative to the axis.
(Default: ‘bottom’)</p>
</dd>
<dt><strong>name (str, optional):</strong></dt><dd><p>Track’s name.</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional;</span></dt><dd><p>Color of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_coverage</span></code>(coverage[, pos])</p></td>
<td><p>Append coverage to this track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.XAxis.fetch_data" title="coolbox.core.track.XAxis.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code>(gr, **kwargs)</p></td>
<td><p>Fetch the plot data for the given GenomeRange.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pile_coverages</span></code>(coverages[, pos])</p></td>
<td><p>Pile a stack of coverages with self’s coverages</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.XAxis.plot" title="coolbox.core.track.XAxis.plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code></a>(ax, gr, **kwargs)</p></td>
<td><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_coverages</span></code>(ax, gr, gr2)</p></td>
<td><p>Plot all coverages on given axes.</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 58%" />
<col style="width: 42%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>has_prop</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_label</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.track.XAxis.fetch_data" title="coolbox.core.track.XAxis.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.track.XAxis.plot" title="coolbox.core.track.XAxis.plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code></a>(ax, gr, **kwargs)</p></td>
<td><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.XAxis.fetch_data">
<span class="sig-name descname"><span class="pre">fetch_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.XAxis.fetch_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetch the raw data for the given GenomeRange.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.track.XAxis.plot">
<span class="sig-name descname"><span class="pre">plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.track.XAxis.plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-coolbox.core.coverage">
<span id="coverage"></span><h2>Coverage<a class="headerlink" href="#module-coolbox.core.coverage" title="Permalink to this headline">¶</a></h2>
<p><strong>Functions:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.coverage.Arcs" title="coolbox.core.coverage.Arcs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Arcs</span></code></a>(file_, *args, **kwargs)</p></td>
<td><p>Compose BEDPE, Pairs track automatically based on type of file extension.(.bedpe, .pairs)</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.coverage.Hist" title="coolbox.core.coverage.Hist"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Hist</span></code></a>(file, *args, **kwargs)</p></td>
<td><p>Compose BigWig, BedGraph, SNP, BAMCov track automatically based on type of file extension.(.bw, .bigwig, .snp, .vcf, .bedgraph, .bam, .sam)</p></td>
</tr>
</tbody>
</table>
<p><strong>Classes:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.coverage.ArcsCoverage" title="coolbox.core.coverage.ArcsCoverage"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ArcsCoverage</span></code></a>(*args, **kwargs)</p></td>
<td><p>Compose BEDPE, Pairs track automatically based on type of file extension.(.bedpe, .pairs)</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.coverage.BEDPE" title="coolbox.core.coverage.BEDPE"><code class="xref py py-obj docutils literal notranslate"><span class="pre">BEDPE</span></code></a>(*args, **kwargs)</p></td>
<td><p>Arcs track from .bedpe file.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.coverage.BEDPECoverage" title="coolbox.core.coverage.BEDPECoverage"><code class="xref py py-obj docutils literal notranslate"><span class="pre">BEDPECoverage</span></code></a>(*args, **kwargs)</p></td>
<td><p>Arcs track from .bedpe file.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.coverage.BedGraph" title="coolbox.core.coverage.BedGraph"><code class="xref py py-obj docutils literal notranslate"><span class="pre">BedGraph</span></code></a>(*args, **kwargs)</p></td>
<td><p>BedGraph track.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.coverage.BedGraphCoverage" title="coolbox.core.coverage.BedGraphCoverage"><code class="xref py py-obj docutils literal notranslate"><span class="pre">BedGraphCoverage</span></code></a>(*args, **kwargs)</p></td>
<td><p>BedGraph track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.coverage.BigWig" title="coolbox.core.coverage.BigWig"><code class="xref py py-obj docutils literal notranslate"><span class="pre">BigWig</span></code></a>(*args, **kwargs)</p></td>
<td><p>BigWig track</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.coverage.BigWigCoverage" title="coolbox.core.coverage.BigWigCoverage"><code class="xref py py-obj docutils literal notranslate"><span class="pre">BigWigCoverage</span></code></a>(*args, **kwargs)</p></td>
<td><p>BigWig track</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.coverage.HLines" title="coolbox.core.coverage.HLines"><code class="xref py py-obj docutils literal notranslate"><span class="pre">HLines</span></code></a>(*args, **kwargs)</p></td>
<td><p>Horizontal line coverage, for show threshold.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.coverage.HiCPeaks" title="coolbox.core.coverage.HiCPeaks"><code class="xref py py-obj docutils literal notranslate"><span class="pre">HiCPeaks</span></code></a>(*args, **kwargs)</p></td>
<td><p>Hi-C Peaks(Loops) from .bedpe file.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.coverage.HiCPeaksCoverage" title="coolbox.core.coverage.HiCPeaksCoverage"><code class="xref py py-obj docutils literal notranslate"><span class="pre">HiCPeaksCoverage</span></code></a>(*args, **kwargs)</p></td>
<td><p>Hi-C Peaks(Loops) from .bedpe file.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.coverage.HighLights" title="coolbox.core.coverage.HighLights"><code class="xref py py-obj docutils literal notranslate"><span class="pre">HighLights</span></code></a>(*args, **kwargs)</p></td>
<td><p>High light region.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.coverage.HighLightsFromFile" title="coolbox.core.coverage.HighLightsFromFile"><code class="xref py py-obj docutils literal notranslate"><span class="pre">HighLightsFromFile</span></code></a>(*args, **kwargs)</p></td>
<td><p>High light regions coverage, read the regions from the file.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.coverage.HistCoverage" title="coolbox.core.coverage.HistCoverage"><code class="xref py py-obj docutils literal notranslate"><span class="pre">HistCoverage</span></code></a>(*args, **kwargs)</p></td>
<td><p>Compose BigWig, BedGraph, SNP, BAMCov track automatically based on type of file extension.(.bw, .bigwig, .snp, .vcf, .bedgraph, .bam, .sam)</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.coverage.Pairs" title="coolbox.core.coverage.Pairs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Pairs</span></code></a>(*args, **kwargs)</p></td>
<td><p>Arcs track from .pairs file.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.coverage.PairsCoverage" title="coolbox.core.coverage.PairsCoverage"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PairsCoverage</span></code></a>(*args, **kwargs)</p></td>
<td><p>Arcs track from .pairs file.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.coverage.TAD" title="coolbox.core.coverage.TAD"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TAD</span></code></a>(*args, **kwargs)</p></td>
<td><p>Tad tack from bed file</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.coverage.TADCoverage" title="coolbox.core.coverage.TADCoverage"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TADCoverage</span></code></a>(*args, **kwargs)</p></td>
<td><p>Tad tack from bed file</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.coverage.Vlines" title="coolbox.core.coverage.Vlines"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Vlines</span></code></a>(*args, **kwargs)</p></td>
<td><p>Vertical lines.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.coverage.VlinesFromFile" title="coolbox.core.coverage.VlinesFromFile"><code class="xref py py-obj docutils literal notranslate"><span class="pre">VlinesFromFile</span></code></a>(*args, **kwargs)</p></td>
<td><p>Vertical lines from the file.</p></td>
</tr>
</tbody>
</table>
<dl class="py function">
<dt class="sig sig-object py" id="coolbox.core.coverage.Arcs">
<span class="sig-prename descclassname"><span class="pre">coolbox.core.coverage.</span></span><span class="sig-name descname"><span class="pre">Arcs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> &#x2192; <a class="reference internal" href="#coolbox.core.track.ArcsBase" title="coolbox.core.track.arcs.base.ArcsBase"><span class="pre">coolbox.core.track.arcs.base.ArcsBase</span></a><a class="headerlink" href="#coolbox.core.coverage.Arcs" title="Permalink to this definition">¶</a></dt>
<dd><p>Compose BEDPE, Pairs track automatically based on type of file extension.(.bedpe, .pairs)</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.coverage.ArcsCoverage">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.coverage.</span></span><span class="sig-name descname"><span class="pre">ArcsCoverage</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.coverage.ArcsCoverage" title="Permalink to this definition">¶</a></dt>
<dd><p>Compose BEDPE, Pairs track automatically based on type of file extension.(.bedpe, .pairs)</p>
<dl class="field-list simple">
<dt class="field-odd">Attributes</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>name</strong></dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 67%" />
<col style="width: 33%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>check_track_type</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.coverage.BEDPE">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.coverage.</span></span><span class="sig-name descname"><span class="pre">BEDPE</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.coverage.BEDPE" title="Permalink to this definition">¶</a></dt>
<dd><p>Arcs track from .bedpe file.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>file: str</strong></dt><dd><p>Path of .bedpe file</p>
</dd>
<dt><strong>pos</strong><span class="classifier">str, optional</span></dt><dd><p>Method for choosing arch anchor for bedpe data: ‘start’, ‘end’, ‘mid’, default ‘mid’</p>
</dd>
<dt><strong>style</strong><span class="classifier">str, optional</span></dt><dd><p>Style of arcs plot: ‘arcs’, ‘hicpeaks’, default ‘arcs’</p>
</dd>
<dt><strong>score_to_width</strong><span class="classifier">str, optional</span></dt><dd><p>Mapping function of score to width, default: ‘0.5 + math.sqrt(score)’</p>
</dd>
<dt><strong>line_width</strong><span class="classifier">float, optional</span></dt><dd><p>Width of arc line.</p>
</dd>
<dt><strong>open_region</strong><span class="classifier">bool, optional</span></dt><dd><p>If specified to True, will fetch the contacts on side in the region,
default True</p>
</dd>
<dt><strong>diameter_to_height</strong><span class="classifier">str, optional</span></dt><dd><p>Mapping function of arc diameter(interval end - start) to height.
You can specify to ‘max_diameter’ let all arcs has same height.
default ‘max_height * diameter / max_diameter’.</p>
</dd>
<dt><strong>orientation</strong><span class="classifier">str, optional</span></dt><dd><p>Track orientation, use ‘inverted’ for inverted track plot. default None</p>
</dd>
<dt><strong>line_style</strong><span class="classifier">str, optional</span></dt><dd><p>Border line style, default ‘solid’</p>
</dd>
<dt><strong>fill</strong><span class="classifier">bool, optional</span></dt><dd><p>Fill center or not, default False.</p>
</dd>
<dt><strong>fill_color</strong><span class="classifier">str, optional</span></dt><dd><p>Fill color, use ‘bed_rgb’ for specify color in file,
default ‘bed_rgb’.</p>
</dd>
<dt><strong>fill_alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Alpha value of fill region.
default 0.2</p>
</dd>
<dt><strong>side</strong><span class="classifier">{‘upper’, ‘lower’, ‘both’}</span></dt><dd><p>Plot peak in which side of the matrix.
NOTE: This parameters is useful only if the Cool track in matrix format.</p>
</dd>
<dt><strong>alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Alpha value of track, default 0.8.</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional;</span></dt><dd><p>Color of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_coverage</span></code>(coverage[, pos])</p></td>
<td><p>Append coverage to this track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.coverage.BEDPE.fetch_data" title="coolbox.core.coverage.BEDPE.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.coverage.BEDPE.fetch_plot_data" title="coolbox.core.coverage.BEDPE.fetch_plot_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code></a>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pile_coverages</span></code>(coverages[, pos])</p></td>
<td><p>Pile a stack of coverages with self’s coverages</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code>(ax, gr, **kwargs)</p></td>
<td><p>Plot arc connecting two positions on a linear scale representing interactions between bins.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_arcs</span></code>(ax, gr, gr2, intervals)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_coverages</span></code>(ax, gr, gr2)</p></td>
<td><p>Plot all coverages on given axes.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_hicpeaks</span></code>(ax, gr, gr2, intervals)</p></td>
<td><p>plot hic peaks uppon a HicMatBase track.</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 66%" />
<col style="width: 34%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>fetch_intervals</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>has_prop</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_contacts</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_label</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.coverage.BEDPE.fetch_data" title="coolbox.core.coverage.BEDPE.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.coverage.BEDPE.fetch_plot_data" title="coolbox.core.coverage.BEDPE.fetch_plot_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code></a>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.coverage.BEDPE.fetch_data">
<span class="sig-name descname"><span class="pre">fetch_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> &#x2192; <span class="pre">pandas.core.frame.DataFrame</span><a class="headerlink" href="#coolbox.core.coverage.BEDPE.fetch_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetch the raw data for the given GenomeRange.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.coverage.BEDPE.fetch_plot_data">
<span class="sig-name descname"><span class="pre">fetch_plot_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> &#x2192; <span class="pre">pandas.core.frame.DataFrame</span><a class="headerlink" href="#coolbox.core.coverage.BEDPE.fetch_plot_data" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd><p>GenomeRange is an object with properties of [chrom, start, end].</p>
</dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>intervals</strong><span class="classifier">pandas.core.frame.DataFrame</span></dt><dd><p>Can be two types:
1: with columns: [‘pos1’, ‘pos2’, ‘score’] ‘score’ is optional
2: with columns: [‘start1’, ‘end1’, ‘start2’, ‘end2’, ‘score’] ‘score’ is optional</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.coverage.BEDPECoverage">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.coverage.</span></span><span class="sig-name descname"><span class="pre">BEDPECoverage</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.coverage.BEDPECoverage" title="Permalink to this definition">¶</a></dt>
<dd><p>Arcs track from .bedpe file.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>file: str</strong></dt><dd><p>Path of .bedpe file</p>
</dd>
<dt><strong>pos</strong><span class="classifier">str, optional</span></dt><dd><p>Method for choosing arch anchor for bedpe data: ‘start’, ‘end’, ‘mid’, default ‘mid’</p>
</dd>
<dt><strong>style</strong><span class="classifier">str, optional</span></dt><dd><p>Style of arcs plot: ‘arcs’, ‘hicpeaks’, default ‘arcs’</p>
</dd>
<dt><strong>score_to_width</strong><span class="classifier">str, optional</span></dt><dd><p>Mapping function of score to width, default: ‘0.5 + math.sqrt(score)’</p>
</dd>
<dt><strong>line_width</strong><span class="classifier">float, optional</span></dt><dd><p>Width of arc line.</p>
</dd>
<dt><strong>open_region</strong><span class="classifier">bool, optional</span></dt><dd><p>If specified to True, will fetch the contacts on side in the region,
default True</p>
</dd>
<dt><strong>diameter_to_height</strong><span class="classifier">str, optional</span></dt><dd><p>Mapping function of arc diameter(interval end - start) to height.
You can specify to ‘max_diameter’ let all arcs has same height.
default ‘max_height * diameter / max_diameter’.</p>
</dd>
<dt><strong>orientation</strong><span class="classifier">str, optional</span></dt><dd><p>Track orientation, use ‘inverted’ for inverted track plot. default None</p>
</dd>
<dt><strong>line_style</strong><span class="classifier">str, optional</span></dt><dd><p>Border line style, default ‘solid’</p>
</dd>
<dt><strong>fill</strong><span class="classifier">bool, optional</span></dt><dd><p>Fill center or not, default False.</p>
</dd>
<dt><strong>fill_color</strong><span class="classifier">str, optional</span></dt><dd><p>Fill color, use ‘bed_rgb’ for specify color in file,
default ‘bed_rgb’.</p>
</dd>
<dt><strong>fill_alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Alpha value of fill region.
default 0.2</p>
</dd>
<dt><strong>side</strong><span class="classifier">{‘upper’, ‘lower’, ‘both’}</span></dt><dd><p>Plot peak in which side of the matrix.
NOTE: This parameters is useful only if the Cool track in matrix format.</p>
</dd>
<dt><strong>alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Alpha value of track, default 0.8.</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional;</span></dt><dd><p>Color of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 67%" />
<col style="width: 33%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>check_track_type</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.coverage.BedGraph">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.coverage.</span></span><span class="sig-name descname"><span class="pre">BedGraph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.coverage.BedGraph" title="Permalink to this definition">¶</a></dt>
<dd><p>BedGraph track.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>file</strong><span class="classifier">str</span></dt><dd><p>File path of bedgraph file.</p>
</dd>
<dt><strong>style</strong><span class="classifier">str, optional</span></dt><dd><p>Track graph type, format {‘line’, ‘fill’, ‘heatmap’, ‘scatter’}</p>
</dd>
<dt><strong>fmt</strong><span class="classifier">str, optional</span></dt><dd><p>Line styles used when the plot style is <cite>line</cite>, will be passed as <cite>fmt</cite> parameter in <cite>matplotlib.pyplot.plot</cite></p>
</dd>
<dt><strong>line_width</strong><span class="classifier">int, optional</span></dt><dd><p>Value used when the plot style is <cite>line</cite>, will be passed as <cite>linewidth</cite> parameter in <cite>matplotlib.pyplot.plot</cite></p>
</dd>
<dt><strong>size</strong><span class="classifier">int, optional</span></dt><dd><p>Value used when the plot style is <cite>scatter</cite>, will be passed as <cite>s</cite> parameter in <cite>matplotlib.pyplot.scatter</cite></p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional</span></dt><dd><p>Main color</p>
</dd>
<dt><strong>threshold_color</strong><span class="classifier">str, optional</span></dt><dd><p>Track’s color for values greater than specified threshold.</p>
</dd>
<dt><strong>threshold</strong><span class="classifier">float, optional</span></dt><dd><p>Threshold used when the plot style is <cite>line</cite> or <cite>scatter</cite>, values greater than this thresh will be
plotted with <cite>color = threshold_color</cite></p>
</dd>
<dt><strong>cmap: str, optional</strong></dt><dd><p>Cmap used when the plot type is heatmap, will be passed as <cite>cmap</cite> paramerter in <cite>matplotlib.pyplot.matshow</cite></p>
</dd>
<dt><strong>alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Alpha value of plot, default 1.0</p>
</dd>
<dt><strong>orientation</strong><span class="classifier">str, optional</span></dt><dd><p>Track orientation, use ‘inverted’ for inverted track plot.</p>
</dd>
<dt><strong>data_range_style</strong><span class="classifier">{bool, ‘text’, ‘y-axis’}, optional</span></dt><dd><p>Show_data_range or not, default True.</p>
</dd>
<dt><strong>max_value</strong><span class="classifier">{float, ‘auto’}, optional</span></dt><dd><p>Max value of track. ‘auto’ for specify max value automatically, default ‘auto’.</p>
</dd>
<dt><strong>min_value</strong><span class="classifier">{float, ‘auto’}, optional</span></dt><dd><p>Min value of track. ‘auto’ for specify max value automatically, default ‘auto’.</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_coverage</span></code>(coverage[, pos])</p></td>
<td><p>Append coverage to this track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.coverage.BedGraph.fetch_data" title="coolbox.core.coverage.BedGraph.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.coverage.BedGraph.fetch_plot_data" title="coolbox.core.coverage.BedGraph.fetch_plot_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code></a>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pile_coverages</span></code>(coverages[, pos])</p></td>
<td><p>Pile a stack of coverages with self’s coverages</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code>(ax, gr, **kwargs)</p></td>
<td><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_coverages</span></code>(ax, gr, gr2)</p></td>
<td><p>Plot all coverages on given axes.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_yaxis_range</span></code>(plot_axis, y_ax)</p></td>
<td><p>Plot the scale of the y axis with respect to the plot_axis</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 66%" />
<col style="width: 34%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>adjust_plot</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>has_prop</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>load</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_data_range</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_fill</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_heatmap</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_hist</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_label</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_line</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_scatter</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_stairs</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_text_range</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.coverage.BedGraph.fetch_data" title="coolbox.core.coverage.BedGraph.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.coverage.BedGraph.fetch_plot_data" title="coolbox.core.coverage.BedGraph.fetch_plot_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code></a>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.coverage.BedGraph.fetch_data">
<span class="sig-name descname"><span class="pre">fetch_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> &#x2192; <span class="pre">pandas.core.frame.DataFrame</span><a class="headerlink" href="#coolbox.core.coverage.BedGraph.fetch_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetch the raw data for the given GenomeRange.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.coverage.BedGraph.fetch_plot_data">
<span class="sig-name descname"><span class="pre">fetch_plot_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> &#x2192; <span class="pre">pandas.core.frame.DataFrame</span><a class="headerlink" href="#coolbox.core.coverage.BedGraph.fetch_plot_data" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd><p>GenomeRange is an object with properties of [chrom, start, end].</p>
</dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>hist_data: Union[Tuple[np.ndarray, np.ndarray], pd.DataFrame, np.ndarray]</dt><dd><p>data used for plotting.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.coverage.BedGraphCoverage">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.coverage.</span></span><span class="sig-name descname"><span class="pre">BedGraphCoverage</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.coverage.BedGraphCoverage" title="Permalink to this definition">¶</a></dt>
<dd><p>BedGraph track.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>file</strong><span class="classifier">str</span></dt><dd><p>File path of bedgraph file.</p>
</dd>
<dt><strong>style</strong><span class="classifier">str, optional</span></dt><dd><p>Track graph type, format {‘line’, ‘fill’, ‘heatmap’, ‘scatter’}</p>
</dd>
<dt><strong>fmt</strong><span class="classifier">str, optional</span></dt><dd><p>Line styles used when the plot style is <cite>line</cite>, will be passed as <cite>fmt</cite> parameter in <cite>matplotlib.pyplot.plot</cite></p>
</dd>
<dt><strong>line_width</strong><span class="classifier">int, optional</span></dt><dd><p>Value used when the plot style is <cite>line</cite>, will be passed as <cite>linewidth</cite> parameter in <cite>matplotlib.pyplot.plot</cite></p>
</dd>
<dt><strong>size</strong><span class="classifier">int, optional</span></dt><dd><p>Value used when the plot style is <cite>scatter</cite>, will be passed as <cite>s</cite> parameter in <cite>matplotlib.pyplot.scatter</cite></p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional</span></dt><dd><p>Main color</p>
</dd>
<dt><strong>threshold_color</strong><span class="classifier">str, optional</span></dt><dd><p>Track’s color for values greater than specified threshold.</p>
</dd>
<dt><strong>threshold</strong><span class="classifier">float, optional</span></dt><dd><p>Threshold used when the plot style is <cite>line</cite> or <cite>scatter</cite>, values greater than this thresh will be
plotted with <cite>color = threshold_color</cite></p>
</dd>
<dt><strong>cmap: str, optional</strong></dt><dd><p>Cmap used when the plot type is heatmap, will be passed as <cite>cmap</cite> paramerter in <cite>matplotlib.pyplot.matshow</cite></p>
</dd>
<dt><strong>alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Alpha value of plot, default 1.0</p>
</dd>
<dt><strong>orientation</strong><span class="classifier">str, optional</span></dt><dd><p>Track orientation, use ‘inverted’ for inverted track plot.</p>
</dd>
<dt><strong>data_range_style</strong><span class="classifier">{bool, ‘text’, ‘y-axis’}, optional</span></dt><dd><p>Show_data_range or not, default True.</p>
</dd>
<dt><strong>max_value</strong><span class="classifier">{float, ‘auto’}, optional</span></dt><dd><p>Max value of track. ‘auto’ for specify max value automatically, default ‘auto’.</p>
</dd>
<dt><strong>min_value</strong><span class="classifier">{float, ‘auto’}, optional</span></dt><dd><p>Min value of track. ‘auto’ for specify max value automatically, default ‘auto’.</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 67%" />
<col style="width: 33%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>check_track_type</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.coverage.BigWig">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.coverage.</span></span><span class="sig-name descname"><span class="pre">BigWig</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.coverage.BigWig" title="Permalink to this definition">¶</a></dt>
<dd><p>BigWig track</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>file</strong><span class="classifier">str</span></dt><dd><p>File path of bigwig file.</p>
</dd>
<dt><strong>num_bins</strong><span class="classifier">int, optional</span></dt><dd><p>Number of bins to plot the hist in current range, default 700.</p>
</dd>
<dt><strong>style</strong><span class="classifier">str, optional</span></dt><dd><p>Track graph type, format {‘line’, ‘fill’, ‘heatmap’, ‘scatter’}</p>
</dd>
<dt><strong>fmt</strong><span class="classifier">str, optional</span></dt><dd><p>Line styles used when the plot style is <cite>line</cite>, will be passed as <cite>fmt</cite> parameter in <cite>matplotlib.pyplot.plot</cite></p>
</dd>
<dt><strong>line_width</strong><span class="classifier">int, optional</span></dt><dd><p>Value used when the plot style is <cite>line</cite>, will be passed as <cite>linewidth</cite> parameter in <cite>matplotlib.pyplot.plot</cite></p>
</dd>
<dt><strong>size</strong><span class="classifier">int, optional</span></dt><dd><p>Value used when the plot style is <cite>scatter</cite>, will be passed as <cite>s</cite> parameter in <cite>matplotlib.pyplot.scatter</cite></p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional</span></dt><dd><p>Main color</p>
</dd>
<dt><strong>threshold_color</strong><span class="classifier">str, optional</span></dt><dd><p>Track’s color for values greater than specified threshold.</p>
</dd>
<dt><strong>threshold</strong><span class="classifier">float, optional</span></dt><dd><p>Threshold used when the plot style is <cite>line</cite> or <cite>scatter</cite>, values greater than this thresh will be
plotted with <cite>color = threshold_color</cite></p>
</dd>
<dt><strong>cmap: str, optional</strong></dt><dd><p>Cmap used when the plot type is heatmap, will be passed as <cite>cmap</cite> paramerter in <cite>matplotlib.pyplot.matshow</cite></p>
</dd>
<dt><strong>alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Alpha value of plot, default 1.0</p>
</dd>
<dt><strong>orientation</strong><span class="classifier">str, optional</span></dt><dd><p>Track orientation, use ‘inverted’ for inverted track plot.</p>
</dd>
<dt><strong>data_range_style</strong><span class="classifier">{bool, ‘text’, ‘y-axis’}, optional</span></dt><dd><p>Show_data_range or not, default True.</p>
</dd>
<dt><strong>max_value</strong><span class="classifier">{float, ‘auto’}, optional</span></dt><dd><p>Max value of track. ‘auto’ for specify max value automatically, default ‘auto’.</p>
</dd>
<dt><strong>min_value</strong><span class="classifier">{float, ‘auto’}, optional</span></dt><dd><p>Min value of track. ‘auto’ for specify max value automatically, default ‘auto’.</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_coverage</span></code>(coverage[, pos])</p></td>
<td><p>Append coverage to this track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.coverage.BigWig.fetch_data" title="coolbox.core.coverage.BigWig.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.coverage.BigWig.fetch_plot_data" title="coolbox.core.coverage.BigWig.fetch_plot_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code></a>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.coverage.BigWig.fetch_scores" title="coolbox.core.coverage.BigWig.fetch_scores"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_scores</span></code></a>(genome_range, num_bins[, …])</p></td>
<td><p>Fetch bins scores within input chromosome range.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pile_coverages</span></code>(coverages[, pos])</p></td>
<td><p>Pile a stack of coverages with self’s coverages</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code>(ax, gr, **kwargs)</p></td>
<td><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_coverages</span></code>(ax, gr, gr2)</p></td>
<td><p>Plot all coverages on given axes.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_yaxis_range</span></code>(plot_axis, y_ax)</p></td>
<td><p>Plot the scale of the y axis with respect to the plot_axis</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 67%" />
<col style="width: 33%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>adjust_plot</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>check_chrom_name</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>get_num_bins</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>has_prop</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_data_range</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_fill</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_heatmap</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_hist</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_label</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_line</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_scatter</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_stairs</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_text_range</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.coverage.BigWig.fetch_data" title="coolbox.core.coverage.BigWig.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.coverage.BigWig.fetch_plot_data" title="coolbox.core.coverage.BigWig.fetch_plot_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code></a>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.coverage.BigWig.fetch_scores" title="coolbox.core.coverage.BigWig.fetch_scores"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_scores</span></code></a>(genome_range, num_bins[, …])</p></td>
<td><p>Fetch bins scores within input chromosome range.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.coverage.BigWig.fetch_data">
<span class="sig-name descname"><span class="pre">fetch_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.coverage.BigWig.fetch_data" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr</strong><span class="classifier">GenomeRange</span></dt><dd></dd>
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.coverage.BigWig.fetch_plot_data">
<span class="sig-name descname"><span class="pre">fetch_plot_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.coverage.BigWig.fetch_plot_data" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd><p>GenomeRange is an object with properties of [chrom, start, end].</p>
</dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>hist_data: Union[Tuple[np.ndarray, np.ndarray], pd.DataFrame, np.ndarray]</dt><dd><p>data used for plotting.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.coverage.BigWig.fetch_scores">
<span class="sig-name descname"><span class="pre">fetch_scores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome_range</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_bins</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_try_nums</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.coverage.BigWig.fetch_scores" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetch bins scores within input chromosome range.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.coverage.BigWigCoverage">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.coverage.</span></span><span class="sig-name descname"><span class="pre">BigWigCoverage</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.coverage.BigWigCoverage" title="Permalink to this definition">¶</a></dt>
<dd><p>BigWig track</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>file</strong><span class="classifier">str</span></dt><dd><p>File path of bigwig file.</p>
</dd>
<dt><strong>num_bins</strong><span class="classifier">int, optional</span></dt><dd><p>Number of bins to plot the hist in current range, default 700.</p>
</dd>
<dt><strong>style</strong><span class="classifier">str, optional</span></dt><dd><p>Track graph type, format {‘line’, ‘fill’, ‘heatmap’, ‘scatter’}</p>
</dd>
<dt><strong>fmt</strong><span class="classifier">str, optional</span></dt><dd><p>Line styles used when the plot style is <cite>line</cite>, will be passed as <cite>fmt</cite> parameter in <cite>matplotlib.pyplot.plot</cite></p>
</dd>
<dt><strong>line_width</strong><span class="classifier">int, optional</span></dt><dd><p>Value used when the plot style is <cite>line</cite>, will be passed as <cite>linewidth</cite> parameter in <cite>matplotlib.pyplot.plot</cite></p>
</dd>
<dt><strong>size</strong><span class="classifier">int, optional</span></dt><dd><p>Value used when the plot style is <cite>scatter</cite>, will be passed as <cite>s</cite> parameter in <cite>matplotlib.pyplot.scatter</cite></p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional</span></dt><dd><p>Main color</p>
</dd>
<dt><strong>threshold_color</strong><span class="classifier">str, optional</span></dt><dd><p>Track’s color for values greater than specified threshold.</p>
</dd>
<dt><strong>threshold</strong><span class="classifier">float, optional</span></dt><dd><p>Threshold used when the plot style is <cite>line</cite> or <cite>scatter</cite>, values greater than this thresh will be
plotted with <cite>color = threshold_color</cite></p>
</dd>
<dt><strong>cmap: str, optional</strong></dt><dd><p>Cmap used when the plot type is heatmap, will be passed as <cite>cmap</cite> paramerter in <cite>matplotlib.pyplot.matshow</cite></p>
</dd>
<dt><strong>alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Alpha value of plot, default 1.0</p>
</dd>
<dt><strong>orientation</strong><span class="classifier">str, optional</span></dt><dd><p>Track orientation, use ‘inverted’ for inverted track plot.</p>
</dd>
<dt><strong>data_range_style</strong><span class="classifier">{bool, ‘text’, ‘y-axis’}, optional</span></dt><dd><p>Show_data_range or not, default True.</p>
</dd>
<dt><strong>max_value</strong><span class="classifier">{float, ‘auto’}, optional</span></dt><dd><p>Max value of track. ‘auto’ for specify max value automatically, default ‘auto’.</p>
</dd>
<dt><strong>min_value</strong><span class="classifier">{float, ‘auto’}, optional</span></dt><dd><p>Min value of track. ‘auto’ for specify max value automatically, default ‘auto’.</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 67%" />
<col style="width: 33%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>check_track_type</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.coverage.HLines">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.coverage.</span></span><span class="sig-name descname"><span class="pre">HLines</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.coverage.HLines" title="Permalink to this definition">¶</a></dt>
<dd><p>Horizontal line coverage, for show threshold.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>values</strong><span class="classifier">float</span></dt><dd><p>Y position. A list of y value or a single value.</p>
</dd>
<dt><strong>percent_mode</strong><span class="classifier">bool</span></dt><dd><p>If in percent mode, y position will equal to ymin + (ymax - ymin) * val.</p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional</span></dt><dd><p>Line color, default ‘#1e1e1e’.</p>
</dd>
<dt><strong>alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Line alpha value, default 0.8.</p>
</dd>
<dt><strong>line_style</strong><span class="classifier">str, optional</span></dt><dd><p>Line style, default ‘dashed’.</p>
</dd>
<dt><strong>line_width</strong><span class="classifier">float, optional</span></dt><dd><p>Line width, default 0.5.</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of thr Coverage.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>name</strong></dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 67%" />
<col style="width: 33%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>check_track_type</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.coverage.HiCPeaks">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.coverage.</span></span><span class="sig-name descname"><span class="pre">HiCPeaks</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.coverage.HiCPeaks" title="Permalink to this definition">¶</a></dt>
<dd><p>Hi-C Peaks(Loops) from .bedpe file. Used to show the peaks on the Hi-C interaction map.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>file: str</strong></dt><dd><p>Path of .bedpe file</p>
</dd>
<dt><strong>pos</strong><span class="classifier">str, optional</span></dt><dd><p>Method for choosing arch anchor for bedpe data: ‘start’, ‘end’, ‘mid’, default ‘mid’</p>
</dd>
<dt><strong>style</strong><span class="classifier">str, optional</span></dt><dd><p>Style of arcs plot: ‘arcs’, ‘hicpeaks’, default ‘arcs’</p>
</dd>
<dt><strong>score_to_width</strong><span class="classifier">str, optional</span></dt><dd><p>Mapping function of score to width, default: ‘0.5 + math.sqrt(score)’</p>
</dd>
<dt><strong>line_width</strong><span class="classifier">float, optional</span></dt><dd><p>Width of arc line.</p>
</dd>
<dt><strong>open_region</strong><span class="classifier">bool, optional</span></dt><dd><p>If specified to True, will fetch the contacts on side in the region,
default True</p>
</dd>
<dt><strong>diameter_to_height</strong><span class="classifier">str, optional</span></dt><dd><p>Mapping function of arc diameter(interval end - start) to height.
You can specify to ‘max_diameter’ let all arcs has same height.
default ‘max_height * diameter / max_diameter’.</p>
</dd>
<dt><strong>orientation</strong><span class="classifier">str, optional</span></dt><dd><p>Track orientation, use ‘inverted’ for inverted track plot. default None</p>
</dd>
<dt><strong>line_style</strong><span class="classifier">str, optional</span></dt><dd><p>Border line style, default ‘solid’</p>
</dd>
<dt><strong>fill</strong><span class="classifier">bool, optional</span></dt><dd><p>Fill center or not, default False.</p>
</dd>
<dt><strong>fill_color</strong><span class="classifier">str, optional</span></dt><dd><p>Fill color, use ‘bed_rgb’ for specify color in file,
default ‘bed_rgb’.</p>
</dd>
<dt><strong>fill_alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Alpha value of fill region.
default 0.2</p>
</dd>
<dt><strong>side</strong><span class="classifier">{‘upper’, ‘lower’, ‘both’}</span></dt><dd><p>Plot peak in which side of the matrix.
NOTE: This parameters is useful only if the Cool track in matrix format.</p>
</dd>
<dt><strong>alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Alpha value of track, default 0.8.</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional;</span></dt><dd><p>Color of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_coverage</span></code>(coverage[, pos])</p></td>
<td><p>Append coverage to this track.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pile_coverages</span></code>(coverages[, pos])</p></td>
<td><p>Pile a stack of coverages with self’s coverages</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code>(ax, gr, **kwargs)</p></td>
<td><p>Plot arc connecting two positions on a linear scale representing interactions between bins.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_arcs</span></code>(ax, gr, gr2, intervals)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_coverages</span></code>(ax, gr, gr2)</p></td>
<td><p>Plot all coverages on given axes.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_hicpeaks</span></code>(ax, gr, gr2, intervals)</p></td>
<td><p>plot hic peaks uppon a HicMatBase track.</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 66%" />
<col style="width: 34%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>fetch_intervals</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>has_prop</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_contacts</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_label</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.coverage.HiCPeaksCoverage">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.coverage.</span></span><span class="sig-name descname"><span class="pre">HiCPeaksCoverage</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.coverage.HiCPeaksCoverage" title="Permalink to this definition">¶</a></dt>
<dd><p>Hi-C Peaks(Loops) from .bedpe file. Used to show the peaks on the Hi-C interaction map.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>file: str</strong></dt><dd><p>Path of .bedpe file</p>
</dd>
<dt><strong>pos</strong><span class="classifier">str, optional</span></dt><dd><p>Method for choosing arch anchor for bedpe data: ‘start’, ‘end’, ‘mid’, default ‘mid’</p>
</dd>
<dt><strong>style</strong><span class="classifier">str, optional</span></dt><dd><p>Style of arcs plot: ‘arcs’, ‘hicpeaks’, default ‘arcs’</p>
</dd>
<dt><strong>score_to_width</strong><span class="classifier">str, optional</span></dt><dd><p>Mapping function of score to width, default: ‘0.5 + math.sqrt(score)’</p>
</dd>
<dt><strong>line_width</strong><span class="classifier">float, optional</span></dt><dd><p>Width of arc line.</p>
</dd>
<dt><strong>open_region</strong><span class="classifier">bool, optional</span></dt><dd><p>If specified to True, will fetch the contacts on side in the region,
default True</p>
</dd>
<dt><strong>diameter_to_height</strong><span class="classifier">str, optional</span></dt><dd><p>Mapping function of arc diameter(interval end - start) to height.
You can specify to ‘max_diameter’ let all arcs has same height.
default ‘max_height * diameter / max_diameter’.</p>
</dd>
<dt><strong>orientation</strong><span class="classifier">str, optional</span></dt><dd><p>Track orientation, use ‘inverted’ for inverted track plot. default None</p>
</dd>
<dt><strong>line_style</strong><span class="classifier">str, optional</span></dt><dd><p>Border line style, default ‘solid’</p>
</dd>
<dt><strong>fill</strong><span class="classifier">bool, optional</span></dt><dd><p>Fill center or not, default False.</p>
</dd>
<dt><strong>fill_color</strong><span class="classifier">str, optional</span></dt><dd><p>Fill color, use ‘bed_rgb’ for specify color in file,
default ‘bed_rgb’.</p>
</dd>
<dt><strong>fill_alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Alpha value of fill region.
default 0.2</p>
</dd>
<dt><strong>side</strong><span class="classifier">{‘upper’, ‘lower’, ‘both’}</span></dt><dd><p>Plot peak in which side of the matrix.
NOTE: This parameters is useful only if the Cool track in matrix format.</p>
</dd>
<dt><strong>alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Alpha value of track, default 0.8.</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional;</span></dt><dd><p>Color of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 67%" />
<col style="width: 33%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>check_track_type</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.coverage.HighLights">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.coverage.</span></span><span class="sig-name descname"><span class="pre">HighLights</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.coverage.HighLights" title="Permalink to this definition">¶</a></dt>
<dd><p>High light region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>highlight_regions</strong><span class="classifier">list of {str, tuple}</span></dt><dd><p>A list of regions for highlights, region can be expressed as a tuple or string.
region tuple like:
[(‘chr1’, 100000, 120000), (‘chr2’, 130000, 150000)]
region string format: <cite>chr:start-end</cite> like:
[‘chr1:100000-120000’, ‘chr2:130000-150000’].</p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional</span></dt><dd><p>High light region color, default HighLights.DEFAULT_COLOR.</p>
</dd>
<dt><strong>alpha</strong><span class="classifier">float, optional</span></dt><dd><p>High light region alpha value, default 0.5</p>
</dd>
<dt><strong>border_line</strong><span class="classifier">bool, optional</span></dt><dd><p>Plot border line or not, default True.</p>
</dd>
<dt><strong>border_line_style</strong><span class="classifier">str, optional</span></dt><dd><p>Border line style, default ‘dashed’</p>
</dd>
<dt><strong>border_line_width</strong><span class="classifier">float, optional</span></dt><dd><p>Border line width, default 1.0</p>
</dd>
<dt><strong>border_line_color</strong><span class="classifier">str, optional</span></dt><dd><p>Border line color, default ‘#000000’</p>
</dd>
<dt><strong>border_line_alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Border line alpha value, default 0.8</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of thr Coverage.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>name</strong></dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 67%" />
<col style="width: 33%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>check_track_type</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>fetch_data</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.coverage.HighLightsFromFile">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.coverage.</span></span><span class="sig-name descname"><span class="pre">HighLightsFromFile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.coverage.HighLightsFromFile" title="Permalink to this definition">¶</a></dt>
<dd><p>High light regions coverage, read the regions from the file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>file_</strong><span class="classifier">str</span></dt><dd><p>Path to the file.</p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional</span></dt><dd><p>High light region color,
use ‘bed_rgb’ for specify color from the file, default ‘bed_rgb’.</p>
</dd>
<dt><strong>alpha</strong><span class="classifier">float, optional</span></dt><dd><p>High light region alpha value, default 0.1.</p>
</dd>
<dt><strong>border_line</strong><span class="classifier">bool, optional</span></dt><dd><p>Plot border line or not, default True.</p>
</dd>
<dt><strong>border_line_style</strong><span class="classifier">str, optional</span></dt><dd><p>Border line style, default ‘dashed’.</p>
</dd>
<dt><strong>border_line_width</strong><span class="classifier">float, optional</span></dt><dd><p>Border line width, default 1.0.</p>
</dd>
<dt><strong>border_line_color</strong><span class="classifier">str, optional</span></dt><dd><p>Border line color, default ‘#000000’</p>
</dd>
<dt><strong>border_line_alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Border line alpha value, default 0.8</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of thr Coverage.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>name</strong></dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 67%" />
<col style="width: 33%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>check_track_type</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>fetch_data</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="coolbox.core.coverage.Hist">
<span class="sig-prename descclassname"><span class="pre">coolbox.core.coverage.</span></span><span class="sig-name descname"><span class="pre">Hist</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> &#x2192; <a class="reference internal" href="#coolbox.core.track.HistBase" title="coolbox.core.track.hist.base.HistBase"><span class="pre">coolbox.core.track.hist.base.HistBase</span></a><a class="headerlink" href="#coolbox.core.coverage.Hist" title="Permalink to this definition">¶</a></dt>
<dd><p>Compose BigWig, BedGraph, SNP, BAMCov track automatically based on type of file
extension.(.bw, .bigwig, .snp, .vcf, .bedgraph, .bam, .sam)</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.coverage.HistCoverage">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.coverage.</span></span><span class="sig-name descname"><span class="pre">HistCoverage</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.coverage.HistCoverage" title="Permalink to this definition">¶</a></dt>
<dd><p>Compose BigWig, BedGraph, SNP, BAMCov track automatically based on type of file
extension.(.bw, .bigwig, .snp, .vcf, .bedgraph, .bam, .sam)</p>
<dl class="field-list simple">
<dt class="field-odd">Attributes</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>name</strong></dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 67%" />
<col style="width: 33%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>check_track_type</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.coverage.Pairs">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.coverage.</span></span><span class="sig-name descname"><span class="pre">Pairs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.coverage.Pairs" title="Permalink to this definition">¶</a></dt>
<dd><p>Arcs track from .pairs file.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>file: str</strong></dt><dd><p>Path of .pairs file</p>
</dd>
<dt><strong>style</strong><span class="classifier">str, optional</span></dt><dd><p>Style of arcs plot: ‘arcs’, ‘hicpeaks’, default ‘arcs’</p>
</dd>
<dt><strong>score_to_width</strong><span class="classifier">str, optional</span></dt><dd><p>Mapping function of score to width, default: ‘0.5 + math.sqrt(score)’</p>
</dd>
<dt><strong>line_width</strong><span class="classifier">float, optional</span></dt><dd><p>Width of arc line.</p>
</dd>
<dt><strong>open_region</strong><span class="classifier">bool, optional</span></dt><dd><p>If specified to True, will fetch the contacts on side in the region,
default True</p>
</dd>
<dt><strong>diameter_to_height</strong><span class="classifier">str, optional</span></dt><dd><p>Mapping function of arc diameter(interval end - start) to height.
You can specify to ‘max_diameter’ let all arcs has same height.
default ‘max_height * diameter / max_diameter’.</p>
</dd>
<dt><strong>orientation</strong><span class="classifier">str, optional</span></dt><dd><p>Track orientation, use ‘inverted’ for inverted track plot. default None</p>
</dd>
<dt><strong>line_style</strong><span class="classifier">str, optional</span></dt><dd><p>Border line style, default ‘solid’</p>
</dd>
<dt><strong>fill</strong><span class="classifier">bool, optional</span></dt><dd><p>Fill center or not, default False.</p>
</dd>
<dt><strong>fill_color</strong><span class="classifier">str, optional</span></dt><dd><p>Fill color, use ‘bed_rgb’ for specify color in file,
default ‘bed_rgb’.</p>
</dd>
<dt><strong>fill_alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Alpha value of fill region.
default 0.2</p>
</dd>
<dt><strong>side</strong><span class="classifier">{‘upper’, ‘lower’, ‘both’}</span></dt><dd><p>Plot peak in which side of the matrix.
NOTE: This parameters is useful only if the Cool track in matrix format.</p>
</dd>
<dt><strong>alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Alpha value of track, default 0.8.</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional;</span></dt><dd><p>Color of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_coverage</span></code>(coverage[, pos])</p></td>
<td><p>Append coverage to this track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.coverage.Pairs.fetch_data" title="coolbox.core.coverage.Pairs.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pile_coverages</span></code>(coverages[, pos])</p></td>
<td><p>Pile a stack of coverages with self’s coverages</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code>(ax, gr, **kwargs)</p></td>
<td><p>Plot arc connecting two positions on a linear scale representing interactions between bins.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_arcs</span></code>(ax, gr, gr2, intervals)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_coverages</span></code>(ax, gr, gr2)</p></td>
<td><p>Plot all coverages on given axes.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_hicpeaks</span></code>(ax, gr, gr2, intervals)</p></td>
<td><p>plot hic peaks uppon a HicMatBase track.</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 66%" />
<col style="width: 34%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>fetch_intervals</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>has_prop</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_contacts</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_label</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.coverage.Pairs.fetch_data" title="coolbox.core.coverage.Pairs.fetch_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code></a>(gr, **kwargs)</p></td>
<td><p>Fetch the raw data for the given GenomeRange.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.coverage.Pairs.fetch_data">
<span class="sig-name descname"><span class="pre">fetch_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.coverage.Pairs.fetch_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetch the raw data for the given GenomeRange.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.coverage.PairsCoverage">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.coverage.</span></span><span class="sig-name descname"><span class="pre">PairsCoverage</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.coverage.PairsCoverage" title="Permalink to this definition">¶</a></dt>
<dd><p>Arcs track from .pairs file.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>file: str</strong></dt><dd><p>Path of .pairs file</p>
</dd>
<dt><strong>style</strong><span class="classifier">str, optional</span></dt><dd><p>Style of arcs plot: ‘arcs’, ‘hicpeaks’, default ‘arcs’</p>
</dd>
<dt><strong>score_to_width</strong><span class="classifier">str, optional</span></dt><dd><p>Mapping function of score to width, default: ‘0.5 + math.sqrt(score)’</p>
</dd>
<dt><strong>line_width</strong><span class="classifier">float, optional</span></dt><dd><p>Width of arc line.</p>
</dd>
<dt><strong>open_region</strong><span class="classifier">bool, optional</span></dt><dd><p>If specified to True, will fetch the contacts on side in the region,
default True</p>
</dd>
<dt><strong>diameter_to_height</strong><span class="classifier">str, optional</span></dt><dd><p>Mapping function of arc diameter(interval end - start) to height.
You can specify to ‘max_diameter’ let all arcs has same height.
default ‘max_height * diameter / max_diameter’.</p>
</dd>
<dt><strong>orientation</strong><span class="classifier">str, optional</span></dt><dd><p>Track orientation, use ‘inverted’ for inverted track plot. default None</p>
</dd>
<dt><strong>line_style</strong><span class="classifier">str, optional</span></dt><dd><p>Border line style, default ‘solid’</p>
</dd>
<dt><strong>fill</strong><span class="classifier">bool, optional</span></dt><dd><p>Fill center or not, default False.</p>
</dd>
<dt><strong>fill_color</strong><span class="classifier">str, optional</span></dt><dd><p>Fill color, use ‘bed_rgb’ for specify color in file,
default ‘bed_rgb’.</p>
</dd>
<dt><strong>fill_alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Alpha value of fill region.
default 0.2</p>
</dd>
<dt><strong>side</strong><span class="classifier">{‘upper’, ‘lower’, ‘both’}</span></dt><dd><p>Plot peak in which side of the matrix.
NOTE: This parameters is useful only if the Cool track in matrix format.</p>
</dd>
<dt><strong>alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Alpha value of track, default 0.8.</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional;</span></dt><dd><p>Color of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 67%" />
<col style="width: 33%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>check_track_type</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.coverage.TAD">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.coverage.</span></span><span class="sig-name descname"><span class="pre">TAD</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.coverage.TAD" title="Permalink to this definition">¶</a></dt>
<dd><p>Tad tack from bed file</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>border_style: str, optional</strong></dt><dd><p>Border style of tad. (Default: ‘solid’)</p>
</dd>
<dt><strong>border_width: int, optional</strong></dt><dd><p>Border width of tad. (Default: ‘2.0’)</p>
</dd>
<dt><strong>show_score</strong><span class="classifier">bool</span></dt><dd><p>Show bed score or not.
default False.</p>
</dd>
<dt><strong>score_font_size</strong><span class="classifier">{‘auto’, int}</span></dt><dd><p>Score text font size.
default ‘auto’</p>
</dd>
<dt><strong>score_font_color</strong><span class="classifier">str</span></dt><dd><p>Score text color.
default ‘#000000’</p>
</dd>
<dt><strong>score_height_ratio</strong><span class="classifier">float</span></dt><dd><p>(text tag height) / (TAD height). used for adjust the position of Score text.
default 0.5</p>
</dd>
<dt><strong>border_only</strong><span class="classifier">bool</span></dt><dd><p>Only show border, default False</p>
</dd>
<dt><strong>depth_ratio</strong><span class="classifier">{float, ‘auto’, ‘full’}</span></dt><dd><p>Depth ratio of triangular, use ‘full’ for full depth, use ‘auto’ for calculate depth by current genome_range. default ‘auto’.</p>
</dd>
<dt><strong>orientation</strong><span class="classifier">{‘normal’, ‘inverted’}</span></dt><dd><p>Invert y-axis or not, default ‘normal’</p>
</dd>
<dt><strong>file: str</strong></dt><dd><p>The file path of <cite>.bed</cite> file.</p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional</span></dt><dd><p>Track color, ‘bed_rgb’ for auto specify color according to bed record.
(Default: ‘bed_rgb’)</p>
</dd>
<dt><strong>border_color</strong><span class="classifier">str, optional</span></dt><dd><p>Border_color of gene. (Default: ‘black’)</p>
</dd>
<dt><strong>max_value</strong><span class="classifier">float, optional</span></dt><dd><p>Max score. (Default: inf)</p>
</dd>
<dt><strong>min_value</strong><span class="classifier">float, optional</span></dt><dd><p>Min score. (Default: -inf)</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_coverage</span></code>(coverage[, pos])</p></td>
<td><p>Append coverage to this track.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_data</span></code>(gr, **kwargs)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_intervals</span></code>(bgz_file, gr)</p></td>
<td><p>Fetch intervals within input chromosome range.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">fetch_plot_data</span></code>(gr, **kwargs)</p></td>
<td><p>Fetch the plot data for the given GenomeRange.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">intervals2dataframe</span></code>(intervals, bed_type)</p></td>
<td><p>Convert intervals list to pandas.DataFrame</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pile_coverages</span></code>(coverages[, pos])</p></td>
<td><p>Pile a stack of coverages with self’s coverages</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.coverage.TAD.plot" title="coolbox.core.coverage.TAD.plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code></a>(ax, gr, **kwargs)</p></td>
<td><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_coverages</span></code>(ax, gr, gr2)</p></td>
<td><p>Plot all coverages on given axes.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_tads</span></code>(ax, gr, tads)</p></td>
<td><p>Plots the boundaries as triangles in the given ax.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_triangular</span></code>(ax, gr, region, ymax, depth)</p></td>
<td><p>/             /              /            _____________________</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_colormap</span></code>(df)</p></td>
<td><p>As min_score and max_score change every plot, we compute them for every plot</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 72%" />
<col style="width: 28%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>get_depth_ratio</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>get_rgb_and_edge_color</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>get_track_height</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>has_prop</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>infer_bed_type</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>init_colormap</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>load_range</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_box</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot_label</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot_score</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.coverage.TAD.plot" title="coolbox.core.coverage.TAD.plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code></a>(ax, gr, **kwargs)</p></td>
<td><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.coverage.TAD.plot">
<span class="sig-name descname"><span class="pre">plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gr</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.genome.GenomeRange"><span class="pre">coolbox.utilities.genome.GenomeRange</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.coverage.TAD.plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot figure based on data returned by <cite>fetch_plot_data</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gr: GenomeRange</strong></dt><dd></dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Other optional parameters including <cite>gr2</cite>, <cite>resolution</cite> for some tracks.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.coverage.TADCoverage">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.coverage.</span></span><span class="sig-name descname"><span class="pre">TADCoverage</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.coverage.TADCoverage" title="Permalink to this definition">¶</a></dt>
<dd><p>Tad tack from bed file</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>border_style: str, optional</strong></dt><dd><p>Border style of tad. (Default: ‘solid’)</p>
</dd>
<dt><strong>border_width: int, optional</strong></dt><dd><p>Border width of tad. (Default: ‘2.0’)</p>
</dd>
<dt><strong>show_score</strong><span class="classifier">bool</span></dt><dd><p>Show bed score or not.
default False.</p>
</dd>
<dt><strong>score_font_size</strong><span class="classifier">{‘auto’, int}</span></dt><dd><p>Score text font size.
default ‘auto’</p>
</dd>
<dt><strong>score_font_color</strong><span class="classifier">str</span></dt><dd><p>Score text color.
default ‘#000000’</p>
</dd>
<dt><strong>score_height_ratio</strong><span class="classifier">float</span></dt><dd><p>(text tag height) / (TAD height). used for adjust the position of Score text.
default 0.5</p>
</dd>
<dt><strong>border_only</strong><span class="classifier">bool</span></dt><dd><p>Only show border, default False</p>
</dd>
<dt><strong>depth_ratio</strong><span class="classifier">{float, ‘auto’, ‘full’}</span></dt><dd><p>Depth ratio of triangular, use ‘full’ for full depth, use ‘auto’ for calculate depth by current genome_range. default ‘auto’.</p>
</dd>
<dt><strong>orientation</strong><span class="classifier">{‘normal’, ‘inverted’}</span></dt><dd><p>Invert y-axis or not, default ‘normal’</p>
</dd>
<dt><strong>file: str</strong></dt><dd><p>The file path of <cite>.bed</cite> file.</p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional</span></dt><dd><p>Track color, ‘bed_rgb’ for auto specify color according to bed record.
(Default: ‘bed_rgb’)</p>
</dd>
<dt><strong>border_color</strong><span class="classifier">str, optional</span></dt><dd><p>Border_color of gene. (Default: ‘black’)</p>
</dd>
<dt><strong>max_value</strong><span class="classifier">float, optional</span></dt><dd><p>Max score. (Default: inf)</p>
</dd>
<dt><strong>min_value</strong><span class="classifier">float, optional</span></dt><dd><p>Min score. (Default: -inf)</p>
</dd>
<dt><strong>properties_dict</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of Track.
(Default: “{self.__class__.__name__}.{self.__class__._counts}”)</p>
</dd>
<dt><strong>title</strong><span class="classifier">str, optional</span></dt><dd><p>Title of ax</p>
</dd>
<dt><strong>height</strong><span class="classifier">int, optional</span></dt><dd><p>Height of ax</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>properties</strong><span class="classifier">dict</span></dt><dd><p>The properties(features) of this track. For example ‘height’, ‘color’…</p>
</dd>
<dt><strong>name</strong><span class="classifier">str</span></dt><dd><p>The name of Track.</p>
</dd>
<dt><strong>coverages</strong><span class="classifier">list of <cite>coolbox.api.coverage.Coverage</cite></span></dt><dd><p>Coverages on this Track.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 67%" />
<col style="width: 33%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>check_track_type</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>plot</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.coverage.Vlines">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.coverage.</span></span><span class="sig-name descname"><span class="pre">Vlines</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.coverage.Vlines" title="Permalink to this definition">¶</a></dt>
<dd><p>Vertical lines.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>vlines</strong><span class="classifier">list of {int, str}</span></dt><dd><p>A list of vline positions. position can be expressed as a tuple like:
[(‘chr1’, 10000), (‘chr2’, 50000)]
or a genome range string like:
[‘chr1:10000-10000’, ‘chr2:50000-50000’]</p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional</span></dt><dd><p>Line color, default ‘#1e1e1e’.</p>
</dd>
<dt><strong>alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Line alpha value, default 0.8.</p>
</dd>
<dt><strong>line_style</strong><span class="classifier">str, optional</span></dt><dd><p>Line style, default ‘dashed’.</p>
</dd>
<dt><strong>line_width</strong><span class="classifier">float, optional</span></dt><dd><p>Line width, default 0.5.</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of thr Coverage.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>name</strong></dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 67%" />
<col style="width: 33%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>check_track_type</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>fetch_data</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.coverage.VlinesFromFile">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.coverage.</span></span><span class="sig-name descname"><span class="pre">VlinesFromFile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.coverage.VlinesFromFile" title="Permalink to this definition">¶</a></dt>
<dd><p>Vertical lines from the file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>file_</strong><span class="classifier">str</span></dt><dd><p>Path to file.</p>
</dd>
<dt><strong>color</strong><span class="classifier">str, optional</span></dt><dd><p>Line color, default ‘#1e1e1e’.</p>
</dd>
<dt><strong>alpha</strong><span class="classifier">float, optional</span></dt><dd><p>Line alpha value, default 0.8.</p>
</dd>
<dt><strong>line_style</strong><span class="classifier">str, optional</span></dt><dd><p>Line style, default ‘dashed’.</p>
</dd>
<dt><strong>line_width</strong><span class="classifier">float, optional</span></dt><dd><p>Line width, default 0.5.</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>The name of thr Coverage.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>name</strong></dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 67%" />
<col style="width: 33%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>check_track_type</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>fetch_data</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>plot</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="module-coolbox.core.feature">
<span id="feature"></span><h2>Feature<a class="headerlink" href="#module-coolbox.core.feature" title="Permalink to this headline">¶</a></h2>
<p><strong>Classes:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.feature.Color" title="coolbox.core.feature.Color"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Color</span></code></a>(value)</p></td>
<td><p>Track color.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.feature.ColorMap" title="coolbox.core.feature.ColorMap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ColorMap</span></code></a>(value)</p></td>
<td><p>Track color map.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.feature.CoolStyle" title="coolbox.core.feature.CoolStyle"><code class="xref py py-obj docutils literal notranslate"><span class="pre">CoolStyle</span></code></a>([style])</p></td>
<td><p>Control Cool track’s style.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.feature.DepthRatio" title="coolbox.core.feature.DepthRatio"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DepthRatio</span></code></a>(depth_ratio)</p></td>
<td><p>Control Cool track’s depth ratio.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.feature.Feature" title="coolbox.core.feature.Feature"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Feature</span></code></a>(**kwargs)</p></td>
<td><p>Feature base class.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.feature.FrameFeature" title="coolbox.core.feature.FrameFeature"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FrameFeature</span></code></a>(**kwargs)</p></td>
<td><p>FrameFeature base class.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.feature.FrameTitle" title="coolbox.core.feature.FrameTitle"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FrameTitle</span></code></a>(value)</p></td>
<td><p>Frame title.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.feature.HistStyle" title="coolbox.core.feature.HistStyle"><code class="xref py py-obj docutils literal notranslate"><span class="pre">HistStyle</span></code></a>([style, fmt, size, line_width])</p></td>
<td><p>Style of BigWig or BedGraph.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.feature.Inverted" title="coolbox.core.feature.Inverted"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Inverted</span></code></a>()</p></td>
<td><p>Invert the orientation of track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.feature.MaxValue" title="coolbox.core.feature.MaxValue"><code class="xref py py-obj docutils literal notranslate"><span class="pre">MaxValue</span></code></a>(value)</p></td>
<td><p>Max value of track.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.feature.MinValue" title="coolbox.core.feature.MinValue"><code class="xref py py-obj docutils literal notranslate"><span class="pre">MinValue</span></code></a>(value)</p></td>
<td><p>Min value of track.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.feature.ShowColorBar" title="coolbox.core.feature.ShowColorBar"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ShowColorBar</span></code></a>([show])</p></td>
<td><p>Show color bar or not.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.feature.ShowDataRange" title="coolbox.core.feature.ShowDataRange"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ShowDataRange</span></code></a>([data_range_style])</p></td>
<td><p>Show data range or not.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.feature.Title" title="coolbox.core.feature.Title"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Title</span></code></a>(value)</p></td>
<td><p>Track title.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.feature.TrackHeight" title="coolbox.core.feature.TrackHeight"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TrackHeight</span></code></a>(value)</p></td>
<td><p>Track height.</p></td>
</tr>
</tbody>
</table>
<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.feature.Color">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.feature.</span></span><span class="sig-name descname"><span class="pre">Color</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">value</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.feature.Color" title="Permalink to this definition">¶</a></dt>
<dd><p>Track color.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.feature.ColorMap">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.feature.</span></span><span class="sig-name descname"><span class="pre">ColorMap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">value</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.feature.ColorMap" title="Permalink to this definition">¶</a></dt>
<dd><p>Track color map.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.feature.CoolStyle">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.feature.</span></span><span class="sig-name descname"><span class="pre">CoolStyle</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">style</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'triangular'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.feature.CoolStyle" title="Permalink to this definition">¶</a></dt>
<dd><p>Control Cool track’s style.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.feature.DepthRatio">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.feature.</span></span><span class="sig-name descname"><span class="pre">DepthRatio</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">depth_ratio</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.feature.DepthRatio" title="Permalink to this definition">¶</a></dt>
<dd><p>Control Cool track’s depth ratio.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.feature.Feature">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.feature.</span></span><span class="sig-name descname"><span class="pre">Feature</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.feature.Feature" title="Permalink to this definition">¶</a></dt>
<dd><p>Feature base class.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.feature.FrameFeature">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.feature.</span></span><span class="sig-name descname"><span class="pre">FrameFeature</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.feature.FrameFeature" title="Permalink to this definition">¶</a></dt>
<dd><p>FrameFeature base class.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.feature.FrameTitle">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.feature.</span></span><span class="sig-name descname"><span class="pre">FrameTitle</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">value</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.feature.FrameTitle" title="Permalink to this definition">¶</a></dt>
<dd><p>Frame title.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.feature.HistStyle">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.feature.</span></span><span class="sig-name descname"><span class="pre">HistStyle</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">style</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'fill'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fmt</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'-'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">line_width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2.0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.feature.HistStyle" title="Permalink to this definition">¶</a></dt>
<dd><p>Style of BigWig or BedGraph.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.feature.Inverted">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.feature.</span></span><span class="sig-name descname"><span class="pre">Inverted</span></span><a class="headerlink" href="#coolbox.core.feature.Inverted" title="Permalink to this definition">¶</a></dt>
<dd><p>Invert the orientation of track.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.feature.MaxValue">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.feature.</span></span><span class="sig-name descname"><span class="pre">MaxValue</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">value</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.feature.MaxValue" title="Permalink to this definition">¶</a></dt>
<dd><p>Max value of track.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.feature.MinValue">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.feature.</span></span><span class="sig-name descname"><span class="pre">MinValue</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">value</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.feature.MinValue" title="Permalink to this definition">¶</a></dt>
<dd><p>Min value of track.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.feature.ShowColorBar">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.feature.</span></span><span class="sig-name descname"><span class="pre">ShowColorBar</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">show</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.feature.ShowColorBar" title="Permalink to this definition">¶</a></dt>
<dd><p>Show color bar or not.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.feature.ShowDataRange">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.feature.</span></span><span class="sig-name descname"><span class="pre">ShowDataRange</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_range_style</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'y-axis'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.feature.ShowDataRange" title="Permalink to this definition">¶</a></dt>
<dd><p>Show data range or not.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.feature.Title">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.feature.</span></span><span class="sig-name descname"><span class="pre">Title</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">value</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.feature.Title" title="Permalink to this definition">¶</a></dt>
<dd><p>Track title.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.feature.TrackHeight">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.feature.</span></span><span class="sig-name descname"><span class="pre">TrackHeight</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">value</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.feature.TrackHeight" title="Permalink to this definition">¶</a></dt>
<dd><p>Track height.</p>
</dd></dl>

</div>
<div class="section" id="module-coolbox.core.frame">
<span id="frame"></span><h2>Frame<a class="headerlink" href="#module-coolbox.core.frame" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="jointview">
<h2>JointView<a class="headerlink" href="#jointview" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-coolbox.core.browser">
<span id="browser"></span><h2>Browser<a class="headerlink" href="#module-coolbox.core.browser" title="Permalink to this headline">¶</a></h2>
<p><strong>Classes:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.browser.Browser" title="coolbox.core.browser.Browser"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Browser</span></code></a>(frame[, reference_genome, …])</p></td>
<td><p>Browser base class.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.browser.WidgetsPanel" title="coolbox.core.browser.WidgetsPanel"><code class="xref py py-obj docutils literal notranslate"><span class="pre">WidgetsPanel</span></code></a>([type, reference_genome])</p></td>
<td><p>Widgets container.</p></td>
</tr>
</tbody>
</table>
<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.browser.Browser">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.browser.</span></span><span class="sig-name descname"><span class="pre">Browser</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">frame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_genome</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'hg19'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">init_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">widgets_box</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'simple'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">img_format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'svg'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.browser.Browser" title="Permalink to this definition">¶</a></dt>
<dd><p>Browser base class.</p>
<dl class="field-list simple">
<dt class="field-odd">Attributes</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>center</strong></dt><dd></dd>
<dt><strong>tracks</strong></dt><dd></dd>
<dt><strong>window_size</strong></dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.browser.Browser.clear_fig_cache" title="coolbox.core.browser.Browser.clear_fig_cache"><code class="xref py py-obj docutils literal notranslate"><span class="pre">clear_fig_cache</span></code></a>()</p></td>
<td><p>Clear the fig cache.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.browser.Browser.get_init_range" title="coolbox.core.browser.Browser.get_init_range"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_init_range</span></code></a>([chrom])</p></td>
<td><p>Generate an initial range within a chromosome.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.browser.Browser.preload_imgs" title="coolbox.core.browser.Browser.preload_imgs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">preload_imgs</span></code></a>(directions)</p></td>
<td><p>Preloading images to self.fig_cache.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.browser.Browser.refresh" title="coolbox.core.browser.Browser.refresh"><code class="xref py py-obj docutils literal notranslate"><span class="pre">refresh</span></code></a>([hard])</p></td>
<td><p>Refresh the image display.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.browser.Browser.save" title="coolbox.core.browser.Browser.save"><code class="xref py py-obj docutils literal notranslate"><span class="pre">save</span></code></a>(path[, dpi])</p></td>
<td><p>Save current frame’s image to file.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.browser.Browser.show" title="coolbox.core.browser.Browser.show"><code class="xref py py-obj docutils literal notranslate"><span class="pre">show</span></code></a>(*args, **kwargs)</p></td>
<td><p>Show widgets and frame.</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 58%" />
<col style="width: 42%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>fetch_data</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>go_left</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>go_right</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>goto</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>zoom_in</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>zoom_out</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.browser.Browser.clear_fig_cache" title="coolbox.core.browser.Browser.clear_fig_cache"><code class="xref py py-obj docutils literal notranslate"><span class="pre">clear_fig_cache</span></code></a>()</p></td>
<td><p>Clear the fig cache.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.browser.Browser.get_init_range" title="coolbox.core.browser.Browser.get_init_range"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_init_range</span></code></a>([chrom])</p></td>
<td><p>Generate an initial range within a chromosome.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.browser.Browser.preload_imgs" title="coolbox.core.browser.Browser.preload_imgs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">preload_imgs</span></code></a>(directions)</p></td>
<td><p>Preloading images to self.fig_cache.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.browser.Browser.refresh" title="coolbox.core.browser.Browser.refresh"><code class="xref py py-obj docutils literal notranslate"><span class="pre">refresh</span></code></a>([hard])</p></td>
<td><p>Refresh the image display.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.core.browser.Browser.save" title="coolbox.core.browser.Browser.save"><code class="xref py py-obj docutils literal notranslate"><span class="pre">save</span></code></a>(path[, dpi])</p></td>
<td><p>Save current frame’s image to file.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.core.browser.Browser.show" title="coolbox.core.browser.Browser.show"><code class="xref py py-obj docutils literal notranslate"><span class="pre">show</span></code></a>(*args, **kwargs)</p></td>
<td><p>Show widgets and frame.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.browser.Browser.clear_fig_cache">
<span class="sig-name descname"><span class="pre">clear_fig_cache</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.browser.Browser.clear_fig_cache" title="Permalink to this definition">¶</a></dt>
<dd><p>Clear the fig cache.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.browser.Browser.get_init_range">
<span class="sig-name descname"><span class="pre">get_init_range</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chrom</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.browser.Browser.get_init_range" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate an initial range within a chromosome.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>chrom (str, optional): initial choromosome.</p>
</dd>
<dt>Return:</dt><dd><p>(<code class="xref py py-obj docutils literal notranslate"><span class="pre">GenomeRange</span></code>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.browser.Browser.preload_imgs">
<span class="sig-name descname"><span class="pre">preload_imgs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">directions</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.browser.Browser.preload_imgs" title="Permalink to this definition">¶</a></dt>
<dd><p>Preloading images to self.fig_cache.</p>
<dl class="simple">
<dt>Can load image in one of 4 directions:</dt><dd><p>left, right, zoom-in, zoom-out</p>
</dd>
</dl>
<p>or load all directions.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.browser.Browser.refresh">
<span class="sig-name descname"><span class="pre">refresh</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">hard</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.browser.Browser.refresh" title="Permalink to this definition">¶</a></dt>
<dd><p>Refresh the image display.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.browser.Browser.save">
<span class="sig-name descname"><span class="pre">save</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.browser.Browser.save" title="Permalink to this definition">¶</a></dt>
<dd><p>Save current frame’s image to file.
The suffix of the path will changing the saving format.
For example: <cite>bsr.save(“fig.png”)</cite> will save file to ‘png’ format.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.core.browser.Browser.show">
<span class="sig-name descname"><span class="pre">show</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.browser.Browser.show" title="Permalink to this definition">¶</a></dt>
<dd><p>Show widgets and frame.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.core.browser.WidgetsPanel">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.core.browser.</span></span><span class="sig-name descname"><span class="pre">WidgetsPanel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'simple'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_genome</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'hg19'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.core.browser.WidgetsPanel" title="Permalink to this definition">¶</a></dt>
<dd><p>Widgets container.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>type</strong><span class="classifier">{‘simple’, ‘full’}, optional</span></dt><dd><p>Widgets type ‘simple’ or ‘full’,
default ‘simple’</p>
</dd>
<dt><strong>reference_genome</strong><span class="classifier">str, optional</span></dt><dd><p>Reference genome,
built-in references:(‘hg19’, ‘hg38’, ‘mm9’, ‘mm10’)
if you want use other genome, you can specify the “chromosome length file”,
that is a tab splited file, first column is the chromosomes,
and second column is the length of correspond chromosome.
default ‘hg19’</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-coolbox.utilities">
<span id="utilities"></span><h2>Utilities<a class="headerlink" href="#module-coolbox.utilities" title="Permalink to this headline">¶</a></h2>
<p><strong>Classes:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.GenomeLength" title="coolbox.utilities.GenomeLength"><code class="xref py py-obj docutils literal notranslate"><span class="pre">GenomeLength</span></code></a>(length_file[, genome_name])</p></td>
<td><p><p class="rubric">Methods</p>
</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.GenomeRange" title="coolbox.utilities.GenomeRange"><code class="xref py py-obj docutils literal notranslate"><span class="pre">GenomeRange</span></code></a>(*args)</p></td>
<td><p>Express a range on the genome.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.Interval" title="coolbox.utilities.Interval"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Interval</span></code></a>(begin, end[, data])</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Attributes</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree" title="coolbox.utilities.IntervalTree"><code class="xref py py-obj docutils literal notranslate"><span class="pre">IntervalTree</span></code></a>([intervals])</p></td>
<td><p>A binary lookup tree of intervals.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.ReadBed" title="coolbox.utilities.ReadBed"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ReadBed</span></code></a>(file_handle)</p></td>
<td><p>Reads a bed file.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.deque" title="coolbox.utilities.deque"><code class="xref py py-obj docutils literal notranslate"><span class="pre">deque</span></code></a></p></td>
<td><p>deque([iterable[, maxlen]]) –&gt; deque object</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.partial" title="coolbox.utilities.partial"><code class="xref py py-obj docutils literal notranslate"><span class="pre">partial</span></code></a></p></td>
<td><p>partial(func, <a href="#id1"><span class="problematic" id="id2">*</span></a>args, <a href="#id3"><span class="problematic" id="id4">**</span></a>keywords) - new function with partial application of the given arguments and keywords.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.refGeneRec" title="coolbox.utilities.refGeneRec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">refGeneRec</span></code></a>(bin, name, chrom, strand, …)</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Attributes</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
</tbody>
</table>
<p><strong>Functions:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.abspath" title="coolbox.utilities.abspath"><code class="xref py py-obj docutils literal notranslate"><span class="pre">abspath</span></code></a>(path)</p></td>
<td><p>Return an absolute path.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.change_chrom_names" title="coolbox.utilities.change_chrom_names"><code class="xref py py-obj docutils literal notranslate"><span class="pre">change_chrom_names</span></code></a>(chrom)</p></td>
<td><p>Changes UCSC chromosome names to ensembl chromosome names and vice versa.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.cm2inch" title="coolbox.utilities.cm2inch"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cm2inch</span></code></a>(*tupl)</p></td>
<td><p>convert length unit from cm to inch.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.correspond_track" title="coolbox.utilities.correspond_track"><code class="xref py py-obj docutils literal notranslate"><span class="pre">correspond_track</span></code></a>(track_or_name)</p></td>
<td><p>Search track by it’s name.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.dirname" title="coolbox.utilities.dirname"><code class="xref py py-obj docutils literal notranslate"><span class="pre">dirname</span></code></a>(p)</p></td>
<td><p>Returns the directory component of a pathname</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.fig2bytes" title="coolbox.utilities.fig2bytes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fig2bytes</span></code></a>(fig[, encode, dpi])</p></td>
<td><p>Convert matplotlib.figure.Figure/IPython.display.SVG object to image bytes.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.file_to_intervaltree" title="coolbox.utilities.file_to_intervaltree"><code class="xref py py-obj docutils literal notranslate"><span class="pre">file_to_intervaltree</span></code></a>(file_name)</p></td>
<td><p>converts a BED like file into a bx python interval tree</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.format_properties" title="coolbox.utilities.format_properties"><code class="xref py py-obj docutils literal notranslate"><span class="pre">format_properties</span></code></a>(properties)</p></td>
<td><p>Convert bool value to ‘yes’ or ‘no’, for compatible with pyGenomeTracks Convert string to float if possible</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.get_size" title="coolbox.utilities.get_size"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_size</span></code></a>(obj[, seen])</p></td>
<td><p>Recursively finds size of objects</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.get_tmp_dir" title="coolbox.utilities.get_tmp_dir"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_tmp_dir</span></code></a>([name])</p></td>
<td><p>Return the path to coolbox temporary directory.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.get_uniq_tmp_file" title="coolbox.utilities.get_uniq_tmp_file"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_uniq_tmp_file</span></code></a>([prefix, suffix, dirname])</p></td>
<td><p>Return the path to a unique(not exists) temporary file.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.hex2rgb" title="coolbox.utilities.hex2rgb"><code class="xref py py-obj docutils literal notranslate"><span class="pre">hex2rgb</span></code></a>(color_hex)</p></td>
<td><p>Convert hex color code to rgb tuple.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.inch2cm" title="coolbox.utilities.inch2cm"><code class="xref py py-obj docutils literal notranslate"><span class="pre">inch2cm</span></code></a>(*tupl)</p></td>
<td><p>convert length unit from inch to cm.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.join" title="coolbox.utilities.join"><code class="xref py py-obj docutils literal notranslate"><span class="pre">join</span></code></a>(a, *p)</p></td>
<td><p>Join two or more pathname components, inserting ‘/’ as needed.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.op_err_msg" title="coolbox.utilities.op_err_msg"><code class="xref py py-obj docutils literal notranslate"><span class="pre">op_err_msg</span></code></a>(a, b[, op])</p></td>
<td><p>Generate the error message of error operand type.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.opener" title="coolbox.utilities.opener"><code class="xref py py-obj docutils literal notranslate"><span class="pre">opener</span></code></a>(filename)</p></td>
<td><p>Determines if a file is compressed or not</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.query_bed" title="coolbox.utilities.query_bed"><code class="xref py py-obj docutils literal notranslate"><span class="pre">query_bed</span></code></a>(filename, chrom, start, end, *[, …])</p></td>
<td><p>Call tabix and generate an array of strings for each line it returns.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.rgb2hex" title="coolbox.utilities.rgb2hex"><code class="xref py py-obj docutils literal notranslate"><span class="pre">rgb2hex</span></code></a>(r, g, b)</p></td>
<td><p>Convert rgb color to hex format.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.shiftedColorMap" title="coolbox.utilities.shiftedColorMap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">shiftedColorMap</span></code></a>(cmap[, start, midpoint, …])</p></td>
<td><p>This function is from: <a class="reference external" href="https://stackoverflow.com/a/20528097/8500469">https://stackoverflow.com/a/20528097/8500469</a></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.split_genome_range" title="coolbox.utilities.split_genome_range"><code class="xref py py-obj docutils literal notranslate"><span class="pre">split_genome_range</span></code></a>(genome_range)</p></td>
<td><p>Little helper func.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.tabix_index" title="coolbox.utilities.tabix_index"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tabix_index</span></code></a>(filename[, preset])</p></td>
<td><p>Call tabix to create an index for a bgzip-compressed file.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.tabix_query" title="coolbox.utilities.tabix_query"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tabix_query</span></code></a>(filename, chrom, start, end[, split])</p></td>
<td><p>Call tabix and generate an array of strings for each line it returns.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.to_bytes" title="coolbox.utilities.to_bytes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_bytes</span></code></a>(s)</p></td>
<td><p>Like toString, to bytes.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.to_gr" title="coolbox.utilities.to_gr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_gr</span></code></a>(obj)</p></td>
<td><p>Convert object to GenomeRange.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.to_string" title="coolbox.utilities.to_string"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_string</span></code></a>(s)</p></td>
<td><p>Convert bytes, bytes list to string, string list.</p></td>
</tr>
</tbody>
</table>
<dl class="py class">
<dt class="sig sig-object py" id="coolbox.utilities.GenomeLength">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.utilities.</span></span><span class="sig-name descname"><span class="pre">GenomeLength</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">length_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genome_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.GenomeLength" title="Permalink to this definition">¶</a></dt>
<dd><p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.GenomeLength.bound_range" title="coolbox.utilities.GenomeLength.bound_range"><code class="xref py py-obj docutils literal notranslate"><span class="pre">bound_range</span></code></a>(genome_range)</p></td>
<td><p>Bound a genome range within reference.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.GenomeLength.check_range" title="coolbox.utilities.GenomeLength.check_range"><code class="xref py py-obj docutils literal notranslate"><span class="pre">check_range</span></code></a>(genome_range)</p></td>
<td><p>Check a genome range is valid or not.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">clear</span></code>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">fromkeys</span></code>(iterable[, value])</p></td>
<td><p>Create a new dictionary with keys from iterable and values set to value.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get</span></code>(key[, default])</p></td>
<td><p>Return the value for key if key is in the dictionary, else default.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">items</span></code>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">keys</span></code>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pop</span></code>(key[, default])</p></td>
<td><p>If key is not found, default is returned if given, otherwise KeyError is raised</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">popitem</span></code>(/)</p></td>
<td><p>Remove and return a (key, value) pair as a 2-tuple.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">setdefault</span></code>(key[, default])</p></td>
<td><p>Insert key with a value of default if key is not in the dictionary.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">update</span></code>([E, ]**F)</p></td>
<td><p>If E is present and has a .keys() method, then does:  for k in E: D[k] = E[k] If E is present and lacks a .keys() method, then does:  for k, v in E: D[k] = v In either case, this is followed by: for k in F:  D[k] = F[k]</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">values</span></code>()</p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 58%" />
<col style="width: 42%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>parse_file</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.GenomeLength.bound_range" title="coolbox.utilities.GenomeLength.bound_range"><code class="xref py py-obj docutils literal notranslate"><span class="pre">bound_range</span></code></a>(genome_range)</p></td>
<td><p>Bound a genome range within reference.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.GenomeLength.check_range" title="coolbox.utilities.GenomeLength.check_range"><code class="xref py py-obj docutils literal notranslate"><span class="pre">check_range</span></code></a>(genome_range)</p></td>
<td><p>Check a genome range is valid or not.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.GenomeLength.bound_range">
<span class="sig-name descname"><span class="pre">bound_range</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome_range</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.GenomeLength.bound_range" title="Permalink to this definition">¶</a></dt>
<dd><p>Bound a genome range within reference.</p>
<p>(both start and end position is one based.)</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.GenomeLength.check_range">
<span class="sig-name descname"><span class="pre">check_range</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome_range</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.GenomeLength.check_range" title="Permalink to this definition">¶</a></dt>
<dd><p>Check a genome range is valid or not.</p>
<p>(both start and end position is one based.)</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.utilities.GenomeRange">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.utilities.</span></span><span class="sig-name descname"><span class="pre">GenomeRange</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.GenomeRange" title="Permalink to this definition">¶</a></dt>
<dd><p>Express a range on the genome.</p>
<dl class="field-list simple">
<dt class="field-odd">Attributes</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>chrom</strong><span class="classifier">str</span></dt><dd><p>chromosome</p>
</dd>
<dt><strong>start</strong><span class="classifier">int</span></dt><dd><p>start position</p>
</dd>
<dt><strong>end</strong><span class="classifier">int</span></dt><dd><p>end position</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.GenomeRange.change_chrom_names" title="coolbox.utilities.GenomeRange.change_chrom_names"><code class="xref py py-obj docutils literal notranslate"><span class="pre">change_chrom_names</span></code></a>()</p></td>
<td><p><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">range1</span> <span class="o">=</span> <span class="n">GenomeRange</span><span class="p">(</span><span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="mi">1000</span><span class="p">,</span> <span class="mi">2000</span><span class="p">)</span>
</pre></div>
</div>
</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.GenomeRange.parse_region_string" title="coolbox.utilities.GenomeRange.parse_region_string"><code class="xref py py-obj docutils literal notranslate"><span class="pre">parse_region_string</span></code></a>(region_string)</p></td>
<td><p>splits a region string into a (chrom, start, end) tuple</p></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.GenomeRange.change_chrom_names" title="coolbox.utilities.GenomeRange.change_chrom_names"><code class="xref py py-obj docutils literal notranslate"><span class="pre">change_chrom_names</span></code></a>()</p></td>
<td><p><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">range1</span> <span class="o">=</span> <span class="n">GenomeRange</span><span class="p">(</span><span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="mi">1000</span><span class="p">,</span> <span class="mi">2000</span><span class="p">)</span>
</pre></div>
</div>
</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.GenomeRange.parse_region_string" title="coolbox.utilities.GenomeRange.parse_region_string"><code class="xref py py-obj docutils literal notranslate"><span class="pre">parse_region_string</span></code></a>(region_string)</p></td>
<td><p>splits a region string into a (chrom, start, end) tuple</p></td>
</tr>
</tbody>
</table>
<p><strong>Attributes:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.GenomeRange.length" title="coolbox.utilities.GenomeRange.length"><code class="xref py py-obj docutils literal notranslate"><span class="pre">length</span></code></a></p></td>
<td><p><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">range1</span> <span class="o">=</span> <span class="n">GenomeRange</span><span class="p">(</span><span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1000</span><span class="p">)</span>
</pre></div>
</div>
</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.GenomeRange.change_chrom_names">
<span class="sig-name descname"><span class="pre">change_chrom_names</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.GenomeRange.change_chrom_names" title="Permalink to this definition">¶</a></dt>
<dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">range1</span> <span class="o">=</span> <span class="n">GenomeRange</span><span class="p">(</span><span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="mi">1000</span><span class="p">,</span> <span class="mi">2000</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">range1</span><span class="o">.</span><span class="n">chrom</span>
<span class="go">&#39;chr1&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">range1</span><span class="o">.</span><span class="n">change_chrom_names</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">range1</span><span class="o">.</span><span class="n">chrom</span>
<span class="go">&#39;1&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">range1</span><span class="o">.</span><span class="n">change_chrom_names</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">range1</span><span class="o">.</span><span class="n">chrom</span>
<span class="go">&#39;chr1&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="coolbox.utilities.GenomeRange.length">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">length</span></span><a class="headerlink" href="#coolbox.utilities.GenomeRange.length" title="Permalink to this definition">¶</a></dt>
<dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">range1</span> <span class="o">=</span> <span class="n">GenomeRange</span><span class="p">(</span><span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1000</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">range1</span><span class="o">.</span><span class="n">length</span>
<span class="go">1000</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.GenomeRange.parse_region_string">
<em class="property"><span class="pre">static</span> </em><span class="sig-name descname"><span class="pre">parse_region_string</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region_string</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.GenomeRange.parse_region_string" title="Permalink to this definition">¶</a></dt>
<dd><p>splits a region string into
a (chrom, start, end) tuple</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>region_string</strong><span class="classifier">str</span></dt><dd><p>Region string to be parsed, like: “chr:start-end”</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.utilities.Interval">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.utilities.</span></span><span class="sig-name descname"><span class="pre">Interval</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">begin</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.Interval" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Attributes</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>begin</strong></dt><dd><p>Alias for field number 0</p>
</dd>
<dt><strong>data</strong></dt><dd><p>Alias for field number 2</p>
</dd>
<dt><strong>end</strong></dt><dd><p>Alias for field number 1</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.Interval.contains_interval" title="coolbox.utilities.Interval.contains_interval"><code class="xref py py-obj docutils literal notranslate"><span class="pre">contains_interval</span></code></a>(other)</p></td>
<td><p>Whether other is contained in this Interval.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.Interval.contains_point" title="coolbox.utilities.Interval.contains_point"><code class="xref py py-obj docutils literal notranslate"><span class="pre">contains_point</span></code></a>(p)</p></td>
<td><p>Whether the Interval contains p.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.Interval.copy" title="coolbox.utilities.Interval.copy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code></a>()</p></td>
<td><p>Shallow copy.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">count</span></code>(value, /)</p></td>
<td><p>Return number of occurrences of value.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.Interval.distance_to" title="coolbox.utilities.Interval.distance_to"><code class="xref py py-obj docutils literal notranslate"><span class="pre">distance_to</span></code></a>(other)</p></td>
<td><p>Returns the size of the gap between intervals, or 0  if they touch or overlap.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.Interval.ge" title="coolbox.utilities.Interval.ge"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ge</span></code></a>(other)</p></td>
<td><p>Greater than or overlaps.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.Interval.gt" title="coolbox.utilities.Interval.gt"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gt</span></code></a>(other)</p></td>
<td><p>Strictly greater than.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">index</span></code>(value[, start, stop])</p></td>
<td><p>Return first index of value.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.Interval.is_null" title="coolbox.utilities.Interval.is_null"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_null</span></code></a>()</p></td>
<td><p>Whether this equals the null interval.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.Interval.le" title="coolbox.utilities.Interval.le"><code class="xref py py-obj docutils literal notranslate"><span class="pre">le</span></code></a>(other)</p></td>
<td><p>Less than or overlaps.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.Interval.length" title="coolbox.utilities.Interval.length"><code class="xref py py-obj docutils literal notranslate"><span class="pre">length</span></code></a>()</p></td>
<td><p>The distance covered by this Interval.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.Interval.lt" title="coolbox.utilities.Interval.lt"><code class="xref py py-obj docutils literal notranslate"><span class="pre">lt</span></code></a>(other)</p></td>
<td><p>Strictly less than.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.Interval.overlap_size" title="coolbox.utilities.Interval.overlap_size"><code class="xref py py-obj docutils literal notranslate"><span class="pre">overlap_size</span></code></a>(begin[, end])</p></td>
<td><p>Return the overlap size between two intervals or a point :param begin: beginning point of the range, or the point, or an Interval :param end: end point of the range.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.Interval.overlaps" title="coolbox.utilities.Interval.overlaps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">overlaps</span></code></a>(begin[, end])</p></td>
<td><p>Whether the interval overlaps the given point, range or Interval.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.Interval.range_matches" title="coolbox.utilities.Interval.range_matches"><code class="xref py py-obj docutils literal notranslate"><span class="pre">range_matches</span></code></a>(other)</p></td>
<td><p>Whether the begins equal and the ends equal.</p></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.Interval.contains_interval" title="coolbox.utilities.Interval.contains_interval"><code class="xref py py-obj docutils literal notranslate"><span class="pre">contains_interval</span></code></a>(other)</p></td>
<td><p>Whether other is contained in this Interval.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.Interval.contains_point" title="coolbox.utilities.Interval.contains_point"><code class="xref py py-obj docutils literal notranslate"><span class="pre">contains_point</span></code></a>(p)</p></td>
<td><p>Whether the Interval contains p.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.Interval.copy" title="coolbox.utilities.Interval.copy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code></a>()</p></td>
<td><p>Shallow copy.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.Interval.distance_to" title="coolbox.utilities.Interval.distance_to"><code class="xref py py-obj docutils literal notranslate"><span class="pre">distance_to</span></code></a>(other)</p></td>
<td><p>Returns the size of the gap between intervals, or 0  if they touch or overlap.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.Interval.ge" title="coolbox.utilities.Interval.ge"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ge</span></code></a>(other)</p></td>
<td><p>Greater than or overlaps.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.Interval.gt" title="coolbox.utilities.Interval.gt"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gt</span></code></a>(other)</p></td>
<td><p>Strictly greater than.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.Interval.is_null" title="coolbox.utilities.Interval.is_null"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_null</span></code></a>()</p></td>
<td><p>Whether this equals the null interval.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.Interval.le" title="coolbox.utilities.Interval.le"><code class="xref py py-obj docutils literal notranslate"><span class="pre">le</span></code></a>(other)</p></td>
<td><p>Less than or overlaps.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.Interval.length" title="coolbox.utilities.Interval.length"><code class="xref py py-obj docutils literal notranslate"><span class="pre">length</span></code></a>()</p></td>
<td><p>The distance covered by this Interval.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.Interval.lt" title="coolbox.utilities.Interval.lt"><code class="xref py py-obj docutils literal notranslate"><span class="pre">lt</span></code></a>(other)</p></td>
<td><p>Strictly less than.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.Interval.overlap_size" title="coolbox.utilities.Interval.overlap_size"><code class="xref py py-obj docutils literal notranslate"><span class="pre">overlap_size</span></code></a>(begin[, end])</p></td>
<td><p>Return the overlap size between two intervals or a point :param begin: beginning point of the range, or the point, or an Interval :param end: end point of the range.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.Interval.overlaps" title="coolbox.utilities.Interval.overlaps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">overlaps</span></code></a>(begin[, end])</p></td>
<td><p>Whether the interval overlaps the given point, range or Interval.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.Interval.range_matches" title="coolbox.utilities.Interval.range_matches"><code class="xref py py-obj docutils literal notranslate"><span class="pre">range_matches</span></code></a>(other)</p></td>
<td><p>Whether the begins equal and the ends equal.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.Interval.contains_interval">
<span class="sig-name descname"><span class="pre">contains_interval</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.Interval.contains_interval" title="Permalink to this definition">¶</a></dt>
<dd><p>Whether other is contained in this Interval.
:param other: Interval
:return: True or False
:rtype: bool</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.Interval.contains_point">
<span class="sig-name descname"><span class="pre">contains_point</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">p</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.Interval.contains_point" title="Permalink to this definition">¶</a></dt>
<dd><p>Whether the Interval contains p.
:param p: a point
:return: True or False
:rtype: bool</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.Interval.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.Interval.copy" title="Permalink to this definition">¶</a></dt>
<dd><p>Shallow copy.
:return: copy of self
:rtype: Interval</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.Interval.distance_to">
<span class="sig-name descname"><span class="pre">distance_to</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.Interval.distance_to" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the size of the gap between intervals, or 0 
if they touch or overlap.
:param other: Interval or point
:return: distance
:rtype: Number</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.Interval.ge">
<span class="sig-name descname"><span class="pre">ge</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.Interval.ge" title="Permalink to this definition">¶</a></dt>
<dd><p>Greater than or overlaps. Returns True if no part of this Interval
extends lower than other.
:raises ValueError: if either self or other is a null Interval
:param other: Interval or point
:return: True or False
:rtype: bool</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.Interval.gt">
<span class="sig-name descname"><span class="pre">gt</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.Interval.gt" title="Permalink to this definition">¶</a></dt>
<dd><p>Strictly greater than. Returns True if no part of this Interval
extends lower than or into other.
:raises ValueError: if either self or other is a null Interval
:param other: Interval or point
:return: True or False
:rtype: bool</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.Interval.is_null">
<span class="sig-name descname"><span class="pre">is_null</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.Interval.is_null" title="Permalink to this definition">¶</a></dt>
<dd><p>Whether this equals the null interval.
:return: True if end &lt;= begin else False
:rtype: bool</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.Interval.le">
<span class="sig-name descname"><span class="pre">le</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.Interval.le" title="Permalink to this definition">¶</a></dt>
<dd><p>Less than or overlaps. Returns True if no part of this Interval
extends higher than other.
:raises ValueError: if either self or other is a null Interval
:param other: Interval or point
:return: True or False
:rtype: bool</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.Interval.length">
<span class="sig-name descname"><span class="pre">length</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.Interval.length" title="Permalink to this definition">¶</a></dt>
<dd><p>The distance covered by this Interval.
:return: length
:type: Number</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.Interval.lt">
<span class="sig-name descname"><span class="pre">lt</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.Interval.lt" title="Permalink to this definition">¶</a></dt>
<dd><p>Strictly less than. Returns True if no part of this Interval
extends higher than or into other.
:raises ValueError: if either self or other is a null Interval
:param other: Interval or point
:return: True or False
:rtype: bool</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.Interval.overlap_size">
<span class="sig-name descname"><span class="pre">overlap_size</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">begin</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.Interval.overlap_size" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the overlap size between two intervals or a point
:param begin: beginning point of the range, or the point, or an Interval
:param end: end point of the range. Optional if not testing ranges.
:return: Return the overlap size, None if not overlap is found
:rtype: depends on the given input (e.g., int will be returned for int interval and timedelta for
datetime intervals)</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.Interval.overlaps">
<span class="sig-name descname"><span class="pre">overlaps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">begin</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.Interval.overlaps" title="Permalink to this definition">¶</a></dt>
<dd><p>Whether the interval overlaps the given point, range or Interval.
:param begin: beginning point of the range, or the point, or an Interval
:param end: end point of the range. Optional if not testing ranges.
:return: True or False
:rtype: bool</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.Interval.range_matches">
<span class="sig-name descname"><span class="pre">range_matches</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.Interval.range_matches" title="Permalink to this definition">¶</a></dt>
<dd><p>Whether the begins equal and the ends equal. Compare __eq__().
:param other: Interval
:return: True or False
:rtype: bool</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.utilities.</span></span><span class="sig-name descname"><span class="pre">IntervalTree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">intervals</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree" title="Permalink to this definition">¶</a></dt>
<dd><p>A binary lookup tree of intervals.
The intervals contained in the tree are represented using <code class="docutils literal notranslate"><span class="pre">Interval(a,</span> <span class="pre">b,</span> <span class="pre">data)</span></code> objects.
Each such object represents a half-open interval <code class="docutils literal notranslate"><span class="pre">[a,</span> <span class="pre">b)</span></code> with optional data.</p>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.add" title="coolbox.utilities.IntervalTree.add"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add</span></code></a>(interval)</p></td>
<td><p>Adds an interval to the tree, if not already present.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.addi" title="coolbox.utilities.IntervalTree.addi"><code class="xref py py-obj docutils literal notranslate"><span class="pre">addi</span></code></a>(begin, end[, data])</p></td>
<td><p>Shortcut for add(Interval(begin, end, data)).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.append" title="coolbox.utilities.IntervalTree.append"><code class="xref py py-obj docutils literal notranslate"><span class="pre">append</span></code></a>(interval)</p></td>
<td><p>Adds an interval to the tree, if not already present.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.appendi" title="coolbox.utilities.IntervalTree.appendi"><code class="xref py py-obj docutils literal notranslate"><span class="pre">appendi</span></code></a>(begin, end[, data])</p></td>
<td><p>Shortcut for add(Interval(begin, end, data)).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.at" title="coolbox.utilities.IntervalTree.at"><code class="xref py py-obj docutils literal notranslate"><span class="pre">at</span></code></a>(p)</p></td>
<td><p>Returns the set of all intervals that contain p.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.begin" title="coolbox.utilities.IntervalTree.begin"><code class="xref py py-obj docutils literal notranslate"><span class="pre">begin</span></code></a>()</p></td>
<td><p>Returns the lower bound of the first interval in the tree.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.chop" title="coolbox.utilities.IntervalTree.chop"><code class="xref py py-obj docutils literal notranslate"><span class="pre">chop</span></code></a>(begin, end[, datafunc])</p></td>
<td><p>Like remove_envelop(), but trims back Intervals hanging into the chopped area so that nothing overlaps.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.clear" title="coolbox.utilities.IntervalTree.clear"><code class="xref py py-obj docutils literal notranslate"><span class="pre">clear</span></code></a>()</p></td>
<td><p>Empties the tree.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.containsi" title="coolbox.utilities.IntervalTree.containsi"><code class="xref py py-obj docutils literal notranslate"><span class="pre">containsi</span></code></a>(begin, end[, data])</p></td>
<td><p>Shortcut for (Interval(begin, end, data) in tree).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.copy" title="coolbox.utilities.IntervalTree.copy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code></a>()</p></td>
<td><p>Construct a new IntervalTree using shallow copies of the intervals in the source tree.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.difference" title="coolbox.utilities.IntervalTree.difference"><code class="xref py py-obj docutils literal notranslate"><span class="pre">difference</span></code></a>(other)</p></td>
<td><p>Returns a new tree, comprising all intervals in self but not in other.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.difference_update" title="coolbox.utilities.IntervalTree.difference_update"><code class="xref py py-obj docutils literal notranslate"><span class="pre">difference_update</span></code></a>(other)</p></td>
<td><p>Removes all intervals in other from self.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.discard" title="coolbox.utilities.IntervalTree.discard"><code class="xref py py-obj docutils literal notranslate"><span class="pre">discard</span></code></a>(interval)</p></td>
<td><p>Removes an interval from the tree, if present.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.discardi" title="coolbox.utilities.IntervalTree.discardi"><code class="xref py py-obj docutils literal notranslate"><span class="pre">discardi</span></code></a>(begin, end[, data])</p></td>
<td><p>Shortcut for discard(Interval(begin, end, data)).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.end" title="coolbox.utilities.IntervalTree.end"><code class="xref py py-obj docutils literal notranslate"><span class="pre">end</span></code></a>()</p></td>
<td><p>Returns the upper bound of the last interval in the tree.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.envelop" title="coolbox.utilities.IntervalTree.envelop"><code class="xref py py-obj docutils literal notranslate"><span class="pre">envelop</span></code></a>(begin[, end])</p></td>
<td><p>Returns the set of all intervals fully contained in the range [begin, end).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.find_nested" title="coolbox.utilities.IntervalTree.find_nested"><code class="xref py py-obj docutils literal notranslate"><span class="pre">find_nested</span></code></a>()</p></td>
<td><p>Returns a dictionary mapping parent intervals to sets of intervals overlapped by and contained in the parent.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.from_tuples" title="coolbox.utilities.IntervalTree.from_tuples"><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_tuples</span></code></a>(tups)</p></td>
<td><p>Create a new IntervalTree from an iterable of 2- or 3-tuples,</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.intersection" title="coolbox.utilities.IntervalTree.intersection"><code class="xref py py-obj docutils literal notranslate"><span class="pre">intersection</span></code></a>(other)</p></td>
<td><p>Returns a new tree of all intervals common to both self and other.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.intersection_update" title="coolbox.utilities.IntervalTree.intersection_update"><code class="xref py py-obj docutils literal notranslate"><span class="pre">intersection_update</span></code></a>(other)</p></td>
<td><p>Removes intervals from self unless they also exist in other.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.is_empty" title="coolbox.utilities.IntervalTree.is_empty"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_empty</span></code></a>()</p></td>
<td><p>Returns whether the tree is empty.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">isdisjoint</span></code>(other)</p></td>
<td><p>Return True if two sets have a null intersection.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.items" title="coolbox.utilities.IntervalTree.items"><code class="xref py py-obj docutils literal notranslate"><span class="pre">items</span></code></a>()</p></td>
<td><p>Constructs and returns a set of all intervals in the tree.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.iter" title="coolbox.utilities.IntervalTree.iter"><code class="xref py py-obj docutils literal notranslate"><span class="pre">iter</span></code></a>()</p></td>
<td><p>Returns an iterator over all the intervals in the tree.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.merge_equals" title="coolbox.utilities.IntervalTree.merge_equals"><code class="xref py py-obj docutils literal notranslate"><span class="pre">merge_equals</span></code></a>([data_reducer, data_initializer])</p></td>
<td><p>Finds all intervals with equal ranges and merges them into a single interval.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.merge_overlaps" title="coolbox.utilities.IntervalTree.merge_overlaps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">merge_overlaps</span></code></a>([data_reducer, …])</p></td>
<td><p>Finds all intervals with overlapping ranges and merges them into a single interval.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.overlap" title="coolbox.utilities.IntervalTree.overlap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">overlap</span></code></a>(begin[, end])</p></td>
<td><p>Returns a set of all intervals overlapping the given range.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.overlaps" title="coolbox.utilities.IntervalTree.overlaps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">overlaps</span></code></a>(begin[, end])</p></td>
<td><p>Returns whether some interval in the tree overlaps the given point or range.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.overlaps_point" title="coolbox.utilities.IntervalTree.overlaps_point"><code class="xref py py-obj docutils literal notranslate"><span class="pre">overlaps_point</span></code></a>(p)</p></td>
<td><p>Returns whether some interval in the tree overlaps p.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.overlaps_range" title="coolbox.utilities.IntervalTree.overlaps_range"><code class="xref py py-obj docutils literal notranslate"><span class="pre">overlaps_range</span></code></a>(begin, end)</p></td>
<td><p>Returns whether some interval in the tree overlaps the given range.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pop</span></code>()</p></td>
<td><p>Return the popped value.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.print_structure" title="coolbox.utilities.IntervalTree.print_structure"><code class="xref py py-obj docutils literal notranslate"><span class="pre">print_structure</span></code></a>([tostring])</p></td>
<td><p>## FOR DEBUGGING ONLY ## Pretty-prints the structure of the tree.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.range" title="coolbox.utilities.IntervalTree.range"><code class="xref py py-obj docutils literal notranslate"><span class="pre">range</span></code></a>()</p></td>
<td><p>Returns a minimum-spanning Interval that encloses all the members of this IntervalTree.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.remove" title="coolbox.utilities.IntervalTree.remove"><code class="xref py py-obj docutils literal notranslate"><span class="pre">remove</span></code></a>(interval)</p></td>
<td><p>Removes an interval from the tree, if present.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.remove_envelop" title="coolbox.utilities.IntervalTree.remove_envelop"><code class="xref py py-obj docutils literal notranslate"><span class="pre">remove_envelop</span></code></a>(begin, end)</p></td>
<td><p>Removes all intervals completely enveloped in the given range.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.remove_overlap" title="coolbox.utilities.IntervalTree.remove_overlap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">remove_overlap</span></code></a>(begin[, end])</p></td>
<td><p>Removes all intervals overlapping the given point or range.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.removei" title="coolbox.utilities.IntervalTree.removei"><code class="xref py py-obj docutils literal notranslate"><span class="pre">removei</span></code></a>(begin, end[, data])</p></td>
<td><p>Shortcut for remove(Interval(begin, end, data)).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.score" title="coolbox.utilities.IntervalTree.score"><code class="xref py py-obj docutils literal notranslate"><span class="pre">score</span></code></a>([full_report])</p></td>
<td><p>Returns a number between 0 and 1, indicating how suboptimal the tree is.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.slice" title="coolbox.utilities.IntervalTree.slice"><code class="xref py py-obj docutils literal notranslate"><span class="pre">slice</span></code></a>(point[, datafunc])</p></td>
<td><p>Split Intervals that overlap point into two new Intervals.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.span" title="coolbox.utilities.IntervalTree.span"><code class="xref py py-obj docutils literal notranslate"><span class="pre">span</span></code></a>()</p></td>
<td><p>Returns the length of the minimum-spanning Interval that encloses all the members of this IntervalTree.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.split_overlaps" title="coolbox.utilities.IntervalTree.split_overlaps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">split_overlaps</span></code></a>()</p></td>
<td><p>Finds all intervals with overlapping ranges and splits them along the range boundaries.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.symmetric_difference" title="coolbox.utilities.IntervalTree.symmetric_difference"><code class="xref py py-obj docutils literal notranslate"><span class="pre">symmetric_difference</span></code></a>(other)</p></td>
<td><p>Return a tree with elements only in self or other but not both.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.symmetric_difference_update" title="coolbox.utilities.IntervalTree.symmetric_difference_update"><code class="xref py py-obj docutils literal notranslate"><span class="pre">symmetric_difference_update</span></code></a>(other)</p></td>
<td><p>Throws out all intervals except those only in self or other, not both.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.union" title="coolbox.utilities.IntervalTree.union"><code class="xref py py-obj docutils literal notranslate"><span class="pre">union</span></code></a>(other)</p></td>
<td><p>Returns a new tree, comprising all intervals from self and other.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.update" title="coolbox.utilities.IntervalTree.update"><code class="xref py py-obj docutils literal notranslate"><span class="pre">update</span></code></a>(intervals)</p></td>
<td><p>Given an iterable of intervals, add them to the tree.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.verify" title="coolbox.utilities.IntervalTree.verify"><code class="xref py py-obj docutils literal notranslate"><span class="pre">verify</span></code></a>()</p></td>
<td><p>## FOR DEBUGGING ONLY ## Checks the table to ensure that the invariants are held.</p></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.add" title="coolbox.utilities.IntervalTree.add"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add</span></code></a>(interval)</p></td>
<td><p>Adds an interval to the tree, if not already present.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.addi" title="coolbox.utilities.IntervalTree.addi"><code class="xref py py-obj docutils literal notranslate"><span class="pre">addi</span></code></a>(begin, end[, data])</p></td>
<td><p>Shortcut for add(Interval(begin, end, data)).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.append" title="coolbox.utilities.IntervalTree.append"><code class="xref py py-obj docutils literal notranslate"><span class="pre">append</span></code></a>(interval)</p></td>
<td><p>Adds an interval to the tree, if not already present.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.appendi" title="coolbox.utilities.IntervalTree.appendi"><code class="xref py py-obj docutils literal notranslate"><span class="pre">appendi</span></code></a>(begin, end[, data])</p></td>
<td><p>Shortcut for add(Interval(begin, end, data)).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.at" title="coolbox.utilities.IntervalTree.at"><code class="xref py py-obj docutils literal notranslate"><span class="pre">at</span></code></a>(p)</p></td>
<td><p>Returns the set of all intervals that contain p.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.begin" title="coolbox.utilities.IntervalTree.begin"><code class="xref py py-obj docutils literal notranslate"><span class="pre">begin</span></code></a>()</p></td>
<td><p>Returns the lower bound of the first interval in the tree.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.chop" title="coolbox.utilities.IntervalTree.chop"><code class="xref py py-obj docutils literal notranslate"><span class="pre">chop</span></code></a>(begin, end[, datafunc])</p></td>
<td><p>Like remove_envelop(), but trims back Intervals hanging into the chopped area so that nothing overlaps.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.clear" title="coolbox.utilities.IntervalTree.clear"><code class="xref py py-obj docutils literal notranslate"><span class="pre">clear</span></code></a>()</p></td>
<td><p>Empties the tree.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.containsi" title="coolbox.utilities.IntervalTree.containsi"><code class="xref py py-obj docutils literal notranslate"><span class="pre">containsi</span></code></a>(begin, end[, data])</p></td>
<td><p>Shortcut for (Interval(begin, end, data) in tree).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.copy" title="coolbox.utilities.IntervalTree.copy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code></a>()</p></td>
<td><p>Construct a new IntervalTree using shallow copies of the intervals in the source tree.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.difference" title="coolbox.utilities.IntervalTree.difference"><code class="xref py py-obj docutils literal notranslate"><span class="pre">difference</span></code></a>(other)</p></td>
<td><p>Returns a new tree, comprising all intervals in self but not in other.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.difference_update" title="coolbox.utilities.IntervalTree.difference_update"><code class="xref py py-obj docutils literal notranslate"><span class="pre">difference_update</span></code></a>(other)</p></td>
<td><p>Removes all intervals in other from self.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.discard" title="coolbox.utilities.IntervalTree.discard"><code class="xref py py-obj docutils literal notranslate"><span class="pre">discard</span></code></a>(interval)</p></td>
<td><p>Removes an interval from the tree, if present.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.discardi" title="coolbox.utilities.IntervalTree.discardi"><code class="xref py py-obj docutils literal notranslate"><span class="pre">discardi</span></code></a>(begin, end[, data])</p></td>
<td><p>Shortcut for discard(Interval(begin, end, data)).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.end" title="coolbox.utilities.IntervalTree.end"><code class="xref py py-obj docutils literal notranslate"><span class="pre">end</span></code></a>()</p></td>
<td><p>Returns the upper bound of the last interval in the tree.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.envelop" title="coolbox.utilities.IntervalTree.envelop"><code class="xref py py-obj docutils literal notranslate"><span class="pre">envelop</span></code></a>(begin[, end])</p></td>
<td><p>Returns the set of all intervals fully contained in the range [begin, end).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.find_nested" title="coolbox.utilities.IntervalTree.find_nested"><code class="xref py py-obj docutils literal notranslate"><span class="pre">find_nested</span></code></a>()</p></td>
<td><p>Returns a dictionary mapping parent intervals to sets of intervals overlapped by and contained in the parent.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.from_tuples" title="coolbox.utilities.IntervalTree.from_tuples"><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_tuples</span></code></a>(tups)</p></td>
<td><p>Create a new IntervalTree from an iterable of 2- or 3-tuples,</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.intersection" title="coolbox.utilities.IntervalTree.intersection"><code class="xref py py-obj docutils literal notranslate"><span class="pre">intersection</span></code></a>(other)</p></td>
<td><p>Returns a new tree of all intervals common to both self and other.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.intersection_update" title="coolbox.utilities.IntervalTree.intersection_update"><code class="xref py py-obj docutils literal notranslate"><span class="pre">intersection_update</span></code></a>(other)</p></td>
<td><p>Removes intervals from self unless they also exist in other.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.is_empty" title="coolbox.utilities.IntervalTree.is_empty"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_empty</span></code></a>()</p></td>
<td><p>Returns whether the tree is empty.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.items" title="coolbox.utilities.IntervalTree.items"><code class="xref py py-obj docutils literal notranslate"><span class="pre">items</span></code></a>()</p></td>
<td><p>Constructs and returns a set of all intervals in the tree.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.iter" title="coolbox.utilities.IntervalTree.iter"><code class="xref py py-obj docutils literal notranslate"><span class="pre">iter</span></code></a>()</p></td>
<td><p>Returns an iterator over all the intervals in the tree.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.merge_equals" title="coolbox.utilities.IntervalTree.merge_equals"><code class="xref py py-obj docutils literal notranslate"><span class="pre">merge_equals</span></code></a>([data_reducer, data_initializer])</p></td>
<td><p>Finds all intervals with equal ranges and merges them into a single interval.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.merge_overlaps" title="coolbox.utilities.IntervalTree.merge_overlaps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">merge_overlaps</span></code></a>([data_reducer, …])</p></td>
<td><p>Finds all intervals with overlapping ranges and merges them into a single interval.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.overlap" title="coolbox.utilities.IntervalTree.overlap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">overlap</span></code></a>(begin[, end])</p></td>
<td><p>Returns a set of all intervals overlapping the given range.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.overlaps" title="coolbox.utilities.IntervalTree.overlaps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">overlaps</span></code></a>(begin[, end])</p></td>
<td><p>Returns whether some interval in the tree overlaps the given point or range.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.overlaps_point" title="coolbox.utilities.IntervalTree.overlaps_point"><code class="xref py py-obj docutils literal notranslate"><span class="pre">overlaps_point</span></code></a>(p)</p></td>
<td><p>Returns whether some interval in the tree overlaps p.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.overlaps_range" title="coolbox.utilities.IntervalTree.overlaps_range"><code class="xref py py-obj docutils literal notranslate"><span class="pre">overlaps_range</span></code></a>(begin, end)</p></td>
<td><p>Returns whether some interval in the tree overlaps the given range.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.print_structure" title="coolbox.utilities.IntervalTree.print_structure"><code class="xref py py-obj docutils literal notranslate"><span class="pre">print_structure</span></code></a>([tostring])</p></td>
<td><p>## FOR DEBUGGING ONLY ## Pretty-prints the structure of the tree.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.range" title="coolbox.utilities.IntervalTree.range"><code class="xref py py-obj docutils literal notranslate"><span class="pre">range</span></code></a>()</p></td>
<td><p>Returns a minimum-spanning Interval that encloses all the members of this IntervalTree.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.remove" title="coolbox.utilities.IntervalTree.remove"><code class="xref py py-obj docutils literal notranslate"><span class="pre">remove</span></code></a>(interval)</p></td>
<td><p>Removes an interval from the tree, if present.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.remove_envelop" title="coolbox.utilities.IntervalTree.remove_envelop"><code class="xref py py-obj docutils literal notranslate"><span class="pre">remove_envelop</span></code></a>(begin, end)</p></td>
<td><p>Removes all intervals completely enveloped in the given range.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.remove_overlap" title="coolbox.utilities.IntervalTree.remove_overlap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">remove_overlap</span></code></a>(begin[, end])</p></td>
<td><p>Removes all intervals overlapping the given point or range.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.removei" title="coolbox.utilities.IntervalTree.removei"><code class="xref py py-obj docutils literal notranslate"><span class="pre">removei</span></code></a>(begin, end[, data])</p></td>
<td><p>Shortcut for remove(Interval(begin, end, data)).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.score" title="coolbox.utilities.IntervalTree.score"><code class="xref py py-obj docutils literal notranslate"><span class="pre">score</span></code></a>([full_report])</p></td>
<td><p>Returns a number between 0 and 1, indicating how suboptimal the tree is.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.slice" title="coolbox.utilities.IntervalTree.slice"><code class="xref py py-obj docutils literal notranslate"><span class="pre">slice</span></code></a>(point[, datafunc])</p></td>
<td><p>Split Intervals that overlap point into two new Intervals.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.span" title="coolbox.utilities.IntervalTree.span"><code class="xref py py-obj docutils literal notranslate"><span class="pre">span</span></code></a>()</p></td>
<td><p>Returns the length of the minimum-spanning Interval that encloses all the members of this IntervalTree.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.split_overlaps" title="coolbox.utilities.IntervalTree.split_overlaps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">split_overlaps</span></code></a>()</p></td>
<td><p>Finds all intervals with overlapping ranges and splits them along the range boundaries.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.symmetric_difference" title="coolbox.utilities.IntervalTree.symmetric_difference"><code class="xref py py-obj docutils literal notranslate"><span class="pre">symmetric_difference</span></code></a>(other)</p></td>
<td><p>Return a tree with elements only in self or other but not both.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.symmetric_difference_update" title="coolbox.utilities.IntervalTree.symmetric_difference_update"><code class="xref py py-obj docutils literal notranslate"><span class="pre">symmetric_difference_update</span></code></a>(other)</p></td>
<td><p>Throws out all intervals except those only in self or other, not both.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.union" title="coolbox.utilities.IntervalTree.union"><code class="xref py py-obj docutils literal notranslate"><span class="pre">union</span></code></a>(other)</p></td>
<td><p>Returns a new tree, comprising all intervals from self and other.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.update" title="coolbox.utilities.IntervalTree.update"><code class="xref py py-obj docutils literal notranslate"><span class="pre">update</span></code></a>(intervals)</p></td>
<td><p>Given an iterable of intervals, add them to the tree.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.IntervalTree.verify" title="coolbox.utilities.IntervalTree.verify"><code class="xref py py-obj docutils literal notranslate"><span class="pre">verify</span></code></a>()</p></td>
<td><p>## FOR DEBUGGING ONLY ## Checks the table to ensure that the invariants are held.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.add">
<span class="sig-name descname"><span class="pre">add</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.add" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds an interval to the tree, if not already present.</p>
<p>Completes in O(log n) time.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.addi">
<span class="sig-name descname"><span class="pre">addi</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">begin</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.addi" title="Permalink to this definition">¶</a></dt>
<dd><p>Shortcut for add(Interval(begin, end, data)).</p>
<p>Completes in O(log n) time.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.append">
<span class="sig-name descname"><span class="pre">append</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.append" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds an interval to the tree, if not already present.</p>
<p>Completes in O(log n) time.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.appendi">
<span class="sig-name descname"><span class="pre">appendi</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">begin</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.appendi" title="Permalink to this definition">¶</a></dt>
<dd><p>Shortcut for add(Interval(begin, end, data)).</p>
<p>Completes in O(log n) time.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.at">
<span class="sig-name descname"><span class="pre">at</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">p</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.at" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the set of all intervals that contain p.</p>
<dl class="simple">
<dt>Completes in O(m + log n) time, where:</dt><dd><ul class="simple">
<li><p>n = size of the tree</p></li>
<li><p>m = number of matches</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>set of Interval</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.begin">
<span class="sig-name descname"><span class="pre">begin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.begin" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the lower bound of the first interval in the tree.</p>
<p>Completes in O(1) time.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.chop">
<span class="sig-name descname"><span class="pre">chop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">begin</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">datafunc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.chop" title="Permalink to this definition">¶</a></dt>
<dd><p>Like remove_envelop(), but trims back Intervals hanging into
the chopped area so that nothing overlaps.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.clear">
<span class="sig-name descname"><span class="pre">clear</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.clear" title="Permalink to this definition">¶</a></dt>
<dd><p>Empties the tree.</p>
<p>Completes in O(1) tine.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.containsi">
<span class="sig-name descname"><span class="pre">containsi</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">begin</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.containsi" title="Permalink to this definition">¶</a></dt>
<dd><p>Shortcut for (Interval(begin, end, data) in tree).</p>
<p>Completes in O(1) time.
:rtype: bool</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.copy" title="Permalink to this definition">¶</a></dt>
<dd><p>Construct a new IntervalTree using shallow copies of the
intervals in the source tree.</p>
<p>Completes in O(n*log n) time.
:rtype: IntervalTree</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.difference">
<span class="sig-name descname"><span class="pre">difference</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.difference" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a new tree, comprising all intervals in self but not
in other.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.difference_update">
<span class="sig-name descname"><span class="pre">difference_update</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.difference_update" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes all intervals in other from self.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.discard">
<span class="sig-name descname"><span class="pre">discard</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.discard" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes an interval from the tree, if present. If not, does
nothing.</p>
<p>Completes in O(log n) time.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.discardi">
<span class="sig-name descname"><span class="pre">discardi</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">begin</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.discardi" title="Permalink to this definition">¶</a></dt>
<dd><p>Shortcut for discard(Interval(begin, end, data)).</p>
<p>Completes in O(log n) time.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.end">
<span class="sig-name descname"><span class="pre">end</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.end" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the upper bound of the last interval in the tree.</p>
<p>Completes in O(1) time.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.envelop">
<span class="sig-name descname"><span class="pre">envelop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">begin</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.envelop" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the set of all intervals fully contained in the range
[begin, end).</p>
<dl class="simple">
<dt>Completes in O(m + k*log n) time, where:</dt><dd><ul class="simple">
<li><p>n = size of the tree</p></li>
<li><p>m = number of matches</p></li>
<li><p>k = size of the search range</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>set of Interval</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.find_nested">
<span class="sig-name descname"><span class="pre">find_nested</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.find_nested" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary mapping parent intervals to sets of
intervals overlapped by and contained in the parent.</p>
<p>Completes in O(n^2) time.
:rtype: dict of [Interval, set of Interval]</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.from_tuples">
<em class="property"><span class="pre">classmethod</span> </em><span class="sig-name descname"><span class="pre">from_tuples</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tups</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.from_tuples" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Create a new IntervalTree from an iterable of 2- or 3-tuples,</dt><dd><p>where the tuple lists begin, end, and optionally data.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.intersection">
<span class="sig-name descname"><span class="pre">intersection</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.intersection" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a new tree of all intervals common to both self and
other.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.intersection_update">
<span class="sig-name descname"><span class="pre">intersection_update</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.intersection_update" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes intervals from self unless they also exist in other.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.is_empty">
<span class="sig-name descname"><span class="pre">is_empty</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.is_empty" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns whether the tree is empty.</p>
<p>Completes in O(1) time.
:rtype: bool</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.items">
<span class="sig-name descname"><span class="pre">items</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.items" title="Permalink to this definition">¶</a></dt>
<dd><p>Constructs and returns a set of all intervals in the tree.</p>
<p>Completes in O(n) time.
:rtype: set of Interval</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.iter">
<span class="sig-name descname"><span class="pre">iter</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.iter" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an iterator over all the intervals in the tree.</p>
<p>Completes in O(1) time.
:rtype: collections.Iterable[Interval]</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.merge_equals">
<span class="sig-name descname"><span class="pre">merge_equals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_reducer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data_initializer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.merge_equals" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds all intervals with equal ranges and merges them
into a single interval. If provided, uses data_reducer and
data_initializer with similar semantics to Python’s built-in
reduce(reducer_func[, initializer]), as follows:</p>
<p>If data_reducer is set to a function, combines the data
fields of the Intervals with</p>
<blockquote>
<div><p>current_reduced_data = data_reducer(current_reduced_data, new_data)</p>
</div></blockquote>
<p>If data_reducer is None, the merged Interval’s data
field will be set to None, ignoring all the data fields
of the merged Intervals.</p>
<p>On encountering the first Interval to merge, if
data_initializer is None (default), uses the first
Interval’s data field as the first value for
current_reduced_data. If data_initializer is not None,
current_reduced_data is set to a shallow copy of
data_initiazer created with</p>
<blockquote>
<div><p>copy.copy(data_initializer).</p>
</div></blockquote>
<p>Completes in O(n*logn).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.merge_overlaps">
<span class="sig-name descname"><span class="pre">merge_overlaps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_reducer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data_initializer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.merge_overlaps" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds all intervals with overlapping ranges and merges them
into a single interval. If provided, uses data_reducer and
data_initializer with similar semantics to Python’s built-in
reduce(reducer_func[, initializer]), as follows:</p>
<p>If data_reducer is set to a function, combines the data
fields of the Intervals with</p>
<blockquote>
<div><p>current_reduced_data = data_reducer(current_reduced_data, new_data)</p>
</div></blockquote>
<p>If data_reducer is None, the merged Interval’s data
field will be set to None, ignoring all the data fields
of the merged Intervals.</p>
<p>On encountering the first Interval to merge, if
data_initializer is None (default), uses the first
Interval’s data field as the first value for
current_reduced_data. If data_initializer is not None,
current_reduced_data is set to a shallow copy of
data_initializer created with copy.copy(data_initializer).</p>
<p>If strict is True (default), intervals are only merged if
their ranges actually overlap; adjacent, touching intervals
will not be merged. If strict is False, intervals are merged
even if they are only end-to-end adjacent.</p>
<p>Completes in O(n*logn).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.overlap">
<span class="sig-name descname"><span class="pre">overlap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">begin</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.overlap" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a set of all intervals overlapping the given range.</p>
<dl class="simple">
<dt>Completes in O(m + k*log n) time, where:</dt><dd><ul class="simple">
<li><p>n = size of the tree</p></li>
<li><p>m = number of matches</p></li>
<li><p>k = size of the search range</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>set of Interval</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.overlaps">
<span class="sig-name descname"><span class="pre">overlaps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">begin</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.overlaps" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns whether some interval in the tree overlaps the given
point or range.</p>
<p>Completes in O(r*log n) time, where r is the size of the
search range.
:rtype: bool</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.overlaps_point">
<span class="sig-name descname"><span class="pre">overlaps_point</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">p</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.overlaps_point" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns whether some interval in the tree overlaps p.</p>
<p>Completes in O(log n) time.
:rtype: bool</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.overlaps_range">
<span class="sig-name descname"><span class="pre">overlaps_range</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">begin</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.overlaps_range" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns whether some interval in the tree overlaps the given
range. Returns False if given a null interval over which to
test.</p>
<p>Completes in O(r*log n) time, where r is the range length and n
is the table size.
:rtype: bool</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.print_structure">
<span class="sig-name descname"><span class="pre">print_structure</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tostring</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.print_structure" title="Permalink to this definition">¶</a></dt>
<dd><p>## FOR DEBUGGING ONLY ##
Pretty-prints the structure of the tree.
If tostring is true, prints nothing and returns a string.
:rtype: None or str</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.range">
<span class="sig-name descname"><span class="pre">range</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.range" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a minimum-spanning Interval that encloses all the
members of this IntervalTree. If the tree is empty, returns
null Interval.
:rtype: Interval</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.remove">
<span class="sig-name descname"><span class="pre">remove</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.remove" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes an interval from the tree, if present. If not, raises
ValueError.</p>
<p>Completes in O(log n) time.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.remove_envelop">
<span class="sig-name descname"><span class="pre">remove_envelop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">begin</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.remove_envelop" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes all intervals completely enveloped in the given range.</p>
<dl class="simple">
<dt>Completes in O((r+m)*log n) time, where:</dt><dd><ul class="simple">
<li><p>n = size of the tree</p></li>
<li><p>m = number of matches</p></li>
<li><p>r = size of the search range</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.remove_overlap">
<span class="sig-name descname"><span class="pre">remove_overlap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">begin</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.remove_overlap" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes all intervals overlapping the given point or range.</p>
<dl class="simple">
<dt>Completes in O((r+m)*log n) time, where:</dt><dd><ul class="simple">
<li><p>n = size of the tree</p></li>
<li><p>m = number of matches</p></li>
<li><p>r = size of the search range (this is 1 for a point)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.removei">
<span class="sig-name descname"><span class="pre">removei</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">begin</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.removei" title="Permalink to this definition">¶</a></dt>
<dd><p>Shortcut for remove(Interval(begin, end, data)).</p>
<p>Completes in O(log n) time.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.score">
<span class="sig-name descname"><span class="pre">score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">full_report</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.score" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a number between 0 and 1, indicating how suboptimal the tree
is. The lower, the better. Roughly, this number represents the
fraction of flawed Intervals in the tree.
:rtype: float</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.slice">
<span class="sig-name descname"><span class="pre">slice</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">point</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">datafunc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.slice" title="Permalink to this definition">¶</a></dt>
<dd><p>Split Intervals that overlap point into two new Intervals. if
specified, uses datafunc(interval, islower=True/False) to
set the data field of the new Intervals.
:param point: where to slice
:param datafunc(interval, isupper): callable returning a new
value for the interval’s data field</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.span">
<span class="sig-name descname"><span class="pre">span</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.span" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the length of the minimum-spanning Interval that
encloses all the members of this IntervalTree. If the tree
is empty, return 0.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.split_overlaps">
<span class="sig-name descname"><span class="pre">split_overlaps</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.split_overlaps" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds all intervals with overlapping ranges and splits them
along the range boundaries.</p>
<p>Completes in worst-case O(n^2*log n) time (many interval
boundaries are inside many intervals), best-case O(n*log n)
time (small number of overlaps &lt;&lt; n per interval).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.symmetric_difference">
<span class="sig-name descname"><span class="pre">symmetric_difference</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.symmetric_difference" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a tree with elements only in self or other but not
both.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.symmetric_difference_update">
<span class="sig-name descname"><span class="pre">symmetric_difference_update</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.symmetric_difference_update" title="Permalink to this definition">¶</a></dt>
<dd><p>Throws out all intervals except those only in self or other,
not both.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.union">
<span class="sig-name descname"><span class="pre">union</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.union" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a new tree, comprising all intervals from self
and other.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.update">
<span class="sig-name descname"><span class="pre">update</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">intervals</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.update" title="Permalink to this definition">¶</a></dt>
<dd><p>Given an iterable of intervals, add them to the tree.</p>
<p>Completes in O(m*log(n+m), where m = number of intervals to
add.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.IntervalTree.verify">
<span class="sig-name descname"><span class="pre">verify</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.IntervalTree.verify" title="Permalink to this definition">¶</a></dt>
<dd><p>## FOR DEBUGGING ONLY ##
Checks the table to ensure that the invariants are held.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.utilities.ReadBed">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.utilities.</span></span><span class="sig-name descname"><span class="pre">ReadBed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_handle</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.ReadBed" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads a bed file. Based on the number of fields
it tries to guess the type of bed file used. Current options
are bed3, bed6 and bed12</p>
<p>Example:
bed = readBed(open(“file.bed”, ‘r’))
for interval in bed:</p>
<blockquote>
<div><p>print(interval[‘start’])</p>
</div></blockquote>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.ReadBed.get_bed_interval" title="coolbox.utilities.ReadBed.get_bed_interval"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_bed_interval</span></code></a>(bed_line)</p></td>
<td><p>Processes each bed line from a bed file, casts the values and returns a namedtuple object</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.ReadBed.get_no_comment_line" title="coolbox.utilities.ReadBed.get_no_comment_line"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_no_comment_line</span></code></a>([iter, count])</p></td>
<td><p>Skips comment lines starting with ‘#’ “track” or “browser” in the bed files</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.ReadBed.guess_file_type" title="coolbox.utilities.ReadBed.guess_file_type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">guess_file_type</span></code></a>(file_iter)</p></td>
<td><p>try to guess type of bed file by counting the fields</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.ReadBed.next" title="coolbox.utilities.ReadBed.next"><code class="xref py py-obj docutils literal notranslate"><span class="pre">next</span></code></a>()</p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.ReadBed.get_bed_interval" title="coolbox.utilities.ReadBed.get_bed_interval"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_bed_interval</span></code></a>(bed_line)</p></td>
<td><p>Processes each bed line from a bed file, casts the values and returns a namedtuple object</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.ReadBed.get_no_comment_line" title="coolbox.utilities.ReadBed.get_no_comment_line"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_no_comment_line</span></code></a>([iter, count])</p></td>
<td><p>Skips comment lines starting with ‘#’ “track” or “browser” in the bed files</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.ReadBed.guess_file_type" title="coolbox.utilities.ReadBed.guess_file_type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">guess_file_type</span></code></a>(file_iter)</p></td>
<td><p>try to guess type of bed file by counting the fields</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.ReadBed.next" title="coolbox.utilities.ReadBed.next"><code class="xref py py-obj docutils literal notranslate"><span class="pre">next</span></code></a>()</p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.ReadBed.get_bed_interval">
<span class="sig-name descname"><span class="pre">get_bed_interval</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bed_line</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.ReadBed.get_bed_interval" title="Permalink to this definition">¶</a></dt>
<dd><p>Processes each bed line from a bed file, casts the values and returns
a namedtuple object</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bed_line</span><span class="o">=</span><span class="s2">&quot;chr1</span><span class="se">\t</span><span class="s2">0</span><span class="se">\t</span><span class="s2">1000</span><span class="se">\t</span><span class="s2">gene_1</span><span class="se">\t</span><span class="s2">0.5</span><span class="se">\t</span><span class="s2">-</span><span class="se">\t</span><span class="s2">0</span><span class="se">\t</span><span class="s2">1000</span><span class="se">\t</span><span class="s2">0</span><span class="se">\t</span><span class="s2">3</span><span class="se">\t</span><span class="s2">10,20,100</span><span class="se">\t</span><span class="s2">20,200,700&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;/tmp/test.bed&#39;</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fh</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">foo</span> <span class="o">=</span> <span class="n">fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">bed_line</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bed_f</span> <span class="o">=</span> <span class="n">ReadBed</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="s1">&#39;/tmp/test.bed&#39;</span><span class="p">,</span><span class="s1">&#39;r&#39;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bed</span> <span class="o">=</span> <span class="n">bed_f</span><span class="o">.</span><span class="n">get_bed_interval</span><span class="p">(</span><span class="n">bed_line</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bed</span><span class="o">.</span><span class="n">chromosome</span>
<span class="go">&#39;chr1&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bed</span><span class="o">.</span><span class="n">block_starts</span>
<span class="go">[20, 200, 700]</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bed_line</span><span class="o">=</span><span class="s2">&quot;chr2</span><span class="se">\t</span><span class="s2">0</span><span class="se">\t</span><span class="s2">1000</span><span class="se">\t</span><span class="s2">gene_1</span><span class="se">\t</span><span class="s2">0.5</span><span class="se">\t</span><span class="s2">-</span><span class="se">\n</span><span class="s2">&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;/tmp/test.bed&#39;</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fh</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">foo</span> <span class="o">=</span> <span class="n">fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">bed_line</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bed_f</span> <span class="o">=</span> <span class="n">ReadBed</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="s1">&#39;/tmp/test.bed&#39;</span><span class="p">,</span><span class="s1">&#39;r&#39;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bed_f</span><span class="o">.</span><span class="n">get_bed_interval</span><span class="p">(</span><span class="n">bed_line</span><span class="p">)</span>
<span class="go">BedInterval(chromosome=&#39;chr2&#39;, start=0, end=1000, name=&#39;gene_1&#39;, score=0.5, strand=&#39;-&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.ReadBed.get_no_comment_line">
<span class="sig-name descname"><span class="pre">get_no_comment_line</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">iter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">count</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.ReadBed.get_no_comment_line" title="Permalink to this definition">¶</a></dt>
<dd><p>Skips comment lines starting with ‘#’
“track” or “browser” in the bed files</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.ReadBed.guess_file_type">
<span class="sig-name descname"><span class="pre">guess_file_type</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_iter</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.ReadBed.guess_file_type" title="Permalink to this definition">¶</a></dt>
<dd><p>try to guess type of bed file by counting the fields</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.ReadBed.next">
<span class="sig-name descname"><span class="pre">next</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.ReadBed.next" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="coolbox.utilities.abspath">
<span class="sig-prename descclassname"><span class="pre">coolbox.utilities.</span></span><span class="sig-name descname"><span class="pre">abspath</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.abspath" title="Permalink to this definition">¶</a></dt>
<dd><p>Return an absolute path.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="coolbox.utilities.change_chrom_names">
<span class="sig-prename descclassname"><span class="pre">coolbox.utilities.</span></span><span class="sig-name descname"><span class="pre">change_chrom_names</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chrom</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.change_chrom_names" title="Permalink to this definition">¶</a></dt>
<dd><p>Changes UCSC chromosome names to ensembl chromosome names
and vice versa.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">change_chrom_names</span><span class="p">(</span><span class="s2">&quot;chr1&quot;</span><span class="p">)</span>
<span class="go">&#39;1&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">change_chrom_names</span><span class="p">(</span><span class="s2">&quot;1&quot;</span><span class="p">)</span>
<span class="go">&#39;chr1&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="coolbox.utilities.cm2inch">
<span class="sig-prename descclassname"><span class="pre">coolbox.utilities.</span></span><span class="sig-name descname"><span class="pre">cm2inch</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">tupl</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.cm2inch" title="Permalink to this definition">¶</a></dt>
<dd><p>convert length unit from cm to inch.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">cm2inch</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
<span class="go">3.937007874015748</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cm2inch</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="go">(0.7874015748031495, 0.7874015748031495)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cm2inch</span><span class="p">((</span><span class="mi">1</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>
<span class="go">(0.39370078740157477, 1.968503937007874)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="coolbox.utilities.correspond_track">
<span class="sig-prename descclassname"><span class="pre">coolbox.utilities.</span></span><span class="sig-name descname"><span class="pre">correspond_track</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">track_or_name</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">Track</span><span class="p"><span class="pre">,</span> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.correspond_track" title="Permalink to this definition">¶</a></dt>
<dd><p>Search track by it’s name.
For mark corresponding track in a element definition, allow
pass track name to find related object in CLI mode.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.utilities.deque">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.utilities.</span></span><span class="sig-name descname"><span class="pre">deque</span></span><a class="headerlink" href="#coolbox.utilities.deque" title="Permalink to this definition">¶</a></dt>
<dd><p>deque([iterable[, maxlen]]) –&gt; deque object</p>
<p>A list-like sequence optimized for data accesses near its endpoints.</p>
<dl class="field-list simple">
<dt class="field-odd">Attributes</dt>
<dd class="field-odd"><dl class="simple">
<dt><a class="reference internal" href="#coolbox.utilities.deque.maxlen" title="coolbox.utilities.deque.maxlen"><code class="xref py py-obj docutils literal notranslate"><span class="pre">maxlen</span></code></a></dt><dd><p>maximum size of a deque or None if unbounded</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.deque.append" title="coolbox.utilities.deque.append"><code class="xref py py-obj docutils literal notranslate"><span class="pre">append</span></code></a></p></td>
<td><p>Add an element to the right side of the deque.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.deque.appendleft" title="coolbox.utilities.deque.appendleft"><code class="xref py py-obj docutils literal notranslate"><span class="pre">appendleft</span></code></a></p></td>
<td><p>Add an element to the left side of the deque.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.deque.clear" title="coolbox.utilities.deque.clear"><code class="xref py py-obj docutils literal notranslate"><span class="pre">clear</span></code></a></p></td>
<td><p>Remove all elements from the deque.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.deque.copy" title="coolbox.utilities.deque.copy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code></a></p></td>
<td><p>Return a shallow copy of a deque.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.deque.count" title="coolbox.utilities.deque.count"><code class="xref py py-obj docutils literal notranslate"><span class="pre">count</span></code></a>(value)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.deque.extend" title="coolbox.utilities.deque.extend"><code class="xref py py-obj docutils literal notranslate"><span class="pre">extend</span></code></a></p></td>
<td><p>Extend the right side of the deque with elements from the iterable</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.deque.extendleft" title="coolbox.utilities.deque.extendleft"><code class="xref py py-obj docutils literal notranslate"><span class="pre">extendleft</span></code></a></p></td>
<td><p>Extend the left side of the deque with elements from the iterable</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.deque.index" title="coolbox.utilities.deque.index"><code class="xref py py-obj docutils literal notranslate"><span class="pre">index</span></code></a>(value, [start, [stop]])</p></td>
<td><p>Raises ValueError if the value is not present.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.deque.insert" title="coolbox.utilities.deque.insert"><code class="xref py py-obj docutils literal notranslate"><span class="pre">insert</span></code></a></p></td>
<td><p>D.insert(index, object) – insert object before index</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.deque.pop" title="coolbox.utilities.deque.pop"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pop</span></code></a></p></td>
<td><p>Remove and return the rightmost element.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.deque.popleft" title="coolbox.utilities.deque.popleft"><code class="xref py py-obj docutils literal notranslate"><span class="pre">popleft</span></code></a></p></td>
<td><p>Remove and return the leftmost element.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.deque.remove" title="coolbox.utilities.deque.remove"><code class="xref py py-obj docutils literal notranslate"><span class="pre">remove</span></code></a></p></td>
<td><p>D.remove(value) – remove first occurrence of value.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.deque.reverse" title="coolbox.utilities.deque.reverse"><code class="xref py py-obj docutils literal notranslate"><span class="pre">reverse</span></code></a></p></td>
<td><p>D.reverse() – reverse <em>IN PLACE</em></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.deque.rotate" title="coolbox.utilities.deque.rotate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">rotate</span></code></a></p></td>
<td><p>Rotate the deque n steps to the right (default n=1).</p></td>
</tr>
</tbody>
</table>
<p><strong>Methods:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.deque.append" title="coolbox.utilities.deque.append"><code class="xref py py-obj docutils literal notranslate"><span class="pre">append</span></code></a></p></td>
<td><p>Add an element to the right side of the deque.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.deque.appendleft" title="coolbox.utilities.deque.appendleft"><code class="xref py py-obj docutils literal notranslate"><span class="pre">appendleft</span></code></a></p></td>
<td><p>Add an element to the left side of the deque.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.deque.clear" title="coolbox.utilities.deque.clear"><code class="xref py py-obj docutils literal notranslate"><span class="pre">clear</span></code></a></p></td>
<td><p>Remove all elements from the deque.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.deque.copy" title="coolbox.utilities.deque.copy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code></a></p></td>
<td><p>Return a shallow copy of a deque.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.deque.count" title="coolbox.utilities.deque.count"><code class="xref py py-obj docutils literal notranslate"><span class="pre">count</span></code></a>(value)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.deque.extend" title="coolbox.utilities.deque.extend"><code class="xref py py-obj docutils literal notranslate"><span class="pre">extend</span></code></a></p></td>
<td><p>Extend the right side of the deque with elements from the iterable</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.deque.extendleft" title="coolbox.utilities.deque.extendleft"><code class="xref py py-obj docutils literal notranslate"><span class="pre">extendleft</span></code></a></p></td>
<td><p>Extend the left side of the deque with elements from the iterable</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.deque.index" title="coolbox.utilities.deque.index"><code class="xref py py-obj docutils literal notranslate"><span class="pre">index</span></code></a>(value, [start, [stop]])</p></td>
<td><p>Raises ValueError if the value is not present.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.deque.insert" title="coolbox.utilities.deque.insert"><code class="xref py py-obj docutils literal notranslate"><span class="pre">insert</span></code></a></p></td>
<td><p>D.insert(index, object) – insert object before index</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.deque.pop" title="coolbox.utilities.deque.pop"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pop</span></code></a></p></td>
<td><p>Remove and return the rightmost element.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.deque.popleft" title="coolbox.utilities.deque.popleft"><code class="xref py py-obj docutils literal notranslate"><span class="pre">popleft</span></code></a></p></td>
<td><p>Remove and return the leftmost element.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.deque.remove" title="coolbox.utilities.deque.remove"><code class="xref py py-obj docutils literal notranslate"><span class="pre">remove</span></code></a></p></td>
<td><p>D.remove(value) – remove first occurrence of value.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.deque.reverse" title="coolbox.utilities.deque.reverse"><code class="xref py py-obj docutils literal notranslate"><span class="pre">reverse</span></code></a></p></td>
<td><p>D.reverse() – reverse <em>IN PLACE</em></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.deque.rotate" title="coolbox.utilities.deque.rotate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">rotate</span></code></a></p></td>
<td><p>Rotate the deque n steps to the right (default n=1).</p></td>
</tr>
</tbody>
</table>
<p><strong>Attributes:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.deque.maxlen" title="coolbox.utilities.deque.maxlen"><code class="xref py py-obj docutils literal notranslate"><span class="pre">maxlen</span></code></a></p></td>
<td><p>maximum size of a deque or None if unbounded</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.deque.append">
<span class="sig-name descname"><span class="pre">append</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.deque.append" title="Permalink to this definition">¶</a></dt>
<dd><p>Add an element to the right side of the deque.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.deque.appendleft">
<span class="sig-name descname"><span class="pre">appendleft</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.deque.appendleft" title="Permalink to this definition">¶</a></dt>
<dd><p>Add an element to the left side of the deque.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.deque.clear">
<span class="sig-name descname"><span class="pre">clear</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.deque.clear" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove all elements from the deque.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.deque.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.deque.copy" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a shallow copy of a deque.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.deque.count">
<span class="sig-name descname"><span class="pre">count</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">value</span></span></em><span class="sig-paren">)</span> &#x2192; <span class="pre">integer</span> <span class="pre">–</span> <span class="pre">return</span> <span class="pre">number</span> <span class="pre">of</span> <span class="pre">occurrences</span> <span class="pre">of</span> <span class="pre">value</span><a class="headerlink" href="#coolbox.utilities.deque.count" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.deque.extend">
<span class="sig-name descname"><span class="pre">extend</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.deque.extend" title="Permalink to this definition">¶</a></dt>
<dd><p>Extend the right side of the deque with elements from the iterable</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.deque.extendleft">
<span class="sig-name descname"><span class="pre">extendleft</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.deque.extendleft" title="Permalink to this definition">¶</a></dt>
<dd><p>Extend the left side of the deque with elements from the iterable</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.deque.index">
<span class="sig-name descname"><span class="pre">index</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">value</span></em><span class="optional">[</span>, <em class="sig-param"><span class="pre">start</span></em><span class="optional">[</span>, <em class="sig-param"><span class="pre">stop</span></em><span class="optional">]</span><span class="optional">]</span><span class="sig-paren">)</span> &#x2192; <span class="pre">integer</span> <span class="pre">–</span> <span class="pre">return</span> <span class="pre">first</span> <span class="pre">index</span> <span class="pre">of</span> <span class="pre">value.</span><a class="headerlink" href="#coolbox.utilities.deque.index" title="Permalink to this definition">¶</a></dt>
<dd><p>Raises ValueError if the value is not present.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.deque.insert">
<span class="sig-name descname"><span class="pre">insert</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.deque.insert" title="Permalink to this definition">¶</a></dt>
<dd><p>D.insert(index, object) – insert object before index</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="coolbox.utilities.deque.maxlen">
<span class="sig-name descname"><span class="pre">maxlen</span></span><a class="headerlink" href="#coolbox.utilities.deque.maxlen" title="Permalink to this definition">¶</a></dt>
<dd><p>maximum size of a deque or None if unbounded</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.deque.pop">
<span class="sig-name descname"><span class="pre">pop</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.deque.pop" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove and return the rightmost element.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.deque.popleft">
<span class="sig-name descname"><span class="pre">popleft</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.deque.popleft" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove and return the leftmost element.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.deque.remove">
<span class="sig-name descname"><span class="pre">remove</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.deque.remove" title="Permalink to this definition">¶</a></dt>
<dd><p>D.remove(value) – remove first occurrence of value.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.deque.reverse">
<span class="sig-name descname"><span class="pre">reverse</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.deque.reverse" title="Permalink to this definition">¶</a></dt>
<dd><p>D.reverse() – reverse <em>IN PLACE</em></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coolbox.utilities.deque.rotate">
<span class="sig-name descname"><span class="pre">rotate</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.deque.rotate" title="Permalink to this definition">¶</a></dt>
<dd><p>Rotate the deque n steps to the right (default n=1).  If n is negative, rotates left.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="coolbox.utilities.dirname">
<span class="sig-prename descclassname"><span class="pre">coolbox.utilities.</span></span><span class="sig-name descname"><span class="pre">dirname</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">p</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.dirname" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the directory component of a pathname</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="coolbox.utilities.fig2bytes">
<span class="sig-prename descclassname"><span class="pre">coolbox.utilities.</span></span><span class="sig-name descname"><span class="pre">fig2bytes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fig</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">encode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'svg'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.fig2bytes" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert matplotlib.figure.Figure/IPython.display.SVG
object to image bytes.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="coolbox.utilities.file_to_intervaltree">
<span class="sig-prename descclassname"><span class="pre">coolbox.utilities.</span></span><span class="sig-name descname"><span class="pre">file_to_intervaltree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.file_to_intervaltree" title="Permalink to this definition">¶</a></dt>
<dd><p>converts a BED like file into a bx python interval tree</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>file_name</strong><span class="classifier">str</span></dt><dd><p>Path to file.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="coolbox.utilities.format_properties">
<span class="sig-prename descclassname"><span class="pre">coolbox.utilities.</span></span><span class="sig-name descname"><span class="pre">format_properties</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">properties</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.format_properties" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert bool value to ‘yes’ or ‘no’, for compatible with pyGenomeTracks
Convert string to float if possible</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="coolbox.utilities.get_size">
<span class="sig-prename descclassname"><span class="pre">coolbox.utilities.</span></span><span class="sig-name descname"><span class="pre">get_size</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">obj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seen</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.get_size" title="Permalink to this definition">¶</a></dt>
<dd><p>Recursively finds size of objects</p>
<dl class="simple">
<dt>From:</dt><dd><p><a class="reference external" href="https://stackoverflow.com/a/40880923/8500469">https://stackoverflow.com/a/40880923/8500469</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="coolbox.utilities.get_tmp_dir">
<span class="sig-prename descclassname"><span class="pre">coolbox.utilities.</span></span><span class="sig-name descname"><span class="pre">get_tmp_dir</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'.coolbox'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.get_tmp_dir" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the path to coolbox temporary directory.
If the file not exists, will make it.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="coolbox.utilities.get_uniq_tmp_file">
<span class="sig-prename descclassname"><span class="pre">coolbox.utilities.</span></span><span class="sig-name descname"><span class="pre">get_uniq_tmp_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'.svg'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dirname</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'.coolbox'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.get_uniq_tmp_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the path to a unique(not exists) temporary file.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="coolbox.utilities.hex2rgb">
<span class="sig-prename descclassname"><span class="pre">coolbox.utilities.</span></span><span class="sig-name descname"><span class="pre">hex2rgb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">color_hex</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.hex2rgb" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert hex color code to rgb tuple.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">hex2rgb</span><span class="p">(</span><span class="s1">&#39;#819a46&#39;</span><span class="p">)</span>
<span class="go">(129, 154, 70)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="coolbox.utilities.inch2cm">
<span class="sig-prename descclassname"><span class="pre">coolbox.utilities.</span></span><span class="sig-name descname"><span class="pre">inch2cm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">tupl</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.inch2cm" title="Permalink to this definition">¶</a></dt>
<dd><p>convert length unit from inch to cm.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="coolbox.utilities.join">
<span class="sig-prename descclassname"><span class="pre">coolbox.utilities.</span></span><span class="sig-name descname"><span class="pre">join</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">a</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">p</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.join" title="Permalink to this definition">¶</a></dt>
<dd><p>Join two or more pathname components, inserting ‘/’ as needed.
If any component is an absolute path, all previous path components
will be discarded.  An empty last part will result in a path that
ends with a separator.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="coolbox.utilities.op_err_msg">
<span class="sig-prename descclassname"><span class="pre">coolbox.utilities.</span></span><span class="sig-name descname"><span class="pre">op_err_msg</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">a</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">b</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">op</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'+'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.op_err_msg" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate the error message of error operand type.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="coolbox.utilities.opener">
<span class="sig-prename descclassname"><span class="pre">coolbox.utilities.</span></span><span class="sig-name descname"><span class="pre">opener</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.opener" title="Permalink to this definition">¶</a></dt>
<dd><p>Determines if a file is compressed or not</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">gzip</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;hello blablabla&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tmp_f_raw</span>  <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;/tmp/test_opener.txt&quot;</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tmp_f_raw</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
<span class="go">15</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tmp_f_raw</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tmp_f_gzip</span> <span class="o">=</span> <span class="n">gzip</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s1">&#39;/tmp/test_opener.txt.gz&#39;</span><span class="p">,</span> <span class="s1">&#39;wb&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tmp_f_gzip</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">to_bytes</span><span class="p">(</span><span class="n">msg</span><span class="p">))</span>
<span class="go">15</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tmp_f_gzip</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">test_raw</span> <span class="o">=</span> <span class="n">opener</span><span class="p">(</span><span class="n">tmp_f_raw</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">type</span><span class="p">(</span><span class="n">test_raw</span><span class="p">)</span>
<span class="go">&lt;class &#39;_io.BufferedReader&#39;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">test_gzip</span> <span class="o">=</span> <span class="n">opener</span><span class="p">(</span><span class="n">tmp_f_gzip</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">type</span><span class="p">(</span><span class="n">test_gzip</span><span class="p">)</span>
<span class="go">&lt;class &#39;gzip.GzipFile&#39;&gt;</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">test_raw</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">test_gzip</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">os</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">test_raw</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">test_gzip</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.utilities.partial">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.utilities.</span></span><span class="sig-name descname"><span class="pre">partial</span></span><a class="headerlink" href="#coolbox.utilities.partial" title="Permalink to this definition">¶</a></dt>
<dd><p>partial(func, <a href="#id5"><span class="problematic" id="id6">*</span></a>args, <a href="#id7"><span class="problematic" id="id8">**</span></a>keywords) - new function with partial application
of the given arguments and keywords.</p>
<dl class="field-list simple">
<dt class="field-odd">Attributes</dt>
<dd class="field-odd"><dl class="simple">
<dt><a class="reference internal" href="#coolbox.utilities.partial.args" title="coolbox.utilities.partial.args"><code class="xref py py-obj docutils literal notranslate"><span class="pre">args</span></code></a></dt><dd><p>tuple of arguments to future partial calls</p>
</dd>
<dt><a class="reference internal" href="#coolbox.utilities.partial.func" title="coolbox.utilities.partial.func"><code class="xref py py-obj docutils literal notranslate"><span class="pre">func</span></code></a></dt><dd><p>function object to use in future partial calls</p>
</dd>
<dt><a class="reference internal" href="#coolbox.utilities.partial.keywords" title="coolbox.utilities.partial.keywords"><code class="xref py py-obj docutils literal notranslate"><span class="pre">keywords</span></code></a></dt><dd><p>dictionary of keyword arguments to future partial calls</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">__call__</span></code>(*args, **kwargs)</p></td>
<td><p>Call self as a function.</p></td>
</tr>
</tbody>
</table>
<p><strong>Attributes:</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.partial.args" title="coolbox.utilities.partial.args"><code class="xref py py-obj docutils literal notranslate"><span class="pre">args</span></code></a></p></td>
<td><p>tuple of arguments to future partial calls</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#coolbox.utilities.partial.func" title="coolbox.utilities.partial.func"><code class="xref py py-obj docutils literal notranslate"><span class="pre">func</span></code></a></p></td>
<td><p>function object to use in future partial calls</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#coolbox.utilities.partial.keywords" title="coolbox.utilities.partial.keywords"><code class="xref py py-obj docutils literal notranslate"><span class="pre">keywords</span></code></a></p></td>
<td><p>dictionary of keyword arguments to future partial calls</p></td>
</tr>
</tbody>
</table>
<dl class="py attribute">
<dt class="sig sig-object py" id="coolbox.utilities.partial.args">
<span class="sig-name descname"><span class="pre">args</span></span><a class="headerlink" href="#coolbox.utilities.partial.args" title="Permalink to this definition">¶</a></dt>
<dd><p>tuple of arguments to future partial calls</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="coolbox.utilities.partial.func">
<span class="sig-name descname"><span class="pre">func</span></span><a class="headerlink" href="#coolbox.utilities.partial.func" title="Permalink to this definition">¶</a></dt>
<dd><p>function object to use in future partial calls</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="coolbox.utilities.partial.keywords">
<span class="sig-name descname"><span class="pre">keywords</span></span><a class="headerlink" href="#coolbox.utilities.partial.keywords" title="Permalink to this definition">¶</a></dt>
<dd><p>dictionary of keyword arguments to future partial calls</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="coolbox.utilities.query_bed">
<span class="sig-prename descclassname"><span class="pre">coolbox.utilities.</span></span><span class="sig-name descname"><span class="pre">query_bed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">split</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.query_bed" title="Permalink to this definition">¶</a></dt>
<dd><p>Call tabix and generate an array of strings for each line it returns.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="coolbox.utilities.refGeneRec">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">coolbox.utilities.</span></span><span class="sig-name descname"><span class="pre">refGeneRec</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bin</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">txStart</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">txEnd</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cdsStart</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cdsEnd</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exonCount</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exonStart</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exonEnds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cdsStartStat</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cdsEndStat</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exonFrames</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.refGeneRec" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Attributes</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>bin</strong></dt><dd><p>Alias for field number 0</p>
</dd>
<dt><strong>cdsEnd</strong></dt><dd><p>Alias for field number 7</p>
</dd>
<dt><strong>cdsEndStat</strong></dt><dd><p>Alias for field number 14</p>
</dd>
<dt><strong>cdsStart</strong></dt><dd><p>Alias for field number 6</p>
</dd>
<dt><strong>cdsStartStat</strong></dt><dd><p>Alias for field number 13</p>
</dd>
<dt><strong>chrom</strong></dt><dd><p>Alias for field number 2</p>
</dd>
<dt><strong>exonCount</strong></dt><dd><p>Alias for field number 8</p>
</dd>
<dt><strong>exonEnds</strong></dt><dd><p>Alias for field number 10</p>
</dd>
<dt><strong>exonFrames</strong></dt><dd><p>Alias for field number 15</p>
</dd>
<dt><strong>exonStart</strong></dt><dd><p>Alias for field number 9</p>
</dd>
<dt><strong>name</strong></dt><dd><p>Alias for field number 1</p>
</dd>
<dt><strong>name2</strong></dt><dd><p>Alias for field number 12</p>
</dd>
<dt><strong>score</strong></dt><dd><p>Alias for field number 11</p>
</dd>
<dt><strong>strand</strong></dt><dd><p>Alias for field number 3</p>
</dd>
<dt><strong>txEnd</strong></dt><dd><p>Alias for field number 5</p>
</dd>
<dt><strong>txStart</strong></dt><dd><p>Alias for field number 4</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">count</span></code>(value, /)</p></td>
<td><p>Return number of occurrences of value.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">index</span></code>(value[, start, stop])</p></td>
<td><p>Return first index of value.</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 64%" />
<col style="width: 36%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>get_exons_size</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>offset_zero</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>to_bed12_line</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>to_line</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="coolbox.utilities.rgb2hex">
<span class="sig-prename descclassname"><span class="pre">coolbox.utilities.</span></span><span class="sig-name descname"><span class="pre">rgb2hex</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">r</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">g</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">b</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.rgb2hex" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert rgb color to hex format.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">rgb2hex</span><span class="p">(</span><span class="mi">129</span><span class="p">,</span> <span class="mi">154</span><span class="p">,</span> <span class="mi">70</span><span class="p">)</span>
<span class="go">&#39;#819a46&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rgb2hex</span><span class="p">(</span><span class="o">-</span><span class="mi">10</span><span class="p">,</span> <span class="mi">256</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
<span class="c">...</span>
<span class="gr">AssertionError</span>: <span class="n">(r, g, b) value must within range 0 ~ 255.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="coolbox.utilities.shiftedColorMap">
<span class="sig-prename descclassname"><span class="pre">coolbox.utilities.</span></span><span class="sig-name descname"><span class="pre">shiftedColorMap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmap</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">midpoint</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'shiftedcmap'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.shiftedColorMap" title="Permalink to this definition">¶</a></dt>
<dd><p>This function is from: <a class="reference external" href="https://stackoverflow.com/a/20528097/8500469">https://stackoverflow.com/a/20528097/8500469</a></p>
<p>Function to offset the “center” of a colormap. Useful for
data with a negative min and positive max and you want the
middle of the colormap’s dynamic range to be at zero</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="coolbox.utilities.split_genome_range">
<span class="sig-prename descclassname"><span class="pre">coolbox.utilities.</span></span><span class="sig-name descname"><span class="pre">split_genome_range</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome_range</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.split_genome_range" title="Permalink to this definition">¶</a></dt>
<dd><p>Little helper func.
enforce gr is a GenomeRange object, and
split gr to chrom, start, end</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="coolbox.utilities.tabix_index">
<span class="sig-prename descclassname"><span class="pre">coolbox.utilities.</span></span><span class="sig-name descname"><span class="pre">tabix_index</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">preset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gff'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.tabix_index" title="Permalink to this definition">¶</a></dt>
<dd><p>Call tabix to create an index for a bgzip-compressed file.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="coolbox.utilities.tabix_query">
<span class="sig-prename descclassname"><span class="pre">coolbox.utilities.</span></span><span class="sig-name descname"><span class="pre">tabix_query</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">split</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.tabix_query" title="Permalink to this definition">¶</a></dt>
<dd><p>Call tabix and generate an array of strings for each line it returns.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="coolbox.utilities.to_bytes">
<span class="sig-prename descclassname"><span class="pre">coolbox.utilities.</span></span><span class="sig-name descname"><span class="pre">to_bytes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.to_bytes" title="Permalink to this definition">¶</a></dt>
<dd><p>Like toString, to bytes.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">to_bytes</span><span class="p">(</span><span class="s1">&#39;hello&#39;</span><span class="p">)</span>
<span class="go">b&#39;hello&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">to_bytes</span><span class="p">([</span><span class="s1">&#39;hello&#39;</span><span class="p">,</span> <span class="s1">&#39;world&#39;</span><span class="p">])</span>
<span class="go">[b&#39;hello&#39;, b&#39;world&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="coolbox.utilities.to_gr">
<span class="sig-prename descclassname"><span class="pre">coolbox.utilities.</span></span><span class="sig-name descname"><span class="pre">to_gr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">obj</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.to_gr" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert object to GenomeRange.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">to_gr</span><span class="p">(</span><span class="s2">&quot;chr1:0-1000&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">(</span><span class="n">gr</span><span class="o">.</span><span class="n">chrom</span><span class="p">,</span> <span class="n">gr</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">gr</span><span class="o">.</span><span class="n">end</span><span class="p">)</span>
<span class="go">(&quot;chr1&quot;, 0, 1000)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="coolbox.utilities.to_string">
<span class="sig-prename descclassname"><span class="pre">coolbox.utilities.</span></span><span class="sig-name descname"><span class="pre">to_string</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coolbox.utilities.to_string" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert bytes, bytes list to string, string list.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">to_string</span><span class="p">(</span><span class="s2">&quot;hello&quot;</span><span class="p">)</span>
<span class="go">&#39;hello&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">to_string</span><span class="p">(</span><span class="sa">b</span><span class="s2">&quot;hello&quot;</span><span class="p">)</span>
<span class="go">&#39;hello&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">to_string</span><span class="p">([</span><span class="sa">b</span><span class="s1">&#39;hello&#39;</span><span class="p">,</span> <span class="sa">b</span><span class="s1">&#39;world&#39;</span><span class="p">])</span>
<span class="go">[&#39;hello&#39;, &#39;world&#39;]</span>
</pre></div>
</div>
</dd></dl>

</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="custom_elements.html" class="btn btn-neutral float-left" title="Customizing CoolBox" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, WeizeXu.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>